---
author: 
  "Jeanie Na"
date: "`r format(Sys.time(), '%d %B %Y')`"
title    : "High throughput study analysis one experiment investigation"


output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    theme: united
    highlight: tango

params:

  Gene:
    label: "Gene:"
    value: "BRCA2"
    input: text 
  Input_dir:
    label: "Input Dir:"
    value: "###results/11_2023/Ind_full/"
    input: text 
  Output_dir:
    label: "Output Dir:"
    value: "###/results/11_2023/xexon_full1526_fc1_exf_Yes_nonsense_syn_exclude_single/"
    input: text    
       
  Exons:
    label: "Exons included:"
    value: "15, 16, 17, 18C, 18N, 19, 20, 21, 22, 23, 24, 25C, 25N, 26"
    input: text   
      
  loessSubset:
    label: "Loess Filter (keep):"
    value: "ModFindlay2"
    choices: ["Findlay", "ModFindlay", "ModFindlay2"]
    input: select
    
  EventCount_Filter:
    label: "EventCount Filter:"
    value: "libD510"
    choices: ["lib10", "libD510"]
    input: select
  vset:
    label: "Variant set:"
    value: "SNV"
    choices: ["allv", "SNV"]
    input: select

  Score_method:
    label: "Score Method:"
    value: "C"
    choices: ["A", "B", "C"]
    input: select
  Lab_curation:
    label: "Lab Curattion:"
    value: "No"
    choices: ["Yes", "No"]
    input: select
  Ratio:
    label: "Ratio:"
    value: "D14vD5"
    choices: ["D14vLib", "D14vD5"]
    input: select
  MM:
    label: "Miture Model:"
    value: "normal"
    choices: ["Robust", "normal"]
    input: select
  Outlierremove:
    label: "Outlierremove Method:"
    value: "No"
    choices: ["p1", "p2", "p5", "No"]
    input: select
  extraFilter:
    label: "Supplementary filter:"
    value: "Yes"
    choices: ["Yes", "No"]
    input: select
    
  clinvarfile:
    label: "Clinvar File:"
    value: "###data/10_2023_curation/ClinVar_BRCA2_DBD_control_111523_nonsense_syn.txt"
    choices: ["###/data/10_2023_curation/ClinVar_BRCA2_DBD_control_111523.xlsx", "###/data/10_2023_curation/ClinVar_BRCA2_DBD_control_111523_nonsense_syn.txt"]
    input: select
    
---

```{=html}

<style type="text/css">
.main-container {
  max-width: 1200px;
  margin-left: auto;
  margin-right: auto;
}
</style>
```
<link rel="stylesheet" type="text/css" href="http://cdn.datatables.net/1.10.5/css/jquery.dataTables.min.css">

```{=html}
<script type="text/javascript">
  $(document).ready(function() {
    $("table").DataTable();
  } );
</script>
```

```{r setup, include=TRUE, warning=FALSE, message=FALSE}
    
knitr::opts_chunk$set(echo = TRUE, warning=FALSE)
library(kableExtra)
library(plotly)
library(arsenal)
library(boot)

library(ggplot2)
library(dplyr)
library(tidyr)
library(readr)
library(readxl)
library(tidyverse)
library(threejs)
library(pROC)


dtdir <- params$Input_dir
resdir <- params$Output_dir
Score_method <- params$Score_method
Lab_curation <- params$Lab_curation
vset <- params$vset
Exons <- params$Exons
Ratio <- params$Ratio
Gene <- params$Gene
extraFilter <- params$extraFilter
```
## Setting of this report

Gene: `r  Gene`\
Exons included : `r Exons`\
Outputdir: `r  resdir`\
loessSubset:  `r params$loessSubset`\
EventCount_Filter: `r params$EventCount_Filter`\
vset: `r vset`\
Score_method: `r Score_method`\
Lab_curation: `r Lab_curation`\
Ratio: `r Ratio`\
Mixture Model: `r params$MM`\
Outlierremove: `r params$Outlierremove`\
clinvarfile: `r params$clinvarfile`
extraFilter:  `r params$extraFilter`

## Aim

Apply Findlay methods on functional scores. `r Ratio`.

```{r}

Varaints_to_be_excluded_when_use_MAVE_internal_control_to_build_model <- read_excel("###/data/10_2023_curation/Varaints to be excluded when use MAVE internal control to build model.xlsx")

Varaints_to_be_excluded_when_use_MAVE_internal_control_to_build_model <- Varaints_to_be_excluded_when_use_MAVE_internal_control_to_build_model %>% mutate(uPOS = paste(GRCh38Location-1, REF, ALT, sep = "_") )
                    
```

```{r readin, include=TRUE, warning=FALSE, message=FALSE}
## read in data
if(Ratio == "D14vLib"){ 

  inputfiles <- list.files(dtdir, pattern = glob2rx(paste0("*_", params$vset, "_", params$loessSubset, "_", params$EventCount_Filter, paste0("*", Ratio, ".csv"))))
  E150 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", 15), inputfiles)]))
  E160 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", 16), inputfiles)]))
  
  E170 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", 17), inputfiles)]))
  E18C0 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", "18C"), inputfiles)]))
  
  E18N0 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", "18N"), inputfiles)]))
  E190 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", 19), inputfiles)]))
  E200 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", 20), inputfiles)]))
  E230 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", 23,"_sg4_3"), inputfiles)])) ## come back here, might be a glitch in future
  
  E240 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", 24), inputfiles)]))
  E25C0 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", "25C"), inputfiles)]))
  E25N0 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", "25N"), inputfiles)]))
  
  E210 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", "21"), inputfiles)]))
  E220 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", "22"), inputfiles)]))
  
  E260 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", "26_sg4_3"), inputfiles)]))
}else{
  inputfiles <- list.files(dtdir, pattern = glob2rx(paste0("*_", params$vset, "_", params$loessSubset, "_", params$EventCount_Filter, paste0("*", Ratio, ".csv"))))
  E150 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", 15), inputfiles)]))
  E160 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", 16), inputfiles)]))
  
  E170 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", 17), inputfiles)]))
  E18C0 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", "18C"), inputfiles)]))
  
  E18N0 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", "18N"), inputfiles)]))
  E190 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", 19), inputfiles)]))
  E200 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", 20), inputfiles)]))
  E230 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", 23,"_sg4_3"), inputfiles)]))

  E240 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", 24), inputfiles)]))
  E25C0 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", "25C"), inputfiles)]))
  E25N0 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", "25N"), inputfiles)]))
  
  E210 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", "21"), inputfiles)]))
  E220 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", "22"), inputfiles)]))
  
  E260 <- read_csv( paste0(dtdir, inputfiles[grep(paste0("E", "26_sg4_3"), inputfiles)]))
}

```

## Data

This analysis does `r ifelse(Lab_curation == "Yes", "", "not" )` take Lab curation into account. 

## Analysis Plan:

0. Start with the adjusted ratio from each individual exon exploration. 

1. run the pipeline with normalization within each exon and across exons including all lab curated and cleaned data files. Files used are `r dtdir`

`r inputfiles`

Reference Method

589 Normalizing scores within and across exons  

590 Position-adjusted log2 day 11 over library ratios were normalized first across exon

591 replicates, and then across all exons assayed. Scores from within each replicate were linearly 

592 scaled such that the median synonymous and median nonsense SNVs within the replicate were 

593 set to the median synonymous and median nonsense SNV values averaged across replicate 

594 experiments. The ensuing SNV scores for each replicate were then normalized across exons in 

595 the same way by again using median synonymous and median nonsense SNVs.


A. Within each exon:

Positional adjusted D14vLib or D14vD5 ratios from within each replicate were linearly scaled such that the median synonymous and median nonsense SNVs within the replicate were set to the median synonymous and median nonsense SNV values averaged across replicate  experiments.

B. Normalization across exons.

The ensuing SNV scores for each replicate were then normalized across exons in  the same way by again using median synonymous and median nonsense SNVs.

C. functional score

Get the average functional score cross replicates.

2. Then calculate both 95% confidence interval (CI) and 99% CI for all stop-gain variants based on the clean data (D8/D5 and D8_olaparib/D5); 95% CI and 99% CI for all synonymous variants based on the clean data (D8/D5 and D8_olaparib/D5). For missense variants, no CI calculation needed, only functional scores from D8/D5 and D8_olaparib/D5. Classify hypomorphs based on the 95% & 99% CI from stop-gains and synonymous.

3. Calculate the p-values for the pairwise mean difference for the 3 categories as a group (between stop-gain vs hypomorphs, synonymous vs hypomorphs, stop-gain vs synonymous) for the no drug experiment and olaparib experiment. Then, once the pairwise mean differences are calculated within no-drug or olaparib experiment, compare the pairwise mean difference between no-drug experiment and olaparib experiment. 

4 Added the mixture model part, also the posterior probability 1-99% 2-98% corresponding functional score. 

```{r prepdt}
if(Ratio == "D14vD5"){
  E15 <- E150 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E15") %>% mutate( AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R3 = as.numeric(R3_D14_D5), R1_raw = as.numeric(R1_D14_D5), R2_raw = as.numeric(R2_D14_D5), R3_raw = as.numeric(R3_D14_D5))
  E16 <- E160 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E16") %>% mutate( AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R3 = as.numeric(R3_D14_D5), R1_raw = as.numeric(R1_D14_D5), R2_raw = as.numeric(R2_D14_D5), R3_raw = as.numeric(R3_D14_D5))
  
  E17 <- E170 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E17") %>% mutate( AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R3 = as.numeric(R3_D14_D5),  R1_raw = as.numeric(R1_D14_D5), R2_raw = as.numeric(R2_D14_D5), R3_raw = as.numeric(R3_D14_D5))
  E18C <- E18C0 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E18C") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R3 = as.numeric(R3_D14_D5), R1_raw = as.numeric(R1_D14_D5), R2_raw = as.numeric(R2_D14_D5), R3_raw = as.numeric(R3_D14_D5)) 
  E18N <- E18N0 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E18N") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R1_raw = as.numeric(R1_D14_D5), R2_raw = as.numeric(R2_D14_D5))
  E19 <- E190 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E19") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R3 = as.numeric(R3_D14_D5), R1_raw = as.numeric(R1_D14_D5), R2_raw = as.numeric(R2_D14_D5), R3_raw = as.numeric(R3_D14_D5))
  E20 <- E200 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E20") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R3 = as.numeric(R3_D14_D5), R1_raw = as.numeric(R1_D14_D5), R2_raw = as.numeric(R2_D14_D5), R3_raw = as.numeric(R3_D14_D5))
  E24 <- E240 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E24") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R3 = as.numeric(R3_D14_D5), R1_raw = as.numeric(R1_D14_D5), R2_raw = as.numeric(R2_D14_D5), R3_raw = as.numeric(R3_D14_D5))
  E25C <- E25C0 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E25C") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R3 = as.numeric(R3_D14_D5), R1_raw = as.numeric(R1_D14_D5), R2_raw = as.numeric(R2_D14_D5), R3_raw = as.numeric(R3_D14_D5))
  E25N <- E25N0 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E25N") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R3 = as.numeric(R3_D14_D5), R1_raw = as.numeric(R1_D14_D5), R2_raw = as.numeric(R2_D14_D5), R3_raw = as.numeric(R3_D14_D5))
  E21 <- E210 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E21") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R3 = as.numeric(R3_D14_D5), R1_raw = as.numeric(R1_D14_D5), R2_raw = as.numeric(R2_D14_D5), R3_raw = as.numeric(R3_D14_D5))
  E22 <- E220 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E22") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R3 = as.numeric(R3_D14_D5), R1_raw = as.numeric(R1_D14_D5), R2_raw = as.numeric(R2_D14_D5), R3_raw = as.numeric(R3_D14_D5)) 
  # E23 <- E230 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E23") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R3 = as.numeric(R3_D14_D5), R4 = as.numeric(R4_D14_D5), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3, R4_D14_lib = as.numeric(R4_D14_lib)) 
    E23 <- E230 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E23") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R3 = as.numeric(R3_D14_D5), R1_raw = as.numeric(R1_D14_D5), R2_raw = as.numeric(R2_D14_D5), R3_raw = as.numeric(R3_D14_D5)) 
  # E26 <- E260 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E26") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R3 = as.numeric(R3_D14_D5),R4 = as.numeric(R4_D14_D5), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3, R4_D14_lib = as.numeric(R4_D14_lib)) 
  E26 <- E260 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E26") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(R1_D14_D5), R2 = as.numeric(R2_D14_D5), R3 = as.numeric(R3_D14_D5), R1_raw = as.numeric(R1_D14_D5), R2_raw = as.numeric(R2_D14_D5), R3_raw = as.numeric(R3_D14_D5)) 
}else{
  E15 <- E150 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E15") %>% mutate( AApos=as.numeric(AApos), R1 = as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R3 = as.numeric(D14_lib_ratio_adjusted_R3), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3)
  E16 <- E160 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E16") %>% mutate( AApos=as.numeric(AApos), R1 = as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R3 = as.numeric(D14_lib_ratio_adjusted_R3), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3)
  
  E17 <- E170 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E17") %>% mutate( AApos=as.numeric(AApos), R1 = as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R3 = as.numeric(D14_lib_ratio_adjusted_R3), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3)
  E18C <- E18C0 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E18C") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R3 = as.numeric(D14_lib_ratio_adjusted_R3), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3) 
  E18N <- E18N0 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E18N") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R1_loess = R1, R2_loess = R2)
  E19 <- E190 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E19") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R3 = as.numeric(D14_lib_ratio_adjusted_R3), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3)
  E20 <- E200 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E20") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R3 = as.numeric(D14_lib_ratio_adjusted_R3), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3)
  E24 <- E240 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E24") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R3 = as.numeric(D14_lib_ratio_adjusted_R3), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3)
  E25C <- E25C0 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E25C") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R3 = as.numeric(D14_lib_ratio_adjusted_R3), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3)
  E25N <- E25N0 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E25N") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R3 = as.numeric(D14_lib_ratio_adjusted_R3), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3)
  E21 <- E210 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E21") %>% mutate(AApos=as.numeric(AApos), R1 =as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R3 = as.numeric(D14_lib_ratio_adjusted_R3), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3)
  E22 <- E220 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E22") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R3 = as.numeric(D14_lib_ratio_adjusted_R3), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3) 

  # E23 <- E230 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E23") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R3 = as.numeric(D14_lib_ratio_adjusted_R3), R4 = as.numeric(D14_lib_ratio_adjusted_R4), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3, R4_D14_lib = as.numeric(R4_D14_lib)) 
  # 
  # E26 <- E260 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E26") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R3 = as.numeric(D14_lib_ratio_adjusted_R3), R4 = as.numeric(D14_lib_ratio_adjusted_R4), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3, R4_D14_lib = as.numeric(R4_D14_lib)) 

  E23 <- E230 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E23") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R3 = as.numeric(D14_lib_ratio_adjusted_R3), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3) 

  E26 <- E260 %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_"), Exon = "E26") %>% mutate(AApos=as.numeric(AApos), R1 = as.numeric(D14_lib_ratio_adjusted_R1), R2 = as.numeric(D14_lib_ratio_adjusted_R2), R3 = as.numeric(D14_lib_ratio_adjusted_R3), R1_raw = as.numeric(R1_D14_lib), R2_raw = as.numeric(R2_D14_lib), R3_raw = as.numeric(R3_D14_lib), R1_loess = R1, R2_loess = R2, R3_loess = R3) 
  
}

if(Lab_curation == "Yes"){ ## remove lab says exclude all
  if(Score_method == "A"){
    E15 <- E15 %>% filter(!`lab CV A` %in% c("exclude all"))
    E16 <- E16 %>% filter(!`lab CV A` %in% c("exclude all"))
    E17 <- E17 %>% filter(!`lab CV A` %in% c("exclude all"))
    E18C <- E18C %>% filter(!`lab CV A` %in% c("exclude all"))
    E18N <- E18N %>% filter(!`lab CV A` %in% c("exclude all"))
    E19 <- E19 %>% filter(!`lab CV A` %in% c("exclude all"))
    E20 <- E20 %>% filter(!`lab CV A` %in% c("exclude all"))
    E21 <- E21 %>% filter(!`lab CV A` %in% c("exclude all"))
    E22 <- E22 %>% filter(!`lab CV A` %in% c("exclude all"))
    E24 <- E24 %>% filter(!`lab CV A` %in% c("exclude all"))
    E25C <- E25C %>% filter(!`lab CV A` %in% c("exclude all"))
    E25N <- E25N %>% filter(!`lab CV A` %in% c("exclude all"))
    
    E26 <- E26 %>% filter(!`lab CV A` %in% c("exclude all"))
  }else if(Score_method == "B"){
    E15 <- E15 %>% filter(!`lab CV B` %in% c("exclude all"))
    E16 <- E16 %>% filter(!`lab CV B` %in% c("exclude all"))
    E17 <- E17 %>% filter(!`lab CV B` %in% c("exclude all"))
    E18C <- E18C %>% filter(!`lab CV B` %in% c("exclude all"))
    E18N <- E18N %>% filter(!`lab CV B` %in% c("exclude all"))
    E19 <- E19 %>% filter(!`lab CV B` %in% c("exclude all"))
    E20 <- E20 %>% filter(!`lab CV B` %in% c("exclude all"))
    E21 <- E21 %>% filter(!`lab CV B` %in% c("exclude all"))
    E22 <- E22 %>% filter(!`lab CV B` %in% c("exclude all"))
    E24 <- E24 %>% filter(!`lab CV B` %in% c("exclude all"))
    E25C <- E25C %>% filter(!`lab CV B` %in% c("exclude all"))
    E25N <- E25N %>% filter(!`lab CV B` %in% c("exclude all"))
    E26 <- E26 %>% filter(!`lab CV B` %in% c("exclude all"))
  }else if(Score_method == "C"){
    E15 <- E15 %>% filter(!`lab CV C` %in% c("exclude all"))
    E16 <- E16 %>% filter(!`lab CV C` %in% c("exclude all"))
    E17 <- E17 %>% filter(!`lab CV C` %in% c("exclude all"))
    E18C <- E18C %>% filter(!`lab CV C` %in% c("exclude all"))
    E18N <- E18N %>% filter(!`lab CV C` %in% c("exclude all"))
    E19 <- E19 %>% filter(!`lab CV C` %in% c("exclude all"))
    E20 <- E20 %>% filter(!`lab CV C` %in% c("exclude all"))
    E21 <- E21 %>% filter(!`lab CV C` %in% c("exclude all"))
    E22 <- E22 %>% filter(!`lab CV C` %in% c("exclude all"))
    E24 <- E24 %>% filter(!`lab CV C` %in% c("exclude all"))
    E25C <- E25C %>% filter(!`lab CV C` %in% c("exclude all"))
    E25N <- E25N %>% filter(!`lab CV C` %in% c("exclude all"))
    E23 <- E23 %>% filter(!`lab CV C` %in% c("exclude all"))
    E26 <- E26 %>% filter(!`lab CV C` %in% c("exclude all"))
  }
}

prefilter3A <- function(data,extraFilter ){
  wide2 <- data 
    if(length(which(wide2$removeA1==TRUE)) > 0){
      wide2$R1[which(wide2$removeA1==TRUE)] <- NA
    }
    if(length(which(wide2$removeA2==TRUE)) > 0){
      wide2$R2[which(wide2$removeA2==TRUE)] <- NA
    }
    if(length(which(wide2$removeA3==TRUE)) > 0){
      wide2$R3[which(wide2$removeA3==TRUE)] <- NA
    }
  
  if(extraFilter == "No"){

  }else{
    ## remove variants if the mean differ from B 
    wide2 <- wide2 %>% filter(abs(functional.scoreA-functional.scoreB) < 1)
  }
  return(wide2)
}

prefilter3C <- function(data, extraFilter){
  wide2 <- data 
  if(length(which(wide2$removeC1==TRUE)) > 0){
      wide2$R1[which(wide2$removeC1==TRUE)] <- NA
    }
    if(length(which(wide2$removeC2==TRUE)) > 0){
      wide2$R2[which(wide2$removeC2==TRUE)] <- NA
    }
    if(length(which(wide2$removeC3==TRUE)) > 0){
      wide2$R3[which(wide2$removeC3==TRUE)] <- NA
    }
  
  if(extraFilter == "No"){
    wide2 <- wide2
  }else{
    ## remove variants if the mean differ from B 
    wide2 <- wide2 %>% filter(abs(functional.scoreA-functional.scoreB) < 1)
  }
  return(wide2)
}

prefilter4A <- function(data, extraFilter){
    wide2 <- data
  if(extraFilter == "No"){
    if(length(which(wide2$removeA1==TRUE)) > 0){
      wide2$R1[which(wide2$removeA1==TRUE)] <- NA
    }
    if(length(which(wide2$removeA2==TRUE)) > 0){
      wide2$R2[which(wide2$removeA2==TRUE)] <- NA
    }
    if(length(which(wide2$removeA3==TRUE)) > 0){
      wide2$R3[which(wide2$removeA3==TRUE)] <- NA
    }
    if(length(which(wide2$removeA4==TRUE)) > 0){
      wide2$R4[which(wide2$removeA4==TRUE)] <- NA
    }
  }else{
    ## remove variants if the mean differ from B 
    wide2 <- wide2 %>% filter(abs(functional.scoreC-functional.scoreB) < 1)
  }
  return(wide2)
}

prefilter4C <- function(data, extraFilter){
  wide2 <- data 
  if(extraFilter == "No"){
    if(length(which(wide2$removeC1==TRUE)) > 0){
      wide2$R1[which(wide2$removeC1==TRUE)] <- NA
    }
    if(length(which(wide2$removeC2==TRUE)) > 0){
      wide2$R2[which(wide2$removeC2==TRUE)] <- NA
    }
    if(length(which(wide2$removeC3==TRUE)) > 0){
      wide2$R3[which(wide2$removeC3==TRUE)] <- NA
    }
    if(length(which(wide2$removeC4==TRUE)) > 0){
      wide2$R4[which(wide2$removeC4==TRUE)] <- NA
    }
  
  }else{
    ## remove variants if the mean differ from B 
    wide2 <- data
    wide2 <- wide2 %>% filter(abs(functional.scoreC-functional.scoreB) < 1)
  }
  return(wide2)
}

if(Score_method == "A"){
  E151 <- prefilter3A(E15, extraFilter)
  E161 <- prefilter3A(E16, extraFilter)
  
  E171 <- prefilter3A(E17, extraFilter)
  E18C1 <- prefilter3A(E18C, extraFilter)
  if(length(which(E18N$removeA == TRUE))>0){
    E18N1 <- E18N[-which(E18N$removeA == TRUE),]
  }else{
      E18N1 <- E18N
  }
  
  E191 <- prefilter3A(E19, extraFilter)
  E201 <- prefilter3A(E20, extraFilter)
  E211 <- prefilter3A(E21, extraFilter)
  E221 <- prefilter3A(E22, extraFilter)
  E241 <- prefilter3A(E24, extraFilter)
  E25C1 <- prefilter3A(E25C, extraFilter)
  E25N1 <- prefilter3A(E25N, extraFilter)
  
  E231 <- prefilter4A(E23, extraFilter)
  E261 <- prefilter4A(E26, extraFilter)
} else if(Score_method == "C"){
  E151 <- prefilter3C(E15, extraFilter)
  E161 <- prefilter3C(E16, extraFilter)
  E171 <- prefilter3C(E17, extraFilter)
  E18C1 <- prefilter3C(E18C, extraFilter)
  if(length(which(E18N$removeC == TRUE))>0){
    E18N1 <- E18N[-which(E18N$removeC == TRUE),]
  }else{
      E18N1 <- E18N
  }
  E191 <- prefilter3C(E19, extraFilter)
  E201 <- prefilter3C(E20, extraFilter)
  E211 <- prefilter3C(E21, extraFilter)
  E221 <- prefilter3C(E22, extraFilter)
  E241 <- prefilter3C(E24, extraFilter)
  E25C1 <- prefilter3C(E25C, extraFilter)
  E25N1 <- prefilter3C(E25N, extraFilter)
  E231 <- prefilter4C(E23, extraFilter)
  E261 <- prefilter4C(E26, extraFilter)

}else{
  E151 <- E15; E161 <- E16; E171 <- E17; E18C1 <- E18C; E18N1 <- E18N; E191 <- E19; E201 <- E20; E211 <- E21; E221 <- E22; E241 <- E24; E25C1 <- E25C; E25N1 <- E25N; E231 <- E23; E261 <- E26
}

colns <- c("Exon", "uPOS", "POS", "REF",                       
"ALT", "AApos","AltAA", "AltCodon", "EventType", #"EventType2",  
"RefAA", "RefCodon", "SpliceAI_ALLELE", "SpliceAI_DP_AG", "SpliceAI_DP_AL", "SpliceAI_DP_DG",  "SpliceAI_DP_DL", "SpliceAI_DS_AG",  "SpliceAI_DS_AL", "SpliceAI_DS_DG", "SpliceAI_DS_DL", "spliceR",  "R1", "R2", "R3", "R4", "functional.scoreA", "functional.scoreB",    "functional.scoreC")

rawcolns <- c("uPOS", "Exon", "R1_raw", "R2_raw", "R3_raw", "R4_raw", "R1_loess", "R2_loess", "R3_loess", "R4_loess")

dt0 <- dplyr::bind_rows(E151[, rawcolns[rawcolns %in% colnames(E151)]], 
                        E161[, rawcolns[rawcolns %in% colnames(E161)]], 
                        E171[, rawcolns[rawcolns %in% colnames(E171)]], 
                        E18C1[, rawcolns[rawcolns %in% colnames(E18C1)]], 
                        E18N1[, rawcolns[rawcolns %in% colnames(E18N1)]], 
                        E191[, rawcolns[rawcolns %in% colnames(E191)]],
                        E201[, rawcolns[rawcolns %in% colnames(E201)]],
                        E211[, rawcolns[rawcolns %in% colnames(E211)]],
                        E221[, rawcolns[rawcolns %in% colnames(E221)]],
                        E241[, rawcolns[rawcolns %in% colnames(E241)]],
                        E25C1[, rawcolns[rawcolns %in% colnames(E25C1)]],
                        E25N1[, rawcolns[rawcolns %in% colnames(E25N1)]],
                        E231[, rawcolns[rawcolns %in% colnames(E231)]],
                        E261[, rawcolns[rawcolns %in% colnames(E261)]])
dt <- dplyr::bind_rows(E151[, colns[colns %in% colnames(E151)]], 
                       E161[, colns[colns %in% colnames(E161)]], 
                       E171[, colns[colns %in% colnames(E171)]], 
                       E18C1[, colns[colns %in% colnames(E18C1)]], 
                       E18N1[, colns[colns %in% colnames(E18N1)]], 
                       E191[, colns[colns %in% colnames(E191)]],
                       E201[, colns[colns %in% colnames(E201)]],
                       E211[, colns[colns %in% colnames(E211)]],
                       E221[, colns[colns %in% colnames(E221)]],
                       E241[, colns[colns %in% colnames(E241)]],
                       E25C1[, colns[colns %in% colnames(E25C1)]],
                       E25N1[, colns[colns %in% colnames(E25N1)]],
                       E231[, colns[colns %in% colnames(E231)]],
                       E261[, colns[colns %in% colnames(E261)]])

exonslist <- paste0("E", unlist(strsplit(Exons, ", ")))

## subset to Exons
dt0 <- dt0 %>% filter(Exon %in% exonslist) %>% select(-Exon)
dt <- dt %>% filter(Exon %in% exonslist)

if(any(c("E23_4", "E26_4") %in% exonslist)){
  dt.long <- tidyr::gather(dt, key = "Experiment", value = "score0", R1, R2, R3, R4)
}else{
  dt.long <- tidyr::gather(dt, key = "Experiment", value = "score0", R1, R2, R3)
}


dt.long$Experiment <- as.numeric(substr(dt.long$Experiment, 2, 2))
dt.long$score0 <- as.numeric(dt.long$score0)

mydt <- dt.long 

## within target region
target <- read.csv("###/data/10_2023_curation/chunling_target.csv")
target$Exon <- gsub("_sg.*_*", "", target$Exon)
target <- target %>% unique()

mydt <- merge(mydt, target) %>% mutate(inTarget = ifelse(POS <= End & POS >= Start, TRUE, FALSE))

```


```{r funs, include=TRUE}


##a = (Syn.Avg.med-StopGain.Avg.med)/(Syn.medi-StopGain.medi),b =Syn.Avg.med â€“ a*Syn.medi 

myscale <- function(x, Syn.Avg.med, StopGain.Avg.med, Syn.med1, StopGain.med1){
  a <- (Syn.Avg.med-StopGain.Avg.med)/(Syn.med1-StopGain.med1)
  b  <- Syn.Avg.med - a*Syn.med1
  x2 <- a*x +b
  return(x2)
}

norm_win_exon <- function(dt){ ## note: remove splice variants for calculation medians inside so to keep the splice variants later in the normalization.
  dt.out <- NULL
  
  for (ei in unique(dt$Exon)){
    dti <- dt %>% filter(Exon == ei)
    meds <- dti %>% ## remove variants in splicing region
      filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>% 
      filter(inTarget == TRUE) %>% 
      filter(!uPOS %in% Varaints_to_be_excluded_when_use_MAVE_internal_control_to_build_model$uPOS) %>% ## remove variants on chunling's list
      group_by(EventType, Experiment) %>%
      dplyr::summarize(M = median(score0, na.rm=TRUE))
    
    meds2  <- meds %>%
    group_by(EventType) %>%
    dplyr::summarize(M2 = mean(M, na.rm=TRUE))
    
    dti$score1 <- NA
    for(ri in 1:max(dt$Experiment)){
      dti[dti$Experiment == ri, "score1"] <- myscale(x = dti[dti$Experiment == ri, "score0"],
                                                                   Syn.Avg.med = meds2$M2[meds2$EventType == "Synonymous"],
                                                                   StopGain.Avg.med = meds2$M2[meds2$EventType == "StopGain"],  
                                                                   Syn.med1 =  meds$M[meds$EventType == "Synonymous" & meds$Experiment == ri],
                                                                   StopGain.med1 = meds$M[meds$EventType == "StopGain" & meds$Experiment == ri])
    }
    dt.out <- rbind(dt.out, dti)
  }
  return(dt.out)
}
 

norm_x_exon <- function(dt){
  dt.out <- dt
  
  meds <- dt %>% ## remove variants in splicing region
      filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>%
    filter(inTarget == TRUE) %>% 
    filter(!uPOS %in% Varaints_to_be_excluded_when_use_MAVE_internal_control_to_build_model$uPOS) %>% ## remove variants on chunling's list
    group_by(Exon, EventType) %>%
    dplyr::summarize(M = median(score1, na.rm=TRUE))
  
  meds2  <- meds %>%
    group_by(EventType) %>%
    dplyr::summarize(M2 = mean(M, na.rm=TRUE))
  
  dt.out$score2 <- NA
  
  for(ei in unique(dt$Exon)){
    dt.out[dt.out$Exon == ei, "score2"] <- myscale(x = dt.out[dt.out$Exon == ei, "score1"],
                                                                   Syn.Avg.med = meds2$M2[meds2$EventType == "Synonymous"],
                                                                   StopGain.Avg.med = meds2$M2[meds2$EventType == "StopGain"],  
                                                                   Syn.med1 =  meds$M[meds$EventType == "Synonymous" & meds$Exon == ei  ],
                                                                   StopGain.med1 = meds$M[meds$EventType == "StopGain" & meds$Exon == ei])
  }
  
  return(dt.out)
}

mydt.snv <- mydt %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") ) %>% filter(inTarget == TRUE)
```

## Results

In total, there are `r length(unique(dt.long$Exon))` Exons in this analysis. And EventType by spice region variants by Exon are 

```{r}

table( mydt.snv$Exon[!is.na(mydt.snv$score0)])
```

```{r}

table( mydt.snv$EventType[!is.na(mydt.snv$score0)], mydt.snv$Exon[!is.na(mydt.snv$score0)], mydt.snv$spliceR[!is.na(mydt.snv$score0)])
```

Since we have two/three replicates in each Exon, the data plotted below were doubled/tripled in number. 


Unique Variants are 

```{r}
dt.av <- mydt %>% dplyr::select(uPOS, Exon, REF, ALT, inTarget, functional.scoreA, functional.scoreB, functional.scoreC, spliceR, EventType) %>% unique()
dt <- dt.av %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") ) %>% filter(inTarget == TRUE)
if(Score_method == "A"){
  print(table(dt$Exon[!is.na(dt$functional.scoreA)]))
  print(table( dt$EventType[!is.na(dt$functional.scoreA)], dt$Exon[!is.na(dt$functional.scoreA)], dt$spliceR[!is.na(dt$functional.scoreA)]))
}else if(Score_method == "B"){
  print(table(dt$Exon[!is.na(dt$functional.scoreB)]))
  print(table( dt$EventType[!is.na(dt$functional.scoreB)], dt$Exon[!is.na(dt$functional.scoreB)], dt$spliceR[!is.na(dt$functional.scoreB)]))
}else{
  print(table(dt$Exon[!is.na(dt$functional.scoreC)]))
  print(table( dt$EventType[!is.na(dt$functional.scoreC)], dt$Exon[!is.na(dt$functional.scoreC)], dt$spliceR[!is.na(dt$functional.scoreC)]))
}
```

## 1. normalized Results for Ratio:


```{r, message=FALSE, warning=FALSE}

dt1 <- norm_win_exon(mydt)
dt2 <- norm_x_exon(dt1)

```


### Step1 normalization.

Plot shows SNV only

```{r, include=TRUE, fig.width=10, fig.height=12}
dt2.av <- dt2 ## dt2.av all variants, dt2 SNV only

dt2 <- dt2.av %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") ) %>% filter(inTarget == TRUE)

etype8 <- unique(dt2$EventType)
if(length(etype8) == 8){
  mycol8 <-  c(1, 5, 6, 2, 3, 4, 7, 8, 9)
}else{
  mycol8 <-  c(1, 5, 3, 4, 2)

}
names(mycol8) <- etype8
mycol2.1 <- mycol8[etype8]


par(mfrow=c(4,3))
for(ei in unique(dt2$Exon)){
plot(dt2$score0[dt2$Exon == ei], dt2$score1[dt2$Exon == ei], pch = 20, col = mycol8[dt2$EventType[dt2$Exon == ei]], main = paste0(ei, " (N =", max(dt2$Experiment[dt2$Exon == ei & !is.na(dt2$score0)]), " , P =", length(unique(dt2$POS[dt2$Exon == ei])), ")"), 
     xlab = "score0", ylab = 'score1')
abline(a = 0, b = 1, col = "gray", lty = 2)
}
plot(0, 0, type = "n", xlab = '', ylab = '', axes = F)
legend(x = "bottom", 
       legend = etype8,
       pch =20,
       col =mycol2.1,
       xpd = TRUE, 
       cex = 1, ncol = 1) # Horizontal legend
```

### Step2 normalization.

Plot shows SNV only


```{r, include=TRUE, fig.width=10, fig.height=12}

par(mfrow=c(4,3))
for(ei in unique(dt$Exon)){
plot(dt2$score1[dt2$Exon == ei], dt2$score2[dt2$Exon == ei], pch = 20, col = mycol8[dt2$EventType[dt2$Exon == ei]], main = paste0(ei, " (N =", max(dt2$Experiment[dt2$Exon == ei & !is.na(dt2$score0)]), " , P =", length(unique(dt2$POS[dt2$Exon == ei])), ")"), 
     xlab = "score1", ylab = 'score2')
abline(a = 0, b = 1, col = "gray", lty = 2)
}

plot(0, 0, type = "n", xlab = '', ylab = '', axes = F)
legend(x = "bottom", 
       legend = etype8,
       pch =20,
       col =mycol2.1,
       xpd = TRUE, 
       cex = 1, ncol = 1) # Horizontal legend
```

### Put together 

Plot shows SNV only


```{r, include=TRUE, fig.width=12, fig.height=12}

par(mfrow=c(3,1))
plot(dt2$POS, dt2$score0, pch = 20, col = mycol8[dt2$EventType], main = "Raw", 
     xlab = "AA position", ylab = 'functional score', ylim = c(-7, 3))
abline(h = 0, col = "gray", lty = 2)
plot(dt2$POS, dt2$score1, pch = 20, col = mycol8[dt2$EventType], main = "Normalization within Exon", 
     xlab = "AA position", ylab = 'functional score', ylim = c(-7, 3))
abline(h = 0, col = "gray", lty = 2)
plot(dt2$POS, dt2$score2, pch = 20, col = mycol8[dt2$EventType], main = "Normalization across Exons", 
     xlab = "AA position", ylab = 'functional score', ylim = c(-7, 3))
abline(h = 0, col = "gray", lty = 2)
legend(x = "bottom", 
       legend = etype8,
       pch =20,
       col =mycol2.1,
       xpd = TRUE, 
       cex = 1, ncol = 3) # Horizontal legend

```


## 2. Score distribution and CI

### Norm check Individual scores ie pooled replicates

Plot shows SNV only

Including splicing region variants

```{r}

dt2plot <- dt2 %>% filter(EventType %in% c("StopGain", "Synonymous")) 
p <- dt2plot %>%
  ggplot( aes(x=score2, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.20) +
  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( paste0(Ratio," individual score after normalization"))
ggplotly(p)


```
Plot shows SNV only

Excluding splicing region variants. Defined as SpliceR == TRUE only.

```{r}
dt2plot <- dt2 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR== FALSE)
p <- dt2plot %>%
  ggplot( aes(x=score2, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.2) +
  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( paste0(Ratio," individual score after normalization excluding splicing region"))
ggplotly(p)


```

Totally, there were `r nrow(dt2plot)` variants left from original `r nrow(dt2)` variants. 

```{r}
shapiro.test(dt2$score2[dt2$EventType == "StopGain"])
shapiro.test(dt2$score2[dt2$EventType == "Synonymous"])

```

### Norm check Average score ie single variant score

Plots show SNV only.

Including splice region variants.

```{r}
temp <- dt2.av %>% select(Exon, uPOS, POS, EventType, REF,     
ALT, AApos,AltAA, AltCodon, EventType, 
RefAA, RefCodon, SpliceAI_ALLELE, SpliceAI_DP_AG, SpliceAI_DP_AL, SpliceAI_DP_DG,  SpliceAI_DP_DL, SpliceAI_DS_AG,  SpliceAI_DS_AL, SpliceAI_DS_DG, SpliceAI_DS_DL, spliceR , Experiment, score2, inTarget) %>% mutate(Experiment = paste0("E", Experiment)) %>% tidyr::spread(key = Experiment, value = score2)


if(any(c("E23_4", "E26_4") %in% exonslist)){
  D14vsLib_fs.2 <- temp %>%rowwise() %>% mutate(functional.score = mean(c(E1, E2, E3,E4), na.rm = T))
}else{
  D14vsLib_fs.2 <- temp %>%rowwise() %>% mutate(functional.score = mean(c(E1, E2, E3), na.rm = T))
}

temp1 <- dt2.av %>% select(uPOS, Experiment, score1) %>% mutate(Experiment = paste0("E", Experiment)) %>% tidyr::spread(key = Experiment, value = score1)
dt2.av.temp <- temp1 %>% rename(R1_wie_norm = E1, R2_wie_norm = E2, R3_wie_norm = E3)
temp1 <- dt2.av %>% select(uPOS, Experiment, score2) %>% mutate(Experiment = paste0("E", Experiment)) %>% tidyr::spread(key = Experiment, value = score2)
dt2.av.temp2 <- temp1 %>% rename(R1_ce_norm = E1, R2_ce_norm = E2, R3_ce_norm = E3)
dt2.av.temp <- merge(merge(dt0, dt2.av.temp), dt2.av.temp2) ## this is to be used to merge with D14vsLib_fs.2 later for output

D14vsLib_fs.2.av <- D14vsLib_fs.2
D14vsLib_fs.2 <- D14vsLib_fs.2.av %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") ) %>% filter(inTarget == TRUE)
dt2plot <- D14vsLib_fs.2 %>% filter(EventType %in% c("StopGain", "Synonymous")) 
p <- dt2plot %>% mutate(name = paste(Exon, uPOS)) %>% 
  ggplot( aes(x=functional.score, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.1) +
  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( paste0("Functional.score ", Score_method, ifelse(Lab_curation == "No", " without ", " with "), "Lab Curation ", "extra filter applied_", extraFilter))
ggplotly(p)


```

Excluding splice region variants (spliceR == TRUE).

```{r}

dt2plot <- D14vsLib_fs.2%>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR== FALSE)
p <- dt2plot %>%
  ggplot( aes(x=functional.score, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.1) +
  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( paste0("Final.functional.score ", Score_method, ifelse(Lab_curation == "No", " w.o.", "with"), " Lab Curation", " extra filter applied_", extraFilter, " Excluding Splice region"))
ggplotly(p)


```

Functional scores By Exon, SNV only

```{r}
dt2plot <- D14vsLib_fs.2%>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR== FALSE)
p <- dt2plot %>%
  ggplot( aes(x=functional.score, y = Exon, fill=EventType, text = uPOS)) +
  geom_point( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.1) +
  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( paste0("Final.functional.score ", Score_method, ifelse(Lab_curation == "No", " w.o.", "with"), " Lab Curation", " extra filter applied_", extraFilter, " Excluding Splice region"))
ggplotly(p)


```

```{r}
temp <- D14vsLib_fs.2
shapiro.test(temp$functional.score[temp$EventType == "StopGain"])
shapiro.test(temp$functional.score[temp$EventType == "Synonymous"])

```


### QQ plots 1. Individual scores

Plots show SNV only.

```{r, fig.width=8, fig.height=4}

par(mfrow=c(1,2))
qqnorm(dt2$score2[dt2$EventType == "StopGain"], main=paste0(Ratio, ' StopGain'))
qqline(dt2$score2[dt2$EventType == "StopGain"], col = "steelblue", lwd = 2)
qqnorm(dt2$score2[dt2$EventType == "Synonymous"], main=paste0(Ratio, ' Synonymous'))
qqline(dt2$score2[dt2$EventType == "Synonymous"], col = "steelblue", lwd = 2)


```

### QQ plots 2. Average score

Plots show SNV only

```{r, fig.width=8, fig.height=4}
temp <- D14vsLib_fs.2
par(mfrow=c(1,2))
qqnorm(temp$functional.score[temp$EventType == "StopGain"], main=paste0(Ratio, ' StopGain'))
qqline(temp$functional.score[temp$EventType == "StopGain"], col = "steelblue", lwd = 2)
qqnorm(temp$functional.score[temp$EventType == "Synonymous"], main=paste0(Ratio, ' Synonymous'))
qqline(temp$functional.score[temp$EventType == "Synonymous"], col = "steelblue", lwd = 2)

```

### Outlier defn 

Individual Score pooled are used. 

Focus on StopGain and Synonymous only.

```{r}

#method1 bootstrap

bootxP<-function(x,y, b,alpha){
# x is your data
# b is the number of bootstraps.
# alpha is your type I error


n <-2000
set.seed(123)
xp <- rep(0, b)
for(i in 1:b){
  x_boot<-c(sample(x,n,replace=TRUE), sample(y,n,replace=TRUE))
  xp[i] <- quantile(x_boot, alpha)
}

return(median(xp))
}

```



Bootstrap Method

1. Estimate of percentile of sample 

We sample large number (N) of samples and each time we get the xth percentile of the bootstrapped sample, estimate of the percentile will be the median of the N number of xth percentile. 


```{r}
x<- dt2$score2[dt2$EventType %in% c("StopGain")]
y<- dt2$score2[dt2$EventType %in% c("Synonymous")]

if(sum(is.na(x))>0) {
  x <- x[-which(is.na(x))]
}
if(sum(is.na(y))>0) {
  y <- y[-which(is.na(y))]
}

bl <- bu <- NA
if(params$Outlierremove != "No"){
  alpha <- ifelse(params$Outlierremove=="p1", 0.01, 
                  ifelse(params$Outlierremove == "p2", 0.02,
                         ifelse(params$Outlierremove == "p5", 0.05, NA)))
  bl <- bootxP(x, y,5000,alpha)
  bu <- bootxP(x, y,5000,1-alpha)
}

```

Plots show SNV only

Including splicing region variants.

```{r}

dt2plot <- dt2 %>% filter(EventType %in% c("Missense", "StopGain", "Synonymous"))
p <- dt2plot %>%
  ggplot( aes(x=score2, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.1) +
  scale_fill_manual(values=c("blue", "red", "#69b3a2")) +  ggtitle ( paste0("score2", Score_method, ifelse(Lab_curation == "No", "without", "with"), "Lab Cureation"))
if(params$Outlierremove != "No"){
  p <- p+ geom_vline(xintercept=c(bl,bu), linetype="dashed", color = "red")  ## upper stopgain
}
ggplotly(p)


```

Excluding splicing region variants (spliceR == TRUE).

```{r}

dt2plot <- dt2 %>% filter(EventType %in% c("Missense", "StopGain", "Synonymous")) %>% filter(spliceR== FALSE)
p <- dt2plot %>%
  ggplot( aes(x=score2, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.1) +
  scale_fill_manual(values=c("blue", "red", "#69b3a2")) +  ggtitle ( paste0("indiviual score after normalization ", Score_method, ifelse(Lab_curation == "No", "without", "with"), "Lab Cureation", "extra filter applied_", extraFilter))
if(params$Outlierremove != "No"){
  p <- p+ geom_vline(xintercept=c(bl,bu), linetype="dashed", color = "red")  ## upper stopgain
}

ggplotly(p)

```



`r ifelse(params$Outlierremove != "No", print(paste0(Ratio, " bootstrapped ", alpha*100, " and ", (1-alpha)*100, " percentile of StopGain & Synonymous are ", bl, " and ", bu)), "")`

```{r removeOutlier}

if(params$Outlierremove != "No"){
  dt.temp <- dt2.av %>% select(uPOS, Experiment, score2) %>% mutate(Experiment = paste0("S", Experiment)) %>% spread(key = Experiment, value = score2)
  dt0 <- merge(dt0, dt.temp)
  temp <- dt2.av %>% mutate(score2 = ifelse(score2 <bl |score2 >bu, NA, score2))
  temp <- temp %>% select(Exon, uPOS, POS, EventType, REF,     
ALT, AApos,AltAA, AltCodon, EventType, RefAA, RefCodon, SpliceAI_ALLELE, SpliceAI_DP_AG, SpliceAI_DP_AL, SpliceAI_DP_DG,  SpliceAI_DP_DL, SpliceAI_DS_AG,  SpliceAI_DS_AL, SpliceAI_DS_DG, SpliceAI_DS_DL, spliceR , Experiment, score2) %>% mutate(Experiment = paste0("E", Experiment)) %>% spread(key = Experiment, value = score2)
if(any(c("E23_4", "E26_4") %in% exonslist)){
  D14vsLib_fs.2o <- temp %>%rowwise() %>% mutate(functional.score = mean(c(E1, E2, E3,E4), na.rm = T))
}else{
  D14vsLib_fs.2o <- temp %>%rowwise() %>% mutate(functional.score = mean(c(E1, E2, E3), na.rm = T))
}
  D14vsLib_fs.2 <- D14vsLib_fs.2o
}

```

## 3. Group distribution

### Average

Plots show SNV only.

```{r}
dt2plot <- D14vsLib_fs.2 %>% filter(EventType %in% c("Missense", "StopGain", "Synonymous")) 

p <- dt2plot %>%
  ggplot( aes(x=functional.score, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.10) +
  scale_fill_manual(values=c( "blue", "red","#69b3a2")) +  ggtitle ( paste0(Ratio," functional score"))
ggplotly(p)

dt2plot <-  D14vsLib_fs.2%>% filter(EventType %in% c("Missense", "StopGain", "Synonymous"))

ggplot(dt2plot, aes(x=EventType, y=functional.score)) + 
  geom_boxplot(outlier.colour="red", outlier.shape=8,
                outlier.size=4) +  ggtitle (Ratio) + geom_jitter(shape=16, position=position_jitter(0.2))

sg <- D14vsLib_fs.2$functional.score[D14vsLib_fs.2$EventType == "StopGain"]
sn <- D14vsLib_fs.2$functional.score[D14vsLib_fs.2$EventType == "Synonymous"]
ms <- D14vsLib_fs.2$functional.score[D14vsLib_fs.2$EventType == "Missense"]

wilcox.test(sg, sn, alternative = "l")
wilcox.test(ms, sn, alternative = "l")
wilcox.test(sg, ms, alternative = "l")

```

## 4. Classification perf 

This section we exclude the splice region variants (spliceR == TRUE) and variants on chunling's list

### 4.1 ROC average 

#### Data Driven

```{r ROC1, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}


temp <- D14vsLib_fs.2 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR== FALSE) %>% arrange(functional.score) %>%  filter(!uPOS %in% Varaints_to_be_excluded_when_use_MAVE_internal_control_to_build_model$uPOS)  ## remove variants on chunling's list
par(mfrow = c(1, 1))
roc_score = roc(controls = temp$functional.score[temp$EventType == "Synonymous"], cases = temp$functional.score[temp$EventType == "StopGain"], direction = ">")
##roc(predictor = temp$functional.score, response = factor(temp$EventType, levels = c( "Synonymous", "StopGain") ))  # AUC score
plot(roc_score, main="ROC curve")
text(0.2, 0.2, paste0("AUC: ", round(roc_score$auc, 3)))

out <- data.frame(functional.score = roc_score$"threshold", sens =  roc_score$"sensitivities", specs = roc_score$"specificities")

out

out$avg <- out$sens+out$specs

auc0 <- as.numeric(roc_score$auc)
cutoff0 <- unique(out$functional.score[which(out$avg == max(out$avg))])
sen0 <- out$sens[which(out$avg == max(out$avg))]
spec0 <- out$specs[which(out$avg == max(out$avg))]

temp$pred.class <- ifelse(temp$functional.score < cutoff0, "StopGain", "Synonymous")

tb0 <- table(truth = temp$EventType, pred = temp$pred.class)
tb0

```

Classification cutoff `r cutoff0`, sensitivity `r sen0`, specificity `r spec0`. 

#### Clinvar data

```{r}
if (grepl("\\.txt$", params$clinvarfile)) {
  # Read the file using read.table
  BRCA2_DBD_missense_classified_in_ClinVar <- read.table(file = params$clinvarfile, header = TRUE, sep = '\t')
} else {
  BRCA2_DBD_missense_classified_in_ClinVar <- read_excel(params$clinvarfile)
}



BRCA2_DBD_missense_classified_in_ClinVar <- BRCA2_DBD_missense_classified_in_ClinVar %>% mutate(uPOS = paste( GRCh38Location-1, Ref, Alt, sep = "_"))

BRCA2_DBD_missense_classified_in_ClinVar$type <- ifelse(BRCA2_DBD_missense_classified_in_ClinVar$classification %in% c("B", "LB", "B (Single)", "B/LB", "LB (Single)"), "Benign", "Pathogenic")

D14vsLib_fs.2.av <- merge(D14vsLib_fs.2.av, BRCA2_DBD_missense_classified_in_ClinVar, by.x = "uPOS", by.y = "uPOS", all.x = T)
D14vsLib_fs.2 <- D14vsLib_fs.2.av %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>% filter(inTarget == TRUE)
```

```{r ROC1.c, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}


temp <- D14vsLib_fs.2 %>% filter(type %in% c("Benign", "Pathogenic")) %>% filter(spliceR== FALSE) %>% arrange(functional.score)%>%  filter(!uPOS %in% Varaints_to_be_excluded_when_use_MAVE_internal_control_to_build_model$uPOS)  ## remove variants on chunling's list
par(mfrow = c(1, 1))
roc_score = roc(controls = temp$functional.score[temp$type == "Benign"], cases = temp$functional.score[temp$type == "Pathogenic"], direction = ">")
plot(roc_score, main="ROC curve")
text(0.2, 0.2, paste0("AUC: ", round(roc_score$auc, 3)))

out <- data.frame(functional.score = roc_score$"threshold", sens =  roc_score$"sensitivities", specs = roc_score$"specificities")

out

out$avg <- out$sens+out$specs

auc0.c <- as.numeric(roc_score$auc)
cutoff0.c <- unique(out$functional.score[which(out$avg == max(out$avg))])
sen0.c <- out$sens[which(out$avg == max(out$avg))]
spec0.c <- out$specs[which(out$avg == max(out$avg))]

temp$pred.class <- ifelse(temp$functional.score < cutoff0.c, "Pathogenic", "Benign")

tb0 <- table(truth = temp$type, pred = temp$pred.class)
tb0

```

Classification cutoff `r cutoff0.c`, sensitivity `r sen0.c`, specificity `r spec0.c`. 

- apply data binary cutoff to clinvar data

```{r ROC1.dc, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}


temp <- D14vsLib_fs.2 %>% filter(type %in% c("Benign", "Pathogenic")) %>% filter(spliceR== FALSE) %>% arrange(functional.score) 

temp$pred.class <- ifelse(temp$functional.score < cutoff0, "Pathogenic", "Benign")

tb0 <- table(truth = temp$type, pred = temp$pred.class)
tb0

```


### 4.2 Mixture Model 

Findlay paper "fitting a two-component Gaussian mixture model in which the parameters of the â€˜non-functionalâ€™ distribution were based on all nonsense SNVs and the â€˜functionalâ€™ distribution based on synonymous SNVs not depleted in RNA (Extended Data Fig. 7). We then used this model to estimate the probability of each SNVâ€™s score being drawn from the non-functional distribution (Pnf). SNVs with Pnf < 0.01 were categorized as functional (72.5%); SNVs with Pnf > 0.99 were categorized as non-functional (21.1%); and SNVs with 0.01 < Pnf < 0.99 (6.4%) were categorized as intermediate."

#### Data driven

Based on distribution from StopGain and synonymous non splicing region (spliceR == FALSE), we also build mixture model.

```{r MixModel, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}
library(mixtools)

temp <- D14vsLib_fs.2 %>% filter(!is.na(functional.score)) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% 
  arrange(functional.score) %>% filter(spliceR== FALSE) %>%  filter(!uPOS %in% Varaints_to_be_excluded_when_use_MAVE_internal_control_to_build_model$uPOS)  ## remove variants on chunling's list

p <- temp %>%
  ggplot( aes(x=functional.score, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.1) +
  scale_fill_manual(values=c("red",  "#69b3a2")) +  ggtitle ( "functional.score for MM building")
ggplotly(p)

m1 <- mean(temp$functional.score[temp$EventType != "StopGain"])
tmp <- temp$functional.score[temp$EventType != "StopGain"]
tmp2 <- tmp ## tmp[which(tmp > median(tmp)-sd(tmp) & tmp < median(tmp)+sd(tmp))]

v1 <- sd(tmp2)
m2 <- mean(temp$functional.score[temp$EventType == "StopGain"])
v2 <- sd(temp$functional.score[temp$EventType == "StopGain"])
l1 <- sum(temp$EventType == "StopGain")/sum(temp$EventType != "StopGain")
if(params$MM == "Robust"){
  # robust calculation
  m2 = mean(temp$functional.score[temp$EventType == "StopGain"], trim = 0.05)
  m1 = mean(temp$functional.score[temp$EventType != "StopGain"], trim = 0.05)
  v2 = IQR(temp$functional.score[temp$EventType == "StopGain"]) / 1.349
  v1 = IQR(temp$functional.score[temp$EventType != "StopGain"]) / 1.349
}
another.error.occured <- FALSE
tryCatch( {
  x <- normalmixEM(temp$functional.score, 
                   lambda = l1, k = 2,
                   mean.constr = c(m2, m1),
                  sd.constr = c(v2, v1),
                 mu = c(m2, m1), 
                 sigma = c(v2, v1))
  
par(mfrow = c(1, 1))
plot(temp$functional.score, x$posterior[,1], col = as.numeric(as.factor(temp$EventType))+1,  ylab = "Prob(nonfunctional)", xlab = "functional score")
f1 <- temp$functional.score[x$posterior[,1]>0.01][order(temp$functional.score[x$posterior[,1]>0.01], decreasing = T)][1]
f2 <- temp$functional.score[x$posterior[,1]>0.99][order(temp$functional.score[x$posterior[,1]>0.99], decreasing = T)][1]

f3 <- temp$functional.score[x$posterior[,1]>0.02][order(temp$functional.score[x$posterior[,1]>0.02], decreasing = T)][1]
f4 <- temp$functional.score[x$posterior[,1]>0.98][order(temp$functional.score[x$posterior[,1]>0.98], decreasing = T)][1]

f5 <- temp$functional.score[x$posterior[,1]>0.05][order(temp$functional.score[x$posterior[,1]>0.05], decreasing = T)][1]
f6 <- temp$functional.score[x$posterior[,1]>0.95][order(temp$functional.score[x$posterior[,1]>0.95], decreasing = T)][1]


abline(v = f1)
abline(v = f2)
abline(h = 0.01)
abline(h = 0.99)

abline(v = f3, lty = 2)
abline(v = f4, lty = 2)
abline(h = 0.02, lty = 2)
abline(h = 0.98, lty = 2)

abline(v = f5, lty = 3)
abline(v = f6, lty = 3)
abline(h = 0.05, lty = 3)
abline(h = 0.95, lty = 3)


par(mfrow = c(1, 1))
plot(x, which=2, xlab2= "", yaxt= "n", xlim = c(range(x$x)[1]-2, range(x$x)[2]+2), main2 ="Two components prob(nonfunctional)", lwd2=0.8, border = "azure3")
lines(density(x$x), lty=2, lwd=0.8)

abline(v = f1)
abline(v = f2)
abline(v = f3, lty = 2)
abline(v = f4, lty = 2)
abline(v = f5, lty = 3)
abline(v = f6, lty = 3)
}
          , error = function(e) {another.error.occured <<- TRUE})
print(another.error.occured)

```


`r paste0("There is",  ifelse(another.error.occured, " AN ", " NO "), "Error Ocurred")`.

`r f2` was the functional score corresponding to the probability of non functional at 99% and `r f1` was the functional score corresponding to the probability of non functional at 1%.

`r f4` was the functional score corresponding to the probability of non functional at 98% and `r f3` was the functional score corresponding to the probability of non functional at 2%.

`r f6` was the functional score corresponding to the probability of non functional at 95% and `r f5` was the functional score corresponding to the probability of non functional at 5%.


If we use these two thresholds to classify the variants of synonymous, stopgain and missense. 

First, 1-99% cutoff
Second, 2-98% cutoff
Third, 5-95% cutoff

Note: Splice regions are included in Missense, not stopgain nor synonymous
variants on chunling's list are included.

```{r MixModel2, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}
temp <- D14vsLib_fs.2 %>% filter(EventType %in% c("Synonymous", "StopGain")) %>% filter(spliceR == FALSE) 

temp2 <- D14vsLib_fs.2 %>% filter(EventType %in% c("Missense")) 

temp <- rbind(temp, temp2)


temp$ctype <- ifelse(temp$functional.score>f1, "neutural", 
                     ifelse(temp$functional.score>f2, "intermediate", "pathogenic"))


temp$ctype2 <- ifelse(temp$functional.score>f3, "neutural", 
                     ifelse(temp$functional.score>f4, "intermediate", "pathogenic"))

temp$ctype3 <- ifelse(temp$functional.score>f5, "neutural", 
                     ifelse(temp$functional.score>f6, "intermediate", "pathogenic"))


temp$ctype <- factor(temp$ctype, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype2 <- factor(temp$ctype2, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype3 <- factor(temp$ctype3, levels = c("pathogenic", "intermediate", "neutural"))

temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))

tb <- table(temp$ctype, temp$EventType)
print(tb)
tb2 <- table(temp$ctype2, temp$EventType)
print(tb2)
tb3 <- table(temp$ctype3, temp$EventType)
print(tb3)
```


- apply cutoffs to clinvar data: 


```{r MixModel2.dc, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}
temp <- D14vsLib_fs.2 %>% filter(type %in% c("Benign", "Pathogenic")) %>% filter(spliceR == FALSE) 

temp$ctype <- ifelse(temp$functional.score>f1, "neutural", 
                     ifelse(temp$functional.score>f2, "intermediate", "pathogenic"))


temp$ctype2 <- ifelse(temp$functional.score>f3, "neutural", 
                     ifelse(temp$functional.score>f4, "intermediate", "pathogenic"))

temp$ctype3 <- ifelse(temp$functional.score>f5, "neutural", 
                     ifelse(temp$functional.score>f6, "intermediate", "pathogenic"))


temp$ctype <- factor(temp$ctype, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype2 <- factor(temp$ctype2, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype3 <- factor(temp$ctype3, levels = c("pathogenic", "intermediate", "neutural"))

tb <- table(temp$ctype, temp$type)
print(tb)
tb2 <- table(temp$ctype2, temp$type)
print(tb2)
tb3 <- table(temp$ctype3, temp$type)
print(tb3)
```

#### ClinVar Data driven

Based on distribution from Clinvar (pathogentic) and (benign), we also build mixture model.

Exclude the spliceR and variants on chunling's list when build model.

```{r}

temp <- D14vsLib_fs.2 %>% filter(!is.na(functional.score)) %>% arrange(functional.score) %>% filter(type %in% c("Pathogenic", "Benign")) %>% filter(spliceR == FALSE) %>%  filter(!uPOS %in% Varaints_to_be_excluded_when_use_MAVE_internal_control_to_build_model$uPOS)  ## remove variants on chunling's list

```

1. Clinvar variants in our Data

```{r}
table(temp$type, temp$EventType, temp$Exon)
```

2. Use Clinvar variants for mixture model

```{r}
table(temp$type)
```

```{r MixModel.C1, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}

p <- temp %>%
  ggplot( aes(x=functional.score, fill=type)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.1) +
  scale_fill_manual(values=c(  "#69b3a2", "red")) +  ggtitle ( "functional.score by clinvar classification")
ggplotly(p)

m1 <- mean(temp$functional.score[temp$type != "Pathogenic"])
tmp <- temp$functional.score[temp$type != "Pathogenic"]
tmp2 <- tmp ## tmp[which(tmp > median(tmp)-sd(tmp) & tmp < median(tmp)+sd(tmp))]

v1 <- sd(tmp2)

m2 <- mean(temp$functional.score[temp$type == "Pathogenic"])
v2 <- sd(temp$functional.score[temp$type == "Pathogenic"])
l1 <- sum(temp$type == "Pathogenic")/sum(temp$type != "Pathogenic")

if(params$MM == "Robust"){
  # robust calculation
  m2 = mean(temp$functional.score[temp$EventType == "StopGain"], trim = 0.05)
  m1 = mean(temp$functional.score[temp$EventType != "StopGain"], trim = 0.05)
  v2 = IQR(temp$functional.score[temp$EventType == "StopGain"]) / 1.349
  v1 = IQR(temp$functional.score[temp$EventType != "StopGain"]) / 1.349
}

an.error.occured <- FALSE
tryCatch( {
x <- normalmixEM(temp$functional.score, 
                   lambda = l1, k = 2,
                 mu = c(m2, m1), 
                 sigma = c(v2, v1))

par(mfrow = c(1, 1))
plot(temp$functional.score, x$posterior[,1], col = 4-as.numeric(as.factor(temp$type)),  ylab = "Prob(nonfunctional)", xlab = "functional score")
f1 <- temp$functional.score[x$posterior[,1]>0.01][order(temp$functional.score[x$posterior[,1]>0.01], decreasing = T)][1]
f2 <- temp$functional.score[x$posterior[,1]>0.99][order(temp$functional.score[x$posterior[,1]>0.99], decreasing = T)][1]

f3 <- temp$functional.score[x$posterior[,1]>0.02][order(temp$functional.score[x$posterior[,1]>0.02], decreasing = T)][1]
f4 <- temp$functional.score[x$posterior[,1]>0.98][order(temp$functional.score[x$posterior[,1]>0.98], decreasing = T)][1]

f5 <- temp$functional.score[x$posterior[,1]>0.05][order(temp$functional.score[x$posterior[,1]>0.05], decreasing = T)][1]
f6 <- temp$functional.score[x$posterior[,1]>0.95][order(temp$functional.score[x$posterior[,1]>0.95], decreasing = T)][1]


abline(v = f1)
abline(v = f2)
abline(h = 0.01)
abline(h = 0.99)

abline(v = f3, lty = 2)
abline(v = f4, lty = 2)
abline(h = 0.02, lty = 2)
abline(h = 0.98, lty = 2)

abline(v = f5, lty = 3)
abline(v = f6, lty = 3)
abline(h = 0.05, lty = 3)
abline(h = 0.95, lty = 3)


par(mfrow = c(1, 1))
plot(x, which=2, xlab2= "", yaxt= "n", xlim = c(range(x$x)[1]-2, range(x$x)[2]+2), main2 ="Two components prob(nonfunctional)", lwd2=0.8, border = "azure3")
lines(density(x$x), lty=2, lwd=0.8)

abline(v = f1)
abline(v = f2)
abline(v = f3, lty = 2)
abline(v = f4, lty = 2)
abline(v = f5, lty = 3)
abline(v = f6, lty = 3)
}, error = function(e) {an.error.occured <<- TRUE})
print(an.error.occured)
```

`r paste0("There is",  ifelse(an.error.occured, " AN ", " NO "), "Error Ocurred")`.


`r f2` was the functional score corresponding to the probability of non functional at 99% and `r f1` was the functional score corresponding to the probability of non functional at 1%.

`r f4` was the functional score corresponding to the probability of non functional at 98% and `r f3` was the functional score corresponding to the probability of non functional at 2%.

`r f6` was the functional score corresponding to the probability of non functional at 95% and `r f5` was the functional score corresponding to the probability of non functional at 5%.


If we use these two thresholds to classify the variants of synonymous, stopgain and missense. 

First, 1-99% cutoff
Second, 2-98% cutoff
Third, 5-95% cutoff

Note: Splice regions are included in Missense, not stopgain nor synonymous

```{r MixModel2.C, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}

if(!an.error.occured){
temp <- D14vsLib_fs.2 %>% filter(EventType %in% c("Synonymous", "StopGain")) %>% filter(spliceR == FALSE) 

temp2 <- D14vsLib_fs.2 %>% filter(EventType %in% c("Missense")) 

temp <- rbind(temp, temp2)


temp$ctype <- ifelse(temp$functional.score>f1, "neutural", 
                     ifelse(temp$functional.score>f2, "intermediate", "pathogenic"))


temp$ctype2 <- ifelse(temp$functional.score>f3, "neutural", 
                     ifelse(temp$functional.score>f4, "intermediate", "pathogenic"))

temp$ctype3 <- ifelse(temp$functional.score>f5, "neutural", 
                     ifelse(temp$functional.score>f6, "intermediate", "pathogenic"))


temp$ctype <- factor(temp$ctype, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype2 <- factor(temp$ctype2, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype3 <- factor(temp$ctype3, levels = c("pathogenic", "intermediate", "neutural"))

temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))

tb <- table(temp$ctype, temp$EventType)
print(tb)

tb2 <- table(temp$ctype2, temp$EventType)
print(tb2)
tb3 <- table(temp$ctype3, temp$EventType)
print(tb3)
}

```

- Apply cutoffs to clinvar data excluding spliceR

```{r MixModel2.Cc, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}

if(!an.error.occured){
temp <- D14vsLib_fs.2 %>% filter(type %in% c("Benign", "Pathogenic")) %>% filter(spliceR == FALSE) 

temp$ctype <- ifelse(temp$functional.score>f1, "neutural", 
                     ifelse(temp$functional.score>f2, "intermediate", "pathogenic"))


temp$ctype2 <- ifelse(temp$functional.score>f3, "neutural", 
                     ifelse(temp$functional.score>f4, "intermediate", "pathogenic"))

temp$ctype3 <- ifelse(temp$functional.score>f5, "neutural", 
                     ifelse(temp$functional.score>f6, "intermediate", "pathogenic"))


temp$ctype <- factor(temp$ctype, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype2 <- factor(temp$ctype2, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype3 <- factor(temp$ctype3, levels = c("pathogenic", "intermediate", "neutural"))


tb <- table(temp$ctype, temp$type)
print(tb)

tb2 <- table(temp$ctype2, temp$type)
print(tb2)
tb3 <- table(temp$ctype3, temp$type)
print(tb3)
}

```

```{r plotout4}
dt2plot <- D14vsLib_fs.2 %>% mutate(spliceremove = ifelse((spliceR == TRUE) & EventType != "Missense", 1, 0)) %>% filter(spliceremove == 0)

p <- dt2plot %>%
  ggplot( aes(x=functional.score, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.1) +
  geom_vline(xintercept=c(f1, f2), linetype="dashed", color = "red") + ## upper stopgain
  geom_vline(xintercept=c(f3, f4), linetype="dotted", color = "green") + ## lower synonymous
  scale_fill_manual(values= mycol8[unique(dt2plot$EventType)]) +
  ggtitle ( "") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  xlab("Functional score") + ylab("Number of variants")+
  theme(legend.title=element_blank()) + scale_y_continuous(expand = c(0, 0))

#ggsave(paste0(resdir, "D14vLib_FS_dist_noSplice.tif"), p, width = 8, height = 5, units="in", device='tiff', dpi=300)

```

```{r plotout5}
dt2plot <- D14vsLib_fs.2 

p <- dt2plot %>%
  ggplot( aes(x=functional.score, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.1) +
  geom_vline(xintercept=c(f1, f2), linetype="dashed", color = "red") + ## upper stopgain
  geom_vline(xintercept=c(f3, f4), linetype="dotted", color = "green") + ## lower synonymous
  scale_fill_manual(values= mycol8[unique(dt2plot$EventType)]) +
  ggtitle ( "") + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  xlab("Functional score") + ylab("Number of variants")+
  theme(legend.title=element_blank()) + scale_y_continuous(expand = c(0, 0))

#ggsave(paste0(resdir, "D14vLib_FS_dist_all.tif"), p, width = 8, height = 5, units="in", device='tiff', dpi=300)

```


### 4.3 Precision

Find cutoff of functional scores with precision of 0.95, 0.98 and 0.99 for both StopGain and Synonymous using our data and Clinvar subsets. 

#### Data driven

Based on distribution from StopGain and synonymous non splicing region (spliceR == FALSE), excluding variants on chunling's list, we also build precision cutoffs.

```{r prec_data, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}
source("###/functions.R")
temp <- D14vsLib_fs.2 %>% filter(!is.na(functional.score)) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% 
  arrange(functional.score) %>% filter(spliceR== FALSE) %>%  filter(!uPOS %in% Varaints_to_be_excluded_when_use_MAVE_internal_control_to_build_model$uPOS)  ## remove variants on chunling's list

score = c(temp$functional.score[temp$EventType == "StopGain"], temp$functional.score[temp$EventType == "Synonymous"])
label = c(temp$EventType[temp$EventType == "StopGain"], temp$EventType[temp$EventType == "Synonymous"])
label = factor(label, levels = c("StopGain", "Synonymous"))
weights = c(1, 1)
precisions = c(0.95, 0.98, 0.99)
cutoffs = precisionBasedThreshold(score, label, weights, precisions)
  
p <- temp %>%
  ggplot( aes(x=functional.score, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.1) +
  geom_vline(xintercept=c(cutoffs[, "StopGain"]), linetype="dashed", color = "red") + ## upper stopgain
  geom_vline(xintercept=c(cutoffs[, "Synonymous"]), linetype="dashed", color = "green") + ## lower synonymous
  scale_fill_manual(values=c("red",  "#69b3a2")) +  ggtitle ( "functional.score with precision cutoffs(weight 1)")
ggplotly(p)

weights = c(1, sum(temp$EventType == "StopGain") / sum(temp$EventType == "Synonymous"))
cutoffs2 = precisionBasedThreshold(score, label, weights, precisions)
  
p <- temp %>%
  ggplot( aes(x=functional.score, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.1) +
  geom_vline(xintercept=c(cutoffs2[, "StopGain"]), linetype="dashed", color = "red") + ## upper stopgain
  geom_vline(xintercept=c(cutoffs2[, "Synonymous"]), linetype="dashed", color = "green") + ## lower synonymous
  scale_fill_manual(values=c("red",  "#69b3a2")) +  ggtitle ( "functional.score with precision cutoffs (data weight)")
ggplotly(p)

weights = c(1, 3*sum(temp$EventType == "StopGain") / sum(temp$EventType == "Synonymous"))
cutoffs3 = precisionBasedThreshold(score, label, weights, precisions)
  
p <- temp %>%
  ggplot( aes(x=functional.score, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.1) +
  geom_vline(xintercept=c(cutoffs3[, "StopGain"]), linetype="dashed", color = "red") + ## upper stopgain
  geom_vline(xintercept=c(cutoffs3[, "Synonymous"]), linetype="dashed", color = "green") + ## lower synonymous
  scale_fill_manual(values=c("red",  "#69b3a2")) +  ggtitle ( "functional.score with precision cutoffs (alpha weight)")
ggplotly(p)
```


for precision cutoffs with weight ratio of 1, 1.
```{r}
print(cutoffs)
```
for precision cutoffs with weight ratio of data (1, `r sum(temp$EventType == "StopGain") / sum(temp$EventType == "Synonymous")`).
```{r}
print(cutoffs2)
```
for precision cutoffs with weight ratio of data (1, `r 3*sum(temp$EventType == "StopGain") / sum(temp$EventType == "Synonymous")`).
```{r}
print(cutoffs3)
```

Applying the cutoffs on three types together (MAVE data).

- cutoffs with weight ratio of 1. 
Note: Splice regions are included in Missense, not stopgain nor synonymous



```{r prec_data2, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}
temp <- D14vsLib_fs.2 %>% filter(EventType %in% c("Synonymous", "StopGain")) %>% filter(spliceR == FALSE) 

temp2 <- D14vsLib_fs.2 %>% filter(EventType %in% c("Missense")) 

temp <- rbind(temp, temp2)


temp$ctype.p95 <- ifelse(temp$functional.score>cutoffs[1, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs[1, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p98 <- ifelse(temp$functional.score>cutoffs[2, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs[2, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p99 <- ifelse(temp$functional.score>cutoffs[3, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs[3, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p95 <- factor(temp$ctype.p95, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p98 <- factor(temp$ctype.p98, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p99 <- factor(temp$ctype.p99, levels = c("pathogenic", "intermediate", "neutural"))

temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))

tb <- table(pred = temp$ctype.p95, truth = temp$EventType)
print(tb)

tb2 <- table(pred = temp$ctype.p98, truth = temp$EventType)
print(tb2)
tb3 <- table(pred = temp$ctype.p99, truth = temp$EventType)
print(tb3)
```

- cutoffs with weight ratio of data ratio. 
Note: Splice regions are included in Missense, not stopgain nor synonymous

```{r prec_data2.1, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}

temp$ctype.p95.1 <- ifelse(temp$functional.score>cutoffs2[1, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs2[1, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p98.1 <- ifelse(temp$functional.score>cutoffs2[2, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs2[2, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p99.1 <- ifelse(temp$functional.score>cutoffs2[3, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs2[3, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p95.1 <- factor(temp$ctype.p95.1, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p98.1 <- factor(temp$ctype.p98.1, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p99.1 <- factor(temp$ctype.p99.1, levels = c("pathogenic", "intermediate", "neutural"))

temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))

tb <- table(pred = temp$ctype.p95.1, truth = temp$EventType)
print(tb)

tb2 <- table(pred = temp$ctype.p98.1, truth = temp$EventType)
print(tb2)
tb3 <- table(pred = temp$ctype.p99.1, truth = temp$EventType)
print(tb3)
```

- cutoffs with weight ratio of alpha ratio. 
Note: Splice regions are included in Missense, not stopgain nor synonymous

```{r prec_data2.1a, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}

temp$ctype.p95.1a <- ifelse(temp$functional.score>cutoffs3[1, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs3[1, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p98.1a <- ifelse(temp$functional.score>cutoffs3[2, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs3[2, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p99.1a <- ifelse(temp$functional.score>cutoffs3[3, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs3[3, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p95.1a <- factor(temp$ctype.p95.1a, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p98.1a <- factor(temp$ctype.p98.1a, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p99.1a <- factor(temp$ctype.p99.1a, levels = c("pathogenic", "intermediate", "neutural"))

temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))

tb <- table(pred = temp$ctype.p95.1a, truth = temp$EventType)
print(tb)

tb2 <- table(pred = temp$ctype.p98.1a, truth = temp$EventType)
print(tb2)
tb3 <- table(pred = temp$ctype.p99.1a, truth = temp$EventType)
print(tb3)
```
Applying the cutoffs on Clinvar data excluding spliceR.

- cutoffs with weight ratio of 1. 


```{r prec_data2c, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}
temp <- D14vsLib_fs.2 %>% filter(type %in% c("Benign", "Pathogenic")) %>% filter(spliceR == FALSE) 

temp$ctype.p95 <- ifelse(temp$functional.score>cutoffs[1, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs[1, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p98 <- ifelse(temp$functional.score>cutoffs[2, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs[2, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p99 <- ifelse(temp$functional.score>cutoffs[3, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs[3, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p95 <- factor(temp$ctype.p95, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p98 <- factor(temp$ctype.p98, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p99 <- factor(temp$ctype.p99, levels = c("pathogenic", "intermediate", "neutural"))


tb <- table(pred = temp$ctype.p95, truth = temp$type)
print(tb)

tb2 <- table(pred = temp$ctype.p98, truth = temp$type)
print(tb2)
tb3 <- table(pred = temp$ctype.p99, truth = temp$type)
print(tb3)
```

- cutoffs with weight ratio of data ratio. 

```{r prec_data2.1cc, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}

temp$ctype.p95.1 <- ifelse(temp$functional.score>cutoffs2[1, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs2[1, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p98.1 <- ifelse(temp$functional.score>cutoffs2[2, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs2[2, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p99.1 <- ifelse(temp$functional.score>cutoffs2[3, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs2[3, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p95.1 <- factor(temp$ctype.p95.1, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p98.1 <- factor(temp$ctype.p98.1, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p99.1 <- factor(temp$ctype.p99.1, levels = c("pathogenic", "intermediate", "neutural"))


tb <- table(pred = temp$ctype.p95.1, truth = temp$type)
print(tb)

tb2 <- table(pred = temp$ctype.p98.1, truth = temp$type)
print(tb2)
tb3 <- table(pred = temp$ctype.p99.1, truth = temp$type)
print(tb3)
```

- cutoffs with weight ratio of alpha ratio. 

```{r prec_data2.1acc, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}

temp$ctype.p95.1a <- ifelse(temp$functional.score>cutoffs3[1, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs3[1, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p98.1a <- ifelse(temp$functional.score>cutoffs3[2, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs3[2, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p99.1a <- ifelse(temp$functional.score>cutoffs3[3, "Synonymous"], "neutural", 
                     ifelse(temp$functional.score>cutoffs3[3, "StopGain"], "intermediate", "pathogenic"))
temp$ctype.p95.1a <- factor(temp$ctype.p95.1a, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p98.1a <- factor(temp$ctype.p98.1a, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p99.1a <- factor(temp$ctype.p99.1a, levels = c("pathogenic", "intermediate", "neutural"))


tb <- table(pred = temp$ctype.p95.1a, truth = temp$type)
print(tb)

tb2 <- table(pred = temp$ctype.p98.1a, truth = temp$type)
print(tb2)
tb3 <- table(pred = temp$ctype.p99.1a, truth = temp$type)
print(tb3)
```

#### ClinVar Data driven

Based on distribution from Clinvar (pathogentic) and (benign), we also build mixture model.
Exlcude spliceR and variants on chunling's list.

```{r prec.clinvar, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}
temp <- D14vsLib_fs.2 %>% filter(!is.na(functional.score)) %>% arrange(functional.score) %>% filter(type %in% c("Pathogenic", "Benign")) %>% filter(spliceR == FALSE) %>% filter(!uPOS %in% Varaints_to_be_excluded_when_use_MAVE_internal_control_to_build_model$uPOS)  ## remove variants on chunling's list


score = c(temp$functional.score[temp$type == "Pathogenic"], temp$functional.score[temp$type == "Benign"])
label = c(temp$type[temp$type == "Pathogenic"], temp$type[temp$type == "Benign"])
label = factor(label, levels = c("Pathogenic", "Benign"))
weights = c(1, 1)
precisions = c(0.95, 0.98, 0.99)
cutoffs.c = precisionBasedThreshold(score, label, weights, precisions)
  
p <- temp %>%
  ggplot( aes(x=functional.score, fill=type)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.1) +
  geom_vline(xintercept=c(cutoffs.c[, "Pathogenic"]), linetype="dashed", color = "red") + ## upper stopgain
  geom_vline(xintercept=c(cutoffs.c[, "Benign"]), linetype="dashed", color = "green") + ## lower synonymous
  scale_fill_manual(values=c("#69b3a2", "red")) +  ggtitle ( "functional.score with precision cutoffs(weight 1)")
ggplotly(p)

weights = c(1, sum(temp$type == "Pathogenic") / sum(temp$type == "Benign"))
cutoffs2.c = precisionBasedThreshold(score, label, weights, precisions)
  
p <- temp %>%
  ggplot( aes(x=functional.score, fill=type)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.1) +
  geom_vline(xintercept=c(cutoffs2.c[, "Pathogenic"]), linetype="dashed", color = "red") + ## upper stopgain
  geom_vline(xintercept=c(cutoffs2.c[, "Benign"]), linetype="dashed", color = "green") + ## lower synonymous
  scale_fill_manual(values=c("#69b3a2", "red")) +  ggtitle ( "functional.score with precision cutoffs (data weight)")
ggplotly(p)

weights = c(1, 3*sum(temp$type == "Pathogenic") / sum(temp$type == "Benign"))
cutoffs3.c = precisionBasedThreshold(score, label, weights, precisions)
  
p <- temp %>%
  ggplot( aes(x=functional.score, fill=type)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity', binwidth = 0.1) +
  geom_vline(xintercept=c(cutoffs3.c[, "Pathogenic"]), linetype="dashed", color = "red") + ## upper stopgain
  geom_vline(xintercept=c(cutoffs3.c[, "Benign"]), linetype="dashed", color = "green") + ## lower synonymous
  scale_fill_manual(values=c("#69b3a2", "red")) +  ggtitle ( "functional.score with precision cutoffs (alpha weight)")
ggplotly(p)
```


for precision cutoffs with weight ratio of 1, 1.
```{r}
print(cutoffs.c)
```

for precision cutoffs with weight ratio of data (1, `r sum(temp$type == "Pathogenic") / sum(temp$type == "Benign")`).
```{r}
print(cutoffs2.c)
```

for precision cutoffs with weight ratio of data (1, `r 3*sum(temp$type == "Pathogenic") / sum(temp$type == "Benign")`).
```{r}
print(cutoffs3.c)
```


Applying the cutoffs on three types together for MAVE data.

- cutoffs with weight ratio of 1. 
Note: Splice regions are included in Missense, not stopgain nor synonymous

```{r prec_data2.c, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}
temp <- D14vsLib_fs.2 %>% filter(EventType %in% c("Synonymous", "StopGain")) %>% filter(spliceR == FALSE) 

temp2 <- D14vsLib_fs.2 %>% filter(EventType %in% c("Missense")) 

temp <- rbind(temp, temp2)


temp$ctype.p95 <- ifelse(temp$functional.score>cutoffs.c[1, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs.c[1, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p98 <- ifelse(temp$functional.score>cutoffs.c[2, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs.c[2, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p99 <- ifelse(temp$functional.score>cutoffs.c[3, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs.c[3, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p95 <- factor(temp$ctype.p95, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p98 <- factor(temp$ctype.p98, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p99 <- factor(temp$ctype.p99, levels = c("pathogenic", "intermediate", "neutural"))

temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))

tb <- table(pred = temp$ctype.p95, truth = temp$EventType)
print(tb)

tb2 <- table(pred = temp$ctype.p98, truth = temp$EventType)
print(tb2)
tb3 <- table(pred = temp$ctype.p99, truth = temp$EventType)
print(tb3)
```

- cutoffs with weight ratio of data ratio. 
Note: Splice regions are included in Missense, not stopgain nor synonymous

```{r prec_data2.1c, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}

temp$ctype.p95.1 <- ifelse(temp$functional.score>cutoffs2.c[1, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs2.c[1, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p98.1 <- ifelse(temp$functional.score>cutoffs2.c[2, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs2.c[2, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p99.1 <- ifelse(temp$functional.score>cutoffs2.c[3, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs2.c[3, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p95.1 <- factor(temp$ctype.p95.1, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p98.1 <- factor(temp$ctype.p98.1, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p99.1 <- factor(temp$ctype.p99.1, levels = c("pathogenic", "intermediate", "neutural"))

temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))

tb <- table(pred = temp$ctype.p95.1, truth = temp$EventType)
print(tb)

tb2 <- table(pred = temp$ctype.p98.1, truth = temp$EventType)
print(tb2)
tb3 <- table(pred = temp$ctype.p99.1, truth = temp$EventType)
print(tb3)
```

- cutoffs with weight ratio of alpha ratio. 
Note: Splice regions are included in Missense, not stopgain nor synonymous

```{r prec_data2.1ca, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}

temp$ctype.p95.1a <- ifelse(temp$functional.score>cutoffs3.c[1, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs3.c[1, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p98.1a <- ifelse(temp$functional.score>cutoffs3.c[2, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs3.c[2, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p99.1a <- ifelse(temp$functional.score>cutoffs3.c[3, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs3.c[3, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p95.1a <- factor(temp$ctype.p95.1a, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p98.1a <- factor(temp$ctype.p98.1a, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p99.1a <- factor(temp$ctype.p99.1a, levels = c("pathogenic", "intermediate", "neutural"))

temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))

tb <- table(pred = temp$ctype.p95.1a, truth = temp$EventType)
print(tb)

tb2 <- table(pred = temp$ctype.p98.1a, truth = temp$EventType)
print(tb2)
tb3 <- table(pred = temp$ctype.p99.1a, truth = temp$EventType)
print(tb3)
```



Applying the cutoffs on Clinvar data

- cutoffs with weight ratio of 1. 

```{r prec_data2.ccc, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}
temp <- D14vsLib_fs.2 %>% filter(type %in% c("Benign", "Pathogenic")) %>% filter(spliceR == FALSE) 

temp$ctype.p95 <- ifelse(temp$functional.score>cutoffs.c[1, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs.c[1, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p98 <- ifelse(temp$functional.score>cutoffs.c[2, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs.c[2, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p99 <- ifelse(temp$functional.score>cutoffs.c[3, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs.c[3, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p95 <- factor(temp$ctype.p95, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p98 <- factor(temp$ctype.p98, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p99 <- factor(temp$ctype.p99, levels = c("pathogenic", "intermediate", "neutural"))


tb <- table(pred = temp$ctype.p95, truth = temp$type)
print(tb)

tb2 <- table(pred = temp$ctype.p98, truth = temp$type)
print(tb2)
tb3 <- table(pred = temp$ctype.p99, truth = temp$type)
print(tb3)
```

- cutoffs with weight ratio of data ratio. 

```{r prec_data2.1ccc, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}

temp$ctype.p95.1 <- ifelse(temp$functional.score>cutoffs2.c[1, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs2.c[1, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p98.1 <- ifelse(temp$functional.score>cutoffs2.c[2, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs2.c[2, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p99.1 <- ifelse(temp$functional.score>cutoffs2.c[3, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs2.c[3, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p95.1 <- factor(temp$ctype.p95.1, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p98.1 <- factor(temp$ctype.p98.1, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p99.1 <- factor(temp$ctype.p99.1, levels = c("pathogenic", "intermediate", "neutural"))


tb <- table(pred = temp$ctype.p95.1, truth = temp$type)
print(tb)

tb2 <- table(pred = temp$ctype.p98.1, truth = temp$type)
print(tb2)
tb3 <- table(pred = temp$ctype.p99.1, truth = temp$type)
print(tb3)
```

- cutoffs with weight ratio of alpha ratio. 

```{r prec_data2.1cacc, include=TRUE, fig.width=9, fig.height=5, message=FALSE, warning=FALSE}

temp$ctype.p95.1a <- ifelse(temp$functional.score>cutoffs3.c[1, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs3.c[1, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p98.1a <- ifelse(temp$functional.score>cutoffs3.c[2, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs3.c[2, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p99.1a <- ifelse(temp$functional.score>cutoffs3.c[3, "Benign"], "neutural", 
                     ifelse(temp$functional.score>cutoffs3.c[3, "Pathogenic"], "intermediate", "pathogenic"))
temp$ctype.p95.1a <- factor(temp$ctype.p95.1a, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p98.1a <- factor(temp$ctype.p98.1a, levels = c("pathogenic", "intermediate", "neutural"))
temp$ctype.p99.1a <- factor(temp$ctype.p99.1a, levels = c("pathogenic", "intermediate", "neutural"))


tb <- table(pred = temp$ctype.p95.1a, truth = temp$type)
print(tb)

tb2 <- table(pred = temp$ctype.p98.1a, truth = temp$type)
print(tb2)
tb3 <- table(pred = temp$ctype.p99.1a, truth = temp$type)
print(tb3)
```


```{r}
## output merge with raw D14_v_lib
out <- merge(D14vsLib_fs.2.av, dt2.av.temp) %>% dplyr::select(-E1, -E2, -E3) %>% relocate(functional.score, .after = R3_ce_norm)
## add the cutoffs
out$p95.sg.w1 <- cutoffs[1, "StopGain"]; out$p95.sn.w1 <- cutoffs[1, "Synonymous"]; 
out$p98.sg.w1 <- cutoffs[2, "StopGain"]; out$p98.sn.w1 <- cutoffs[2, "Synonymous"]; 
out$p99.sg.w1 <- cutoffs[3, "StopGain"]; out$p99.sn.w1 <- cutoffs[3, "Synonymous"]; 
out$p95.sg.wd <- cutoffs2[1, "StopGain"]; out$p95.sn.wd <- cutoffs2[1, "Synonymous"]; 
out$p98.sg.wd <- cutoffs2[2, "StopGain"]; out$p98.sn.wd <- cutoffs2[2, "Synonymous"]; 
out$p99.sg.wd <- cutoffs2[3, "StopGain"]; out$p99.sn.wd <- cutoffs2[3, "Synonymous"]; 

out$p95.sg.w1.c <- cutoffs.c[1, "Pathogenic"]; out$p95.sn.w1.c <- cutoffs.c[1, "Benign"]; 
out$p98.sg.w1.c <- cutoffs.c[2, "Pathogenic"]; out$p98.sn.w1.c <- cutoffs.c[2, "Benign"]; 
out$p99.sg.w1.c <- cutoffs.c[3, "Pathogenic"]; out$p99.sn.w1.c <- cutoffs.c[3, "Benign"]; 
out$p95.sg.wd.c <- cutoffs2.c[1, "Pathogenic"]; out$p95.sn.wd.c <- cutoffs2.c[1, "Benign"]; 
out$p98.sg.wd.c <- cutoffs2.c[2, "Pathogenic"]; out$p98.sn.wd.c <- cutoffs2.c[2, "Benign"]; 
out$p99.sg.wd.c <- cutoffs2.c[3, "Pathogenic"]; out$p99.sn.wd.c <- cutoffs2.c[3, "Benign"]; 

write.csv(out, paste0(resdir, Ratio, "_fs_", Score_method, "_", (paste0(ifelse(Lab_curation == "No", "without", "with"), "_LabCureation_")), "E", gsub(", ", "", Exons), "_", params$loessSubset, "_", params$EventCount_Filter, "_MM_", params$MM, "_Outlierremove_",  params$Outlierremove,  "_extrafilter_", extraFilter,".csv"), row.names = FALSE)

out2 <- out %>% filter(uPOS %in% D14vsLib_fs.2$uPOS)
write.csv(out2, paste0(resdir, "SNV_", Ratio, "_fs_", Score_method, "_", (paste0(ifelse(Lab_curation == "No", "without", "with"), "_LabCureation_")), "E", gsub(", ", "", Exons), "_", params$loessSubset, "_", params$EventCount_Filter, "_MM_", params$MM, "_Outlierremove_",  params$Outlierremove,  "_extrafilter_", extraFilter,".csv"), row.names = FALSE)

``` 


```{r}
## output the metrics in one file
metrics <- data.frame(Exons = params$Exons,
                      EvenFilter = params$EventCount_Filter,
                      vset = params$vset,
                      Method = params$loessSubset,
                      score = Score_method,
                      Ratio = params$Ratio,
                      MM = params$MM,
                      Outlierremove = params$Outlierremove,
                      AUC_D14_lib=as.numeric(auc0),
                      sens_D14_lib=sen0,
                      spec_D14_lib=spec0,
                      confusion.matrix_D14_lib_roc = c(paste(as.vector((c(tb0[1,], tb0[2,]))), collapse = " ")),
                      miscall_D14_lib_roc=c(sum(tb0) - sum(diag(tb0))),  
                      confusion.matrix_D14_lib_data_1 = c(paste(as.vector((c(tb[1,], tb[2,], tb[3,]))), collapse = " ")),
                      miscall_D14_lib_data_1=c(sum(tb) - sum(diag(tb))),                      
                      confusion.matrix_D14_lib_data_2 = c(paste(as.vector((c(tb2[1,], tb2[2,], tb2[3,]))), collapse = " ")),
                      
                      miscall_D14_lib_data_2=c(sum(tb2) - sum(diag(tb2))),

                      confusion.matrix_D14_lib_data_5 = c(paste(as.vector((c(tb3[1,], tb3[2,], tb3[3,]))), collapse = " ")),
                      
                      miscall_D14_lib_data_5=c(sum(tb3) - sum(diag(tb3))))

write.csv(metrics, paste0(params$Output_dir, "BRCA2_", Ratio, "_", params$vset,"_", params$loessSubset, "_", params$EventCount_Filter, "_", params$Score_method, "_", (paste0(ifelse(Lab_curation == "No", "without", "with"), "_LabCureation_")), "E", gsub(", ", "", Exons), "_", params$loessSubset, "_", params$EventCount_Filter, "_MM_", params$MM, "_Outlierremove_",  params$Outlierremove, "_extrafilter_", extraFilter, "_metrics.csv"), row.names = FALSE)

```
