---
author: 
  "Jeanie Na"
date: "`r format(Sys.time(), '%d %B %Y')`"
title    : "High throughput study analysis one experiment investigation individual exon report"


output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    theme: united
    highlight: tango

params:

  Exon:
    label: "Exon:"
    value: "18N_NS3"
    input: text
    
  Output_dir:
    label: "Output Dir:"
    value: "/research/bsi/projects/breast/s108235.tripneg_Couch/projects/s108235.tripneg/breast_requests/Predisposition_genes/Chunling/BRCA2/results/11_2023/Ind_1.5/"
    input: text       
  
  loessSubset:
    label: "Loess Filter (keep):"
    value: "ModFindlay2"
    choices: ["Findlay", "ModFindlay", "ModFindlay2", "NegFindlay"]
    input: select
    
  EventCount_Filter:
    label: "EventCount Filter:"
    value: "libD510"
    choices: ["lib10", "libD510"]
    input: select
  vset:
    label: "Variant set:"
    value: "SNV"
    choices: ["allv", "SNV"]
    input: select
  Gene: 
    label: "Gene:"
    value: "BRCA2"
    choices: ["BRCA2", "PALB2", "RAD51C"]
    input: select
  C_fc:
    value: 1.5
---

```{=html}
<style type="text/css">
.main-container {
  max-width: 1200px;
  margin-left: auto;
  margin-right: auto;
}
</style>
```
<link rel="stylesheet" type="text/css" href="http://cdn.datatables.net/1.10.5/css/jquery.dataTables.min.css">

```{=html}
<script type="text/javascript">
  $(document).ready(function() {
    $("table").DataTable();
  } );
</script>
```

```{r setup, include=TRUE, warning=FALSE, message=FALSE}

knitr::opts_chunk$set(echo = TRUE, warning=FALSE)
library(kableExtra)
library(plotly)
library(arsenal)
require(patchwork)

library(ggplot2)
library(dplyr)
library(tidyr)
library(readr)
library(readxl)
library(tidyverse)
library(threejs)
library (ROCR);
library(pROC)

rdir <- '//research/bsi/projects/breast/s108235.tripneg_Couch/projects/s108235.tripneg/breast_requests/Predisposition_genes/Chunling/BRCA2/rpgm/'

exonDataList <- readr::read_table("/research/bsi/projects/breast/s108235.tripneg_Couch/projects/s108235.tripneg/breast_requests/Predisposition_genes/Chunling/BRCA2/data/exonDataList.txt")
Exn <- paste0("E", params$Exon)
loessSubset <- params$loessSubset
EventCount_Filter <- params$EventCount_Filter
vset <- params$vset
Gene <- params$Gene 
evc <- ifelse(length(grep("100", EventCount_Filter)), 100, 10)
C_fc <- params$C_fc
filenames <- unlist(strsplit(exonDataList$filenames[which(exonDataList$Gene == params$Gene & exonDataList$Exon == Exn)],","))
excludeAApos <- exonDataList$AApos[which(exonDataList$Gene == params$Gene & exonDataList$Exon == Exn)]
dtdir <- exonDataList$path[which(exonDataList$Gene == Gene& exonDataList$Exon == Exn)]
```

## Setting of this report

Gene: `r Gene`\ 
Exon: `r Exn`\
dtdir: `r  dtdir`\
filenames: `r filenames`\
loessSubset:  `r params$loessSubset`\
EventCount_Filter: `r params$EventCount_Filter`\
vset: `r vset`\
Score A and C _foldchange : `r params$C_fc`\

## Aim

To propose functional scores to classify variants for `r params$Gene` `r Exn` only. We are looking into `r length(filenames)` experiment(s) in this report.

```{r readin, include=TRUE, warning=FALSE, message=FALSE}
# read in data
tests <- list()
for (i in 1:length(filenames)){
  
  test1 <-read_delim(paste0(dtdir, filenames[i]), 
                                      delim = "\t", escape_double = FALSE, 
                                      trim_ws = TRUE)

  test1 <- test1 %>% mutate(SpliceAI_DS_AG = round(SpliceAI_DS_AG, 5),
                           SpliceAI_DS_AL = round(SpliceAI_DS_AL, 5),
                           SpliceAI_DS_DG = round(SpliceAI_DS_DG, 5),
                           SpliceAI_DS_DL = round(SpliceAI_DS_DL, 5))
  
  test1$sample_id <- paste0("R", i, "_", test1$sample_id)
  print(unique(test1$sample_id))
  sns <- unique(test1$sample_id)

  if(length(grep("ssm", tolower(sns)))>0){
    test1.lib <- sns[grep("ssm", tolower(sns))]
  }else if (length(grep("lib", tolower(sns)))>0){
    test1.lib <- sns[grep("lib", tolower(sns))]
  }
  
  test1.D5 <- sns[grep("d5", tolower(sns))]
  
  if(length(grep("d14", tolower(sns)))>0){
    test1.D14 <- sns[grep("d14", tolower(sns))]
  }else if (length(grep("d20", tolower(sns)))>0){
    test1.D14 <- sns[grep("d20", tolower(sns))]
  }
  
  
  test1 <- test1 %>% filter(sample_id %in% c(test1.lib, test1.D5, test1.D14))
  
  test1$sample_id[test1$sample_id == test1.lib] <- paste0("R", i, "_lib")
  test1$sample_id[test1$sample_id == test1.D5] <- paste0("R", i, "_D5")
  test1$sample_id[test1$sample_id == test1.D14] <- paste0("R", i, "_D14")
  
  tests[[i]] <- test1
}


```

## Data


`r filenames` was used in this report.

Time points:

Lib, D5,  `r paste0(ifelse(grep('D14', test1.D14), "D14", "D20"), "used in this analysis)")`.

## Target region

We will subset data within the target region as shown below.

```{r}

target <- read.csv("/research/bsi/projects/breast/s108235.tripneg_Couch/projects/s108235.tripneg/breast_requests/Predisposition_genes/Chunling/BRCA2/data/10_2023_curation//chunling_target.csv")

kable(target[, c("Exon", "Start", "End")])
```

## Method

Please refer to the email Chunling sent on 7/20/2022.

A. Within each replicates:

1.  Before Loess:

-- Exclude rows with "NA" in "EventType" column; 

-- Exclude rows with "EventCount" \< `r evc` and " sample_id" = lib column or D5 

-- Calculate the normalized freq. (evenCount/sum(eventCount)) for lib, d5, `r ifelse(grep('D14', test1.D14), "D14", "D20")`


2.  Apply Loess position adjustment: 

-- Subset variants according to the filtering `r params$loesssubset`

-- Use the above subset variants to fit loess line then apply to all dataset for loess transformation (use span of 0.15 or to be determined), extend flatly outward to include any positions not fit. In our data, we are not excluding any event type: "UpstreamNoncoding", "PartialCodingAndUpstreamNoncoding" , "PartialCodingUpStream", "StopGain", "Missense", "Synonymous", "DownstreamNonCodingAndPartialCodingDownstream", "DownstreamNonCoding". Note: Findlay may have used different approach.

3.  After Loess transformation

-- Exclude rows with " SpliceAI" (any one of the SpliceAI 4 "DS" columns) \>0.2; 

-- restrict to SNVs (variants with single nt change column ALT and REF: "REF"="A" or "T" or "C" or "G" and ALT ="A" or "T" or "C" or "G")

B. Normalization across replicates 

-- Scores from within each replicate were linearly scaled such that the median synonymous and median nonsense SNVs within the replicate were set to the median synonymous and median nonsense SNV values averaged across replicate experiments.

C. Functional scores

-- Single Experiment report
-- Multiple Experiments

D. Classification

-- Mixture model.

-- ROC balance of accuracy, ie, max(sens + spec)

E. Output

-- In addition to this report, there will be a csv file output with all the intermediate results. 
For example, the file name would be like EX19_2E_4lab_cutoff_0_sen0.84_spec0.78_wFS.csv

It is the Exon name + how many Experiments + loess subset + 4lab_cutoff_ + cutoff  + Sensitivity + specificity from ROC.


Refer to Findlay's paper for methods.

## Results

### A. Within each replicate D5_lib Loess

```{r}
## exclude with "NA" in "EventType" column;
## exclude with "EventCount" <10 and " sample_id" =" E3_pUCSSMmu_lib" or D5
if(EventCount_Filter %in% c("lib10", "lib100")){
  for( i in 1:length(filenames)){
    tests[[i]] <- tests[[i]] %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_")) %>% filter(!is.na(EventType)) %>% filter(!(EventCount < evc & sample_id %in% c(paste0("R", i, "_lib"))))
  }
  
} else {
    for( i in 1:length(filenames)){
    tests[[i]] <- tests[[i]] %>% mutate(uPOS = paste(POS, REF, ALT, sep = "_")) %>% filter(!is.na(EventType)) %>% filter(!(EventCount < evc & sample_id %in% c(paste0("R", i, "_lib"), paste0("R", i, "_D5"))))
    }
}
  
```

After filtering NA for EventType, Event Count < `r evc` for `r EventCount_Filter`, replicate1 has `r length(unique(tests[[1]]$uPOS))` variants. 
`r ifelse(length(filenames) > 1, paste0("replicate2 has ", length(unique(tests[[2]]$uPOS)), " variants"), "")` 
`r ifelse(length(filenames) > 2, paste0("replicate3 has ", length(unique(tests[[3]]$uPOS)), " variants"), "")`
`r ifelse(length(filenames) > 3, paste0("replicate4 has ", length(unique(tests[[4]]$uPOS)), " variants"), "")`

```{r}
test <- bind_rows(tests)
wide <- test %>% arrange(POS) %>% select(-AAEventCount, -AANormalizedEventCount, -NormalizedEventCount)  %>% 
        spread(sample_id, EventCount)
      
colnames(wide) <- make.names(colnames(wide))
# # one experiments
wide <- wide %>% mutate(R1_lib_freq = log2(R1_lib/sum(R1_lib, na.rm = T)),
                              R1_D5_freq = log2(R1_D5/sum(R1_D5, na.rm = T)),
                              R1_D14_freq = log2(R1_D14/sum(R1_D14, na.rm = T)),
                              
                              R1_D5_lib = R1_D5_freq - R1_lib_freq,
                              R1_D14_lib= R1_D14_freq - R1_lib_freq,
                              R1_D14_D5 = R1_D14_freq - R1_D5_freq
                              )%>% arrange(POS)
    
if(length(filenames) >1){ # two experiments
    ### calculate ratio
    wide <- wide %>% mutate(
                              R2_lib_freq = log2(R2_lib/sum(R2_lib, na.rm = T)),
                              R2_D5_freq = log2(R2_D5/sum(R2_D5, na.rm = T)),
                              R2_D14_freq = log2(R2_D14/sum(R2_D14, na.rm = T)),
                              
                              R2_D5_lib = R2_D5_freq - R2_lib_freq,
                              R2_D14_lib= R2_D14_freq - R2_lib_freq,
                              R2_D14_D5 = R2_D14_freq - R2_D5_freq
                              )%>% arrange(POS)
  } 
  
  if(length(filenames) >2) {# three experiments
    ### calculate ratio
    wide <- wide %>% mutate(R3_lib_freq = log2(R3_lib/sum(R3_lib, na.rm = T)),
                              R3_D5_freq = log2(R3_D5/sum(R3_D5, na.rm = T)),
                              R3_D14_freq = log2(R3_D14/sum(R3_D14, na.rm = T)),
                              
                              R3_D5_lib = R3_D5_freq - R3_lib_freq,
                              R3_D14_lib= R3_D14_freq - R3_lib_freq,
                              R3_D14_D5 = R3_D14_freq - R3_D5_freq
                            )%>% arrange(POS)
  }

  if(length(filenames) >3) { # four experiments
    ### calculate ratio
    wide <- wide %>% mutate(R4_lib_freq = log2(R4_lib/sum(R4_lib, na.rm = T)),
                              R4_D5_freq = log2(R4_D5/sum(R4_D5, na.rm = T)),
                              R4_D14_freq = log2(R4_D14/sum(R4_D14, na.rm = T)),
                              
                              R4_D5_lib = R4_D5_freq - R4_lib_freq,
                              R4_D14_lib= R4_D14_freq - R4_lib_freq,
                              R4_D14_D5 = R4_D14_freq - R4_D5_freq
                            )%>% arrange(POS)
  }
```



```{r}
## update EvenType using Chunling's latest file
ET <- read.csv("/research/bsi/projects/breast/s108235.tripneg_Couch/projects/s108235.tripneg/breast_requests/Predisposition_genes/Chunling/BRCA2/data/10_2023_curation/chunling_ET.csv") %>% rename(EventType = ET) %>% select(-uid)

#find the row indices
row_idx <- match(ET$uPOS, wide$uPOS)

#replace the values
wide[row_idx[-which(is.na(row_idx))], names(ET)] <- ET[-which(is.na(row_idx)),]

unique(wide$EventType)
wide <- wide %>% arrange(POS) %>% filter(!AApos %in% excludeAApos) ## remove pam site
```

After merging all `r length(filenames)` files, we have total `r nrow(wide)` variants.


Before loess, plot the D5_lib and D14_D5 ratio, dashed lines are log2(0.5) and log2(0.8) that were used in Findlay.

```{r lowess, include=TRUE, fig.width=10, fig.height= 4*length(filenames) }
par(mfrow = c(length(filenames), 1))
sspan <- 0.15

etype <- c("UpstreamNoncoding", "PartialCodingAndUpstreamNoncoding",  "PartialCodingUpStream", "StopGain", "Missense", "Synonymous", "DownstreamNonCodingAndPartialCodingDownstream", "DownstreamNonCoding", "PartialCodingDownStream", "Intron", "Splice")
mycol <- c(1, 5, 6, 2, 3, 4, 7, 8, 9, 15, 16)
names(mycol) <- etype

if(loessSubset == "Findlay"){
  test1 <- wide %>% filter(R1_D5_lib > log2(0.5) & R1_D14_D5 > log2(0.8))
} else if(loessSubset == "ModFindlay"){
  test1 <- wide %>% filter(!(R1_D5_lib < log2(0.5) & R1_D14_D5 < log2(0.8)))
} else if(loessSubset == "NegFindlay"){
  test1 <- wide %>% filter(!(R1_D5_lib > log2(0.5) & R1_D14_D5 > log2(0.8)))
}else{ ## ModFindlay2
  test1 <- wide %>% filter(R1_D14_D5 > log2(0.8))
}

if(vset == "SNV"){
  test1 <- test1 %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )
}

loess15.1 <- loess(R1_D5_lib ~ POS, data=test1, span=sspan)

# get smoothed output
smoothed15 <- predict(loess15.1, data.frame(POS = wide$POS), se = TRUE)
frontendNA <- rearendNA <- NULL
frontendNA <- which(is.na(smoothed15$fit) & as.numeric(names(smoothed15$fit)) < 1/2*nrow(wide))
if(length(frontendNA)){
  smoothed15$fit[frontendNA] <- smoothed15$fit[max(frontendNA)+1]
}
rearendNA <- which(is.na(smoothed15$fit) & as.numeric(names(smoothed15$fit)) > 1/2*nrow(wide))
if(length(rearendNA)){
  smoothed15$fit[rearendNA] <- smoothed15$fit[min(rearendNA)-1]
}

temp <- mycol[match(wide$EventType, etype)]
plot(wide$R1_D5_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R1"), xlab="Position", ylab="D5_lib_ratio (log2)", col = temp)
abline(h = log2(0.5), lty = 2)

wide$D5_lib_loess_R1 <- smoothed15$fit


plot(wide$R1_D14_D5, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R1"), xlab="Position", ylab="D14_D5_ratio (log2)", col = temp)
abline(h = log2(0.8), lty = 2)

wide$D14_lib_ratio_adjusted_R1 <- wide$R1_D14_lib-wide$D5_lib_loess_R1

if(length(filenames) >1 ){ ## two experiments
  if(loessSubset == "Findlay"){
    test2 <- wide %>% filter(R2_D5_lib > log2(0.5) & R2_D14_D5 > log2(0.8))
  } else if(loessSubset == "ModFindlay"){
    test2 <- wide %>% filter(!(R2_D5_lib < log2(0.5) & R2_D14_D5 < log2(0.8)))
  } else if(loessSubset == "NegFindlay"){
    test2 <- wide %>% filter(!(R2_D5_lib > log2(0.5) & R2_D14_D5 > log2(0.8)))
  }else{ ## ModFindlay2
    test2 <- wide %>% filter(R2_D14_D5 > log2(0.8))
  }
  if(vset == "SNV"){
    test2 <- test2 %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )
  }
  
  loess15.2 <- loess(R2_D5_lib ~ POS, data=test2, span=sspan)
  
  # get smoothed output
  smoothed15 <- predict(loess15.2, data.frame(POS = wide$POS), se = TRUE)
  frontendNA <- rearendNA <- NULL
  frontendNA <- which(is.na(smoothed15$fit) & as.numeric(names(smoothed15$fit)) < 1/2*nrow(wide))
  if(length(frontendNA)){
    smoothed15$fit[frontendNA] <- smoothed15$fit[max(frontendNA)+1]
  }
  rearendNA <- which(is.na(smoothed15$fit) & as.numeric(names(smoothed15$fit)) > 1/2*nrow(wide))
  if(length(rearendNA)){
    smoothed15$fit[rearendNA] <- smoothed15$fit[min(rearendNA)-1]
  }
  
  temp <- mycol[match(wide$EventType, etype)]
  plot(wide$R2_D5_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R2"), xlab="Position", ylab="D5_lib_ratio (log2)", col = temp)
  abline(h = log2(0.5), lty = 2)
  
  wide$D5_lib_loess_R2 <- smoothed15$fit
  
  
  plot(wide$R2_D14_D5, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R2"), xlab="Position", ylab="D14_D5_ratio (log2)", col = temp)
  abline(h = log2(0.8), lty = 2)
  
  wide$D14_lib_ratio_adjusted_R2 <- wide$R2_D14_lib-wide$D5_lib_loess_R2
  
}

if(length(filenames) >2 ){ ## three experiments
  if(loessSubset == "Findlay"){
    test3 <- wide %>% filter(R3_D5_lib > log2(0.5) & R3_D14_D5 > log2(0.8))
  } else if(loessSubset == "ModFindlay"){
    test3 <- wide %>% filter(!(R3_D5_lib < log2(0.5) & R3_D14_D5 < log2(0.8)))
  } else if(loessSubset == "NegFindlay"){
    test3 <- wide %>% filter(!(R3_D5_lib > log2(0.5) & R3_D14_D5 > log2(0.8)))
  }else{ ## ModFindlay2
    test3 <- wide %>% filter(R3_D14_D5 > log2(0.8))
  }
  
  if(vset == "SNV"){
    test3 <- test3 %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )
  }
  
  loess15.3 <- loess(R3_D5_lib ~ POS, data=test3, span=sspan)
  
  # get smoothed output
  smoothed15 <- predict(loess15.3, data.frame(POS = wide$POS), se = TRUE)
  frontendNA <- rearendNA <- NULL
  frontendNA <- which(is.na(smoothed15$fit) & as.numeric(names(smoothed15$fit)) < 1/2*nrow(wide))
  if(length(frontendNA)){
    smoothed15$fit[frontendNA] <- smoothed15$fit[max(frontendNA)+1]
  }
  rearendNA <- which(is.na(smoothed15$fit) & as.numeric(names(smoothed15$fit)) > 1/2*nrow(wide))
  if(length(rearendNA)){
    smoothed15$fit[rearendNA] <- smoothed15$fit[min(rearendNA)-1]
  }
  
  temp <- mycol[match(wide$EventType, etype)]
  plot(wide$R3_D5_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R3"), xlab="Position", ylab="D5_lib_ratio (log2)", col = temp)
  abline(h = log2(0.5), lty = 2)
  
  wide$D5_lib_loess_R3 <- smoothed15$fit
  
  
  plot(wide$R3_D14_D5, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R3"), xlab="Position", ylab="D14_D5_ratio (log2)", col = temp)
  abline(h = log2(0.8), lty = 2)
  
  wide$D14_lib_ratio_adjusted_R3 <- wide$R3_D14_lib-wide$D5_lib_loess_R3
  
}

if(length(filenames) >3 ){ ## four experiments
  if(loessSubset == "Findlay"){
    test4 <- wide %>% filter(R4_D5_lib > log2(0.5) & R4_D14_D5 > log2(0.8))
  } else if(loessSubset == "ModFindlay"){
    test4 <- wide %>% filter(!(R4_D5_lib < log2(0.5) & R4_D14_D5 < log2(0.8)))
  } else if(loessSubset == "NegFindlay"){
    test4 <- wide %>% filter(!(R4_D5_lib > log2(0.5) & R4_D14_D5 > log2(0.8)))
  }else{ ## ModFindlay2
    test4 <- wide %>% filter(R4_D14_D5 > log2(0.8))
  }
  
  if(vset == "SNV"){
    test4 <- test4 %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )
  }
  
  loess15.4 <- loess(R4_D5_lib ~ POS, data=test4, span=sspan)
  
  # get smoothed output
  smoothed15 <- predict(loess15.4, data.frame(POS = wide$POS), se = TRUE)
  frontendNA <- rearendNA <- NULL
  frontendNA <- which(is.na(smoothed15$fit) & as.numeric(names(smoothed15$fit)) < 1/2*nrow(wide))
  if(length(frontendNA)){
    smoothed15$fit[frontendNA] <- smoothed15$fit[max(frontendNA)+1]
  }
  rearendNA <- which(is.na(smoothed15$fit) & as.numeric(names(smoothed15$fit)) > 1/2*nrow(wide))
  if(length(rearendNA)){
    smoothed15$fit[rearendNA] <- smoothed15$fit[min(rearendNA)-1]
  }
  
  temp <- mycol[match(wide$EventType, etype)]
  plot(wide$R4_D5_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R4"), xlab="Position", ylab="D5_lib_ratio (log2)", col = temp)
  abline(h = log2(0.5), lty = 2)
  
  wide$D5_lib_loess_R4 <- smoothed15$fit
  
  
  plot(wide$R4_D14_D5, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R4"), xlab="Position", ylab="D14_D5_ratio (log2)", col = temp)
  abline(h = log2(0.8), lty = 2)
  
  wide$D14_lib_ratio_adjusted_R4 <- wide$R4_D14_lib-wide$D5_lib_loess_R4
  
}
```

After subsetting , replicate1 has `r nrow(test1[!is.na(test1$R1_D5_lib), ])` variants used for LOESS fitting then applied back to all the position in the previous data to get fitted D5 values for each variant. 

```{r lowess2.1, include=TRUE, fig.width=10, fig.height=9 }
par(mfrow = c(3, 1))
temp <- mycol[match(test1$EventType, etype)]
plot(test1$R1_D5_lib, x=test1$POS, type="p", pch = 20, main=paste0(Exn, " R1 variants used for loess"), xlab="Position", ylab="D5_lib_ratio (log2)", col = temp)
smoothed15.0 <- predict(loess15.1, data.frame(POS = test1$POS), se = TRUE)

lines(smoothed15.0$fit, x=test1$POS, col="red")
abline(h = log2(0.5), lty = 2)
plot(test1$R1_D14_D5, x=test1$POS, type="p", pch = 20, main=paste0(Exn, " R1 variants used for loess"), xlab="Position", ylab="D14_D5_ratio (log2)", col = temp)
abline(h = log2(0.8), lty = 2)
temp <- mycol[match(wide$EventType, etype)]

plot(wide$R1_D5_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R1 all variants"), xlab="Position", ylab="D5_lib_ratio (log2)", col = temp)
lines(wide$D5_lib_loess_R1, x=wide$POS, col="red")

```

`r ifelse(length(filenames) == 2, paste0("After subsetting , replicate2 has ", nrow(test2[!is.na(test2$R2_D5_lib), ]), " variants used for LOESS fitting then applied back to all the position in the previous data to get fitted D5 values for each variant."), "") `


```{r lowess2.2, include=TRUE, fig.width=10, fig.height=9 }
if(length(filenames) >1 ) {
  par(mfrow = c(3, 1))
  temp <- mycol[match(test2$EventType, etype)]
  plot(test2$R2_D5_lib, x=test2$POS, type="p", pch = 20, main=paste0(Exn, " R2 variants used for loess"), xlab="Position", ylab="D5_lib_ratio (log2)", col = temp)
  smoothed15.0 <- predict(loess15.2, data.frame(POS = test2$POS), se = TRUE)
  
  lines(smoothed15.0$fit, x=test2$POS, col="red")
  abline(h = log2(0.5), lty = 2)
  plot(test2$R2_D14_D5, x=test2$POS, type="p", pch = 20, main=paste0(Exn, " R2 variants used for loess"), xlab="Position", ylab="D14_D5_ratio (log2)", col = temp)
  abline(h = log2(0.8), lty = 2)
  temp <- mycol[match(wide$EventType, etype)]
  
  plot(wide$R2_D5_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R2 all variants"), xlab="Position", ylab="D5_lib_ratio (log2)", col = temp)
  lines(wide$D5_lib_loess_R2, x=wide$POS, col="red")
}
```

`r ifelse(length(filenames) == 3, paste0("After subsetting , replicate3 has ", nrow(test3[!is.na(test3$R3_D5_lib), ]), " variants used for LOESS fitting then applied back to all the position in the previous data to get fitted D5 values for each variant."), "") `


```{r lowess2.3, include=TRUE, fig.width=10, fig.height=9 }
if(length(filenames) >2) {
  par(mfrow = c(3, 1))
  temp <- mycol[match(test3$EventType, etype)]
  plot(test3$R3_D5_lib, x=test3$POS, type="p", pch = 20, main=paste0(Exn, " R3 variants used for loess"), xlab="Position", ylab="D5_lib_ratio (log2)", col = temp)
  smoothed15.0 <- predict(loess15.3, data.frame(POS = test3$POS), se = TRUE)
  
  lines(smoothed15.0$fit, x=test3$POS, col="red")
  abline(h = log2(0.5), lty = 2)
  plot(test3$R3_D14_D5, x=test3$POS, type="p", pch = 20, main=paste0(Exn, " R3 variants used for loess"), xlab="Position", ylab="D14_D5_ratio (log2)", col = temp)
  abline(h = log2(0.8), lty = 2)
  temp <- mycol[match(wide$EventType, etype)]
  
  plot(wide$R3_D5_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R3 all variants"), xlab="Position", ylab="D5_lib_ratio (log2)", col = temp)
  lines(wide$D5_lib_loess_R3, x=wide$POS, col="red")
}

```
`r ifelse(length(filenames) == 4, paste0("After subsetting , replicate4 has ", nrow(test4[!is.na(test4$R4_D5_lib), ]), " variants used for LOESS fitting then applied back to all the position in the previous data to get fitted D5 values for each variant."), "") `


```{r lowess2.4, include=TRUE, fig.width=10, fig.height=9 }
if(length(filenames) >3) {
  par(mfrow = c(3, 1))
  temp <- mycol[match(test4$EventType, etype)]
  plot(test4$R4_D5_lib, x=test4$POS, type="p", pch = 20, main=paste0(Exn, " R4 variants used for loess"), xlab="Position", ylab="D5_lib_ratio (log2)", col = temp)
  smoothed15.0 <- predict(loess15.4, data.frame(POS = test4$POS), se = TRUE)
  
  lines(smoothed15.0$fit, x=test4$POS, col="red")
  abline(h = log2(0.5), lty = 2)
  plot(test4$R4_D14_D5, x=test4$POS, type="p", pch = 20, main=paste0(Exn, " R4 variants used for loess"), xlab="Position", ylab="D14_D5_ratio (log2)", col = temp)
  abline(h = log2(0.8), lty = 2)
  temp <- mycol[match(wide$EventType, etype)]
  
  plot(wide$R4_D5_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R4 all variants"), xlab="Position", ylab="D5_lib_ratio (log2)", col = temp)
  lines(wide$D5_lib_loess_R4, x=wide$POS, col="red")
}

```

#### Position adjustment for D14_lib ratio

Check D5 fitting.

```{r adjustment.1, include=TRUE, fig.width=10, fig.height=8}

mycol2 <- mycol[etype]



### 
## plot the before and after D5
par(mfrow = c(2, 1))
plot(wide$R1_D5_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R1"), xlab="Position", ylab="D5_lib_ratio (log2) before", col = mycol2[match(wide$EventType, etype)])
lines(wide$D5_lib_loess_R1, x=wide$POS, col="red")

plot(wide$R1_D5_lib-wide$D5_lib_loess_R1, x=wide$POS, type="p",  pch = 20, main=paste0(Exn, " R1"), xlab="Position", ylab="D5_lib_ratio (log2) after", col =mycol2[match(wide$EventType, etype)])
lgcol <- unique(mycol2[match(wide$EventType, etype)])
lgnm <- names(lgcol)
legend(x = "bottom",
       inset = -0.15, 
       legend = etype,
       pch =20,
       col =mycol2,
       xpd = TRUE, 
       cex = 1, ncol = 3) # Horizontal legend


```

```{r adjustment.2, include=TRUE, fig.width=10, fig.height=8}

mycol2 <- mycol[etype]


if(length(filenames) >1) {
  ### 
  ## plot the before and after D5
  par(mfrow = c(2, 1))
  plot(wide$R2_D5_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R2"), xlab="Position", ylab="D5_lib_ratio (log2) before", col = mycol2[match(wide$EventType, etype)])
  lines(wide$D5_lib_loess_R2, x=wide$POS, col="red")
  
  plot(wide$R2_D5_lib-wide$D5_lib_loess_R2, x=wide$POS, type="p",  pch = 20, main=paste0(Exn, " R2"), xlab="Position", ylab="D5_lib_ratio (log2) after", col =mycol2[match(wide$EventType, etype)])
  legend(x = "bottom",
         inset = -0.15, 
         legend = etype,
         pch =20,
         col =mycol2,
         xpd = TRUE, 
         cex = 1, ncol = 3) # Horizontal legend
}

```

```{r adjustment.3, include=TRUE, fig.width=10, fig.height=8}

mycol2 <- mycol[etype]

if(length(filenames) >2) {

  ### 
  ## plot the before and after D5
  par(mfrow = c(2, 1))
  plot(wide$R3_D5_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R3"), xlab="Position", ylab="D5_lib_ratio (log2) before", col = mycol2[match(wide$EventType, etype)])
  lines(wide$D5_lib_loess_R3, x=wide$POS, col="red")
  
  plot(wide$R3_D5_lib-wide$D5_lib_loess_R3, x=wide$POS, type="p",  pch = 20, main=paste0(Exn, " R3"), xlab="Position", ylab="D5_lib_ratio (log2) after", col =mycol2[match(wide$EventType, etype)])
  legend(x = "bottom",
         inset = -0.15, 
         legend = etype,
         pch =20,
         col =mycol2,
         xpd = TRUE, 
         cex = 1, ncol = 3) # Horizontal legend
}

```


```{r adjustment.4, include=TRUE, fig.width=10, fig.height=8}

mycol2 <- mycol[etype]

if(length(filenames) >3) {

  ### 
  ## plot the before and after D5
  par(mfrow = c(2, 1))
  plot(wide$R4_D5_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn, " R4"), xlab="Position", ylab="D5_lib_ratio (log2) before", col = mycol2[match(wide$EventType, etype)])
  lines(wide$D5_lib_loess_R4, x=wide$POS, col="red")
  
  plot(wide$R4_D5_lib-wide$D5_lib_loess_R4, x=wide$POS, type="p",  pch = 20, main=paste0(Exn, " R4"), xlab="Position", ylab="D5_lib_ratio (log2) after", col =mycol2[match(wide$EventType, etype)])
  legend(x = "bottom",
         inset = -0.15, 
         legend = etype,
         pch =20,
         col =mycol2,
         xpd = TRUE, 
         cex = 1, ncol = 3) # Horizontal legend
}

```


D14_lib ratio position adjustment

```{r adjustment2.1, include=TRUE, fig.width=10, fig.height=8}

### 
## plot the before and after D14
par(mfrow = c(2, 1))
plot(wide$R1_D14_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn," R1"), xlab="Position", ylab="D14_lib_ratio (log2) before", col = mycol2[match(wide$EventType, etype)])
plot(wide$D14_lib_ratio_adjusted_R1, x=wide$POS, type="p",  pch = 20, main=paste0(Exn, " R1"), xlab="Position", ylab="D14_lib_ratio (log2) after", col =mycol2[match(wide$EventType, etype)])
legend(x = "bottom",
       inset = -0.15, 
       legend = etype,
       pch =20,
       col =mycol2,
       xpd = TRUE, 
       cex = 1, ncol = 3) # Horizontal legend
```

```{r adjustment2.2, include=TRUE, fig.width=10, fig.height=8}
if(length(filenames) > 1) {
  ### 
  ## plot the before and after D14
  par(mfrow = c(2, 1))
  plot(wide$R2_D14_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn," R2"), xlab="Position", ylab="D14_lib_ratio (log2) before", col = mycol2[match(wide$EventType, etype)])
  plot(wide$D14_lib_ratio_adjusted_R2, x=wide$POS, type="p",  pch = 20, main=paste0(Exn, " R2"), xlab="Position", ylab="D14_lib_ratio (log2) after", col =mycol2[match(wide$EventType, etype)])
  legend(x = "bottom",
         inset = -0.15, 
         legend = etype,
         pch =20,
         col =mycol2,
         xpd = TRUE, 
         cex = 1, ncol = 3) # Horizontal legend
}
```

```{r adjustment2.3, include=TRUE, fig.width=10, fig.height=8}
if(length(filenames) >2) {
  ### 
  ## plot the before and after D14
  par(mfrow = c(2, 1))
  plot(wide$R3_D14_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn," R3"), xlab="Position", ylab="D14_lib_ratio (log2) before", col = mycol2[match(wide$EventType, etype)])
  plot(wide$D14_lib_ratio_adjusted_R3, x=wide$POS, type="p",  pch = 20, main=paste0(Exn, " R3"), xlab="Position", ylab="D14_lib_ratio (log2) after", col =mycol2[match(wide$EventType, etype)])
  legend(x = "bottom",
         inset = -0.15, 
         legend = etype,
         pch =20,
         col =mycol2,
         xpd = TRUE, 
         cex = 1, ncol = 3) # Horizontal legend
}
```


```{r adjustment2.4, include=TRUE, fig.width=10, fig.height=8}
if(length(filenames) >3) {
  ### 
  ## plot the before and after D14
  par(mfrow = c(2, 1))
  plot(wide$R4_D14_lib, x=wide$POS, type="p", pch = 20, main=paste0(Exn," R4"), xlab="Position", ylab="D14_lib_ratio (log2) before", col = mycol2[match(wide$EventType, etype)])
  plot(wide$D14_lib_ratio_adjusted_R4, x=wide$POS, type="p",  pch = 20, main=paste0(Exn, " R4"), xlab="Position", ylab="D14_lib_ratio (log2) after", col =mycol2[match(wide$EventType, etype)])
  legend(x = "bottom",
         inset = -0.15, 
         legend = etype,
         pch =20,
         col =mycol2,
         xpd = TRUE, 
         cex = 1, ncol = 3) # Horizontal legend
}
```


```{r}

# After Loess and position normalization:
# --exclude with " SpliceAI" (any one of the SpliceAI 4 columns) >0.2;
# --restrict to variants with single nt change (filter by column "REF"="A" or "T" or "C" or "G")
##  mutate(splice = rowSums(select(., 8:11) > 0.2)>0) %>% filter(splice == FALSE) %>% arrange(position)
wide <- wide %>% mutate(spliceR = rowSums(select(., 16:19) > 0.2, na.rm = TRUE)>0)
if(Exn == "E18N_NS3"){
  wide <- wide %>% filter(POS <= target[target$Exon == "E18N", "End"] & POS >= target[target$Exon == "E18N", "Start"])

}else{
  wide <- wide %>% filter(POS <= target[target$Exon == Exn, "End"] & POS >= target[target$Exon == Exn, "Start"])

}


wide2 <- wide %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") ) ## SNV no splice, this is to use as model building or normalization, also subset to target region only

sliceRs <- wide %>% filter(spliceR == TRUE)

```


```{r}
library(readxl)
Varaints_to_be_excluded_when_use_MAVE_internal_control_to_build_model <- read_excel("/research/bsi/projects/breast/s108235.tripneg_Couch/projects/s108235.tripneg/breast_requests/Predisposition_genes/Chunling/BRCA2/data/10_2023_curation/Varaints to be excluded when use MAVE internal control to build model.xlsx")

Varaints_to_be_excluded_when_use_MAVE_internal_control_to_build_model <- Varaints_to_be_excluded_when_use_MAVE_internal_control_to_build_model %>% mutate(uPOS = paste(GRCh38Location-1, REF, ALT, sep = "_") )
                                                                                                                    
wide2 <- wide2 %>% filter(!wide2$uPOS %in% Varaints_to_be_excluded_when_use_MAVE_internal_control_to_build_model$uPOS)    
                                                                                                                    
## SNV no splice, this is to use as model building or normalization

```


### B. Within Exon Normalization 

From Findlay paper, "Scores from within each replicate were linearly scaled such that the median synonymous and median nonsense SNVs within the replicate were set to the median synonymous and median nonsense SNV values averaged across replicate experiments.

We follow the same approach.

Within each replicate, scores were linearly scaled such that the median synonymous and median stopgain SNVs within the replicate were set to the median synonymous and median stopgain SNV values averaged across replicate experiments.

Note: Here we exclude with " SpliceAI" (any one of the SpliceAI 4 columns) >0.2; SNV only

After removing Splice variants, variants size decreased from `r length(unique(wide$uPOS))` to `r length(unique(wide2$uPOS))`. We use this subset to get the median of StopGain and Synonymous within replicate and across replicates. Then normalize on all the variants.

NOTE::: single experiment, there will be no change after normalization due to the nature of the scaling.


#### D14vlib

```{r D14vLib_normalize, include=TRUE, fig.width= length(filenames)*4, fig.height=5}
Syn.med1<- Syn.med2<- Syn.med3 <- Syn.med4 <- StopGain.med1 <- StopGain.med2 <- StopGain.med3 <- StopGain.med4 <-NA

Syn.med1 <- median(wide2$D14_lib_ratio_adjusted_R1[which(wide2$EventType == "Synonymous")], na.rm = T); 
if(length(filenames) >1) {
  Syn.med2 <- median(wide2$D14_lib_ratio_adjusted_R2[which(wide2$EventType == "Synonymous")], na.rm = T); 
  if(length(filenames) > 2){
    Syn.med3 <- median(wide2$D14_lib_ratio_adjusted_R3[which(wide2$EventType == "Synonymous")], na.rm = T); 
      if(length(filenames) > 3){
        Syn.med4 <- median(wide2$D14_lib_ratio_adjusted_R4[which(wide2$EventType == "Synonymous")], na.rm = T); 
      }
  }
}

StopGain.med1 <- median(wide2$D14_lib_ratio_adjusted_R1[which(wide2$EventType == "StopGain")], na.rm = T); 
if(length(filenames) >1) {
  StopGain.med2 <- median(wide2$D14_lib_ratio_adjusted_R2[which(wide2$EventType == "StopGain")], na.rm = T); 
  if(length(filenames) > 2){
    StopGain.med3 <- median(wide2$D14_lib_ratio_adjusted_R3[which(wide2$EventType == "StopGain")], na.rm = T); 
      if(length(filenames) > 2){
         StopGain.med4 <- median(wide2$D14_lib_ratio_adjusted_R4[which(wide2$EventType == "StopGain")], na.rm = T); 
      }
  }
}

Syn.Avg.med <- mean(c(Syn.med1, Syn.med2, Syn.med3, Syn.med4), na.rm = T) 
StopGain.Avg.med <- mean(c(StopGain.med1, StopGain.med2, StopGain.med3, StopGain.med4), na.rm = T)

scale1 <- function(x){
  a <- (Syn.Avg.med-StopGain.Avg.med)/(Syn.med1-StopGain.med1)
  b  <- Syn.Avg.med - a*Syn.med1
  x2 <- a*x +b
  return(x2)
}

scale2 <- function(x){
  a <- (Syn.Avg.med-StopGain.Avg.med)/(Syn.med2-StopGain.med2)
  b  <- Syn.Avg.med - a*Syn.med2
  x2 <- a*x +b
  return(x2)
}

scale3 <- function(x){
  a <- (Syn.Avg.med-StopGain.Avg.med)/(Syn.med3-StopGain.med3)
  b  <- Syn.Avg.med - a*Syn.med3
  x2 <- a*x +b
  return(x2)
}

scale4 <- function(x){
  a <- (Syn.Avg.med-StopGain.Avg.med)/(Syn.med3-StopGain.med4)
  b  <- Syn.Avg.med - a*Syn.med4
  x2 <- a*x +b
  return(x2)
}

wide$D14_lib_ratio_adjusted_R1.1 <- scale1(wide$D14_lib_ratio_adjusted_R1)
if(length(filenames) >1) {
  wide$D14_lib_ratio_adjusted_R2.1 <- scale2(wide$D14_lib_ratio_adjusted_R2) 
  if(length(filenames) > 2){
    wide$D14_lib_ratio_adjusted_R3.1 <- scale3(wide$D14_lib_ratio_adjusted_R3)
    if(length(filenames) > 3){
        wide$D14_lib_ratio_adjusted_R4.1 <- scale3(wide$D14_lib_ratio_adjusted_R4)
    }
  }
}

etype <- unique(wide$EventType)
mycol2 <- mycol[etype]

par(mfrow=c(1, length(filenames)))

plot(wide$D14_lib_ratio_adjusted_R1, wide$D14_lib_ratio_adjusted_R1.1, pch = 20, col = mycol2[match(wide$EventType, etype)], main = "R1 Normalization")
abline(v = c(Syn.med1,StopGain.med1), col = c(3, 2))
abline(h = c(Syn.Avg.med,StopGain.Avg.med), col = c(3, 2))
abline(a=0, b=1)
if(length(filenames) >1) {
  plot(wide$D14_lib_ratio_adjusted_R2, wide$D14_lib_ratio_adjusted_R2.1, pch = 20, col = mycol2[match(wide$EventType, etype)], main = "R2 Normalization")
  abline(v = c(Syn.med2,StopGain.med2), col = c(3, 2))
  abline(h = c(Syn.Avg.med,StopGain.Avg.med), col = c(3, 2))
  abline(a=0, b=1)
 
  if(length(filenames) > 2){
    plot(wide$D14_lib_ratio_adjusted_R3, wide$D14_lib_ratio_adjusted_R3.1, pch = 20, col = mycol2[match(wide$EventType, etype)], main = "R3 Normalization")
    abline(v = c(Syn.med3,StopGain.med3), col = c(3, 2))
    abline(h = c(Syn.Avg.med,StopGain.Avg.med), col = c(3, 2))
    abline(a=0, b=1)
      
    if(length(filenames) > 3){
    plot(wide$D14_lib_ratio_adjusted_R4, wide$D14_lib_ratio_adjusted_R4.1, pch = 20, col = mycol2[match(wide$EventType, etype)], main = "R4 Normalization")
    abline(v = c(Syn.med4,StopGain.med4), col = c(3, 2))
    abline(h = c(Syn.Avg.med,StopGain.Avg.med), col = c(3, 2))
    abline(a=0, b=1)
  }
  }
}


legend(par('usr')[2], par('usr')[4], ##"bottomright",
       legend = etype,
       pch =20,
       col =mycol2,
       cex = 1, ncol = 1) # Horizontal legend
```

Check each experiment

```{r D14vLib_normalize2, include=TRUE, fig.width=10, fig.height=16}
for (i in 1:length(filenames)){
  par(mfrow=c(4,1))

  plot(wide$POS, unlist(wide[, paste0("R", i, "_D5_lib")]), col = mycol2[match(wide$EventType, etype)], pch = 20, main = paste0("R", i, " Before Normalization"), xlab = "POS", ylab = "D5_lib_ratio original")
  abline(h = Syn.med1, col = 3, main = paste0("R", "i", "_D5_lib ratio"))
  abline(h = StopGain.med1, col = 2)
  
  plot(wide$POS, unlist(wide[, paste0("R", i, "_D14_lib")]), col = mycol2[match(wide$EventType, etype)], pch = 20, main =  paste0("R", i, "Before Normalization"), xlab = "POS", ylab = "D14_lib_ratio original")
  abline(h = Syn.med1, col = 3, main = paste0("R", i, " D14_lib ratio"))
  abline(h = StopGain.med1, col = 2)
  
  
  plot(wide$POS, unlist(wide[, paste0("D14_lib_ratio_adjusted_R", i)]), col = mycol2[match(wide$EventType, etype)], pch = 20, main = paste0("R", i, " Before Normalization"), xlab = "POS", ylab = "D14_lib_ratio_adjusted")
  abline(h = Syn.med1, col = 3, main =  paste0("R", i, " position adjusted D14_lib ratio"))
  abline(h = StopGain.med1, col = 2)
  plot(wide$POS, unlist(wide[, paste0("D14_lib_ratio_adjusted_R", i, ".1")]), col = mycol2[match(wide$EventType, etype)], pch = 20, main = paste0("R", i, " After Normalization"), xlab = "POS", ylab = "D14_lib_ratio_adjusted_normalized")
  abline(h = Syn.Avg.med, col = 3)
  abline(h = StopGain.Avg.med, col = 2)
  legend(x = "bottom",
         inset = -0.15,
         legend = etype,
         pch =20,
         col =mycol2,
         xpd = TRUE,
         cex = 1, ncol = 3) # Horizontal legend

}


```


#### D14vD5

```{r D14vD5_normalize, include=TRUE, fig.width= length(filenames)*4, fig.height=5}
Syn.med1<- Syn.med2<- Syn.med3 <-Syn.med4 <- StopGain.med1 <- StopGain.med2 <- StopGain.med3 <-StopGain.med4 <- NA

Syn.med1 <- median(wide2$R1_D14_D5[which(wide2$EventType == "Synonymous")], na.rm = T); 
if(length(filenames) >1) {
  Syn.med2 <- median(wide2$R2_D14_D5[which(wide2$EventType == "Synonymous")], na.rm = T); 
  if(length(filenames) > 2){
    Syn.med3 <- median(wide2$R3_D14_D5[which(wide2$EventType == "Synonymous")], na.rm = T); 
    
    if(length(filenames) > 3){
    Syn.med4 <- median(wide2$R4_D14_D5[which(wide2$EventType == "Synonymous")], na.rm = T); 
  }
  }
}

StopGain.med1 <- median(wide2$R1_D14_D5[which(wide2$EventType == "StopGain")], na.rm = T); 
if(length(filenames) >1) {
  StopGain.med2 <- median(wide2$R2_D14_D5[which(wide2$EventType == "StopGain")], na.rm = T); 
  if(length(filenames) > 2){
    StopGain.med3 <- median(wide2$R3_D14_D5[which(wide2$EventType == "StopGain")], na.rm = T);
    if(length(filenames) > 3){
    StopGain.med4 <- median(wide2$R4_D14_D5[which(wide2$EventType == "StopGain")], na.rm = T); 
  }
  }
}

Syn.Avg.med <- mean(c(Syn.med1, Syn.med2, Syn.med3, Syn.med4), na.rm = T) 
StopGain.Avg.med <- mean(c(StopGain.med1, StopGain.med2, StopGain.med3, StopGain.med4), na.rm = T)


wide$R1_D14_D5.1 <- scale1(wide$R1_D14_D5)
if(length(filenames) >1) {
  wide$R2_D14_D5.1 <- scale2(wide$R2_D14_D5) 
  if(length(filenames) > 2){
    wide$R3_D14_D5.1 <- scale3(wide$R3_D14_D5)
      if(length(filenames) > 3){
    wide$R4_D14_D5.1 <- scale4(wide$R4_D14_D5)
  }
  }
}

etype <- unique(wide$EventType)
mycol2 <- mycol[etype]

par(mfrow=c(1, length(filenames)))

plot(wide$R1_D14_D5, wide$R1_D14_D5.1, pch = 20, col = mycol2[match(wide$EventType, etype)], main = "R1 Normalization")
abline(v = c(Syn.med1,StopGain.med1), col = c(3, 2))
abline(h = c(Syn.Avg.med,StopGain.Avg.med), col = c(3, 2))
abline(a=0, b=1)
if(length(filenames) >1) {
  plot(wide$R2_D14_D5, wide$R2_D14_D5.1, pch = 20, col = mycol2[match(wide$EventType, etype)], main = "R2 Normalization")
  abline(v = c(Syn.med2,StopGain.med2), col = c(3, 2))
  abline(h = c(Syn.Avg.med,StopGain.Avg.med), col = c(3, 2))
  abline(a=0, b=1)
 
  if(length(filenames) > 2){
    plot(wide$R3_D14_D5, wide$R3_D14_D5.1, pch = 20, col = mycol2[match(wide$EventType, etype)], main = "R3 Normalization")
    abline(v = c(Syn.med3,StopGain.med3), col = c(3, 2))
    abline(h = c(Syn.Avg.med,StopGain.Avg.med), col = c(3, 2))
    abline(a=0, b=1)
      
    if(length(filenames) > 3){
    plot(wide$R4_D14_D5, wide$R4_D14_D5.1, pch = 20, col = mycol2[match(wide$EventType, etype)], main = "R4 Normalization")
    abline(v = c(Syn.med4,StopGain.med4), col = c(3, 2))
    abline(h = c(Syn.Avg.med,StopGain.Avg.med), col = c(3, 2))
    abline(a=0, b=1)
  }
  }
}


legend(par('usr')[2], par('usr')[4], ##"bottomright",
       legend = etype,
       pch =20,
       col =mycol2,
       cex = 1, ncol = 1) # Horizontal legend
```

Check each experiment

```{r D14vD5_normalize2, include=TRUE, fig.width=10, fig.height=8}

par(mfrow=c(2,1))
plot(wide$POS, wide$R1_D14_D5, col = mycol2[match(wide$EventType, etype)], pch = 20, main = "R1 Before Normalization", xlab = "POS", ylab = "R1_D14_D5")
abline(h = Syn.med1, col = 3, main = "R1  D14_D5 ratio")
abline(h = StopGain.med1, col = 2)
plot(wide$POS, wide$R1_D14_D5.1, col = mycol2[match(wide$EventType, etype)], pch = 20, main = "R1 After Normalization", xlab = "POS", ylab = "R1_D14_D5.1")
abline(h = Syn.Avg.med, col = 3)
abline(h = StopGain.Avg.med, col = 2)
legend(x = "bottom",
       inset = -0.15,
       legend = etype,
       pch =20,
       col =mycol2,
       xpd = TRUE,
       cex = 1, ncol = 3) # Horizontal legend

if(length(filenames) >1) {
  par(mfrow=c(2,1))
  plot(wide$POS, wide$R2_D14_D5, col = mycol2[match(wide$EventType, etype)], pch = 20, main = "R2 Before Normalization", xlab = "POS", ylab = "R2_D14_D5")
  abline(h = Syn.med2, col = 3, main = "R2 D14_D5 ratio")
  abline(h = StopGain.med2, col = 2)
  plot(wide$POS, wide$R2_D14_D5.1, col = mycol2[match(wide$EventType, etype)], pch = 20, main = "R2 After Normalization", xlab = "POS", ylab = "R2_D14_D5.1")
  abline(h = Syn.Avg.med, col = 3)
  abline(h = StopGain.Avg.med, col = 2)
  legend(x = "bottom",
         inset = -0.7, 
         legend = etype,
         pch =20,
         col =mycol2,
         xpd = TRUE, 
         cex = 0.8, ncol = 3) # Horizontal legend
 if(length(filenames) >2) {
    par(mfrow=c(2,1))
    plot(wide$POS, wide$R3_D14_D5, pch = 20, col = mycol2[match(wide$EventType, etype)], main = "R3 Before Normalization", xlab = "POS", ylab = "R3_D14_D5")
    abline(h = Syn.med3, col = 3, main = "R3 D14_D5 ratio")
    abline(h = StopGain.med3, col = 2)
    plot(wide$POS,wide$R3_D14_D5.1, pch = 20, col = mycol2[match(wide$EventType, etype)], main = "R3 After Normalization", xlab = "POS", ylab = "R3_D14_D5.1")
    abline(h = Syn.Avg.med, col = 3)
    abline(h = StopGain.Avg.med, col = 2)
    legend(x = "bottom",
           inset = -0.7, 
           legend = etype,
           pch =20,
           col =mycol2,
           xpd = TRUE, 
           cex = 0.8, ncol = 3) # Horizontal legend
     if(length(filenames) >3) {
    par(mfrow=c(2,1))
    plot(wide$POS, wide$R4_D14_D5, pch = 20, col = mycol2[match(wide$EventType, etype)], main = "R4 Before Normalization", xlab = "POS", ylab = "R4_D14_D5")
    abline(h = Syn.med4, col = 3, main = "R4 D14_D5 ratio")
    abline(h = StopGain.med4, col = 2)
    plot(wide$POS,wide$R4_D14_D5.1, pch = 20, col = mycol2[match(wide$EventType, etype)], main = "R4 After Normalization", xlab = "POS", ylab = "R4_D14_D5.1")
    abline(h = Syn.Avg.med, col = 3)
    abline(h = StopGain.Avg.med, col = 2)
    legend(x = "bottom",
           inset = -0.7, 
           legend = etype,
           pch =20,
           col =mycol2,
           xpd = TRUE, 
           cex = 0.8, ncol = 3) # Horizontal legend
 }
 }
}

```

### C. Pre-Functional Scores

So called Pre-Functional scores is individual experiment scores later in the output noted as R1, R2, R3. They are pulled together to check overall distribution separation of the stopGain vs Synonymous. Also used later for ROC and mixture model checking for pre-classification use (removeA1, removeA2, removeA3, removeA4), which in turn inform the later calculation of the final functional score.

#### Distributions

SNV StopGain and Synonymous only, Including splice region variants 

```{r}
temp1<- temp2 <- temp3 <- temp4 <- temp11 <- temp22 <- temp33 <- temp44 <- NULL

temp1 <- wide %>% mutate(functional.score = D14_lib_ratio_adjusted_R1.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") ) %>% 
    arrange(functional.score)
temp11 <- wide %>% mutate(functional.score = R1_D14_D5.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>% 
    arrange(functional.score)
  
if(length(filenames) >1){
  temp2 <- wide %>% mutate(functional.score = D14_lib_ratio_adjusted_R2.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>% 
    arrange(functional.score)
  temp22 <- wide %>% mutate(functional.score = R2_D14_D5.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>% 
    arrange(functional.score)
}
if(length(filenames)> 2){
  temp3 <- wide %>% mutate(functional.score = D14_lib_ratio_adjusted_R3.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>% 
    arrange(functional.score)
  temp33 <- wide %>% mutate(functional.score = R3_D14_D5.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>% 
    arrange(functional.score)
}

if(length(filenames)> 3){
  temp4 <- wide %>% mutate(functional.score = D14_lib_ratio_adjusted_R4.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>% 
    arrange(functional.score)
  temp44 <- wide %>% mutate(functional.score = R4_D14_D5.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>% 
    arrange(functional.score)
}
temp.1 <- rbind(temp1, temp2, temp3, temp4)
temp.2 <- rbind(temp11, temp22, temp33, temp44)
  
p1 <- temp.1 %>%
  ggplot( aes(x=functional.score, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1) + scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14vsLib")


p2 <- temp.2 %>%
  ggplot( aes(x=functional.score, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1) +  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14_D5")
# 
# combined <- p1 + p2 & theme(legend.position = "bottom")
# (combined + plot_layout(guides = "collect"))

fig <- subplot(ggplotly(p1), ggplotly(p2)) %>% 
  layout(title = 'w. splice region variants: Left: D14_Lib, Right: D14_D5')
fig
```

Excluding spice region

```{r}
temp1<- temp2 <- temp3 <- temp4 <- temp11 <- temp22 <- temp33 <- temp44 <- NULL

temp1 <- wide %>% mutate(functional.score = D14_lib_ratio_adjusted_R1.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>%
    arrange(functional.score)
temp11 <- wide %>% mutate(functional.score = R1_D14_D5.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR == FALSE)%>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") ) %>%
    arrange(functional.score)
  
if(length(filenames) >1){
  temp2 <- wide %>% mutate(functional.score = D14_lib_ratio_adjusted_R2.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>%
    arrange(functional.score)
  temp22 <- wide %>% mutate(functional.score = R2_D14_D5.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>%
    arrange(functional.score)
}
if(length(filenames)> 2){
  temp3 <- wide %>% mutate(functional.score = D14_lib_ratio_adjusted_R3.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>%
    arrange(functional.score)
  temp33 <- wide %>% mutate(functional.score = R3_D14_D5.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>%
    arrange(functional.score)
}
if(length(filenames)> 3){
  temp4 <- wide %>% mutate(functional.score = D14_lib_ratio_adjusted_R4.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR == FALSE)%>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") ) %>%
    arrange(functional.score)
  temp44 <- wide %>% mutate(functional.score = R4_D14_D5.1) %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR == FALSE)%>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") ) %>%
    arrange(functional.score)
}

temp.1 <- rbind(temp1, temp2, temp3, temp4)
temp.2 <- rbind(temp11, temp22, temp33, temp44)
  
p1 <- temp.1 %>% arrange(functional.score) %>%
  ggplot( aes(x=functional.score, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1) + 
  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14vsLib")


p2 <- temp.2 %>%
  ggplot( aes(x=functional.score, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1)  +
  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14_D5")

#combined <- p1 + p2 & theme(legend.position = "bottom")
#(combined + plot_layout(guides = "collect")) 
fig <- subplot(ggplotly(p1), ggplotly(p2)) %>% 
  layout(title = 'wo. splice region variants: Left: D14_Lib, Right: D14_D5')
fig

```


StopGain and Synonymous SNV only, Excluding splice region variants 

```{r}
temp.1  %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR == TRUE) %>% select(EventType  , functional.score)
```


```{r}
temp.2  %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR == TRUE) %>% select(EventType  , functional.score)
```



### D. Classification on Pre-Functional Scores

#### ROC on D14_lib ratios

SNVs excluding splice region, excluding variants on chunling's file.

```{r ROC0, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

library(pROC)

data_long <- wide %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>% filter(EventType %in% c("Synonymous", "StopGain")) %>% mutate(functional.score = D14_lib_ratio_adjusted_R1.1) %>% filter(uPOS %in% wide2$uPOS)
if(length(filenames) >1){
  temp <- wide %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>% filter(EventType %in% c("Synonymous", "StopGain")) %>% mutate(functional.score = D14_lib_ratio_adjusted_R2.1)
  data_long <- rbind(data_long, temp)
}

if(length(filenames) >2){
  temp <- wide %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>% filter(EventType %in% c("Synonymous", "StopGain")) %>% mutate(functional.score = D14_lib_ratio_adjusted_R3.1)
  data_long <- rbind(data_long, temp)
}

if(length(filenames) >3){
  temp <- wide %>% filter(spliceR == FALSE)%>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") ) %>% filter(EventType %in% c("Synonymous", "StopGain")) %>% mutate(functional.score = D14_lib_ratio_adjusted_R4.1)
  data_long <- rbind(data_long, temp)
}

par(mfrow = c(1, 1))
roc_score = roc(controls = data_long$functional.score[data_long$EventType == "Synonymous"], cases = data_long$functional.score[data_long$EventType == "StopGain"], direction = ">")  # AUC score
plot(roc_score, main="ROC curve")
text(0.2, 0.2, paste0("AUC: ", round(roc_score$auc, 3)))

out <- data.frame(functional.score = roc_score$"threshold", sens =  roc_score$"sensitivities", specs = roc_score$"specificities")
options(digits = 21)
out$avg <- out$sens+out$specs

ind <- which(out$avg == max(out$avg))
if(length(ind) == 1){
  cutoff0 <- unique(out$functional.score[ind])
  sen0 <- out$sens[ind]
  spec0 <- out$specs[ind]

} else{
  cutoff0 <- unique(out$functional.score[ind[1]])
  sen0 <- out$sens[ind[1]]
  spec0 <- out$specs[ind[1]]
}
auc0 <- roc_score$auc


data_long$pred.class <- ifelse(data_long$functional.score < cutoff0, "StopGain", "Synonymous")

tb0 <- table(truth = data_long$EventType, pred = data_long$pred.class)
tb0

```

`r cutoff0` was used for the performance. sensitivity is `r sen0`, specificity is `r spec0`. 
Note, pre-classification was done on replicates pooled together. The number of the variants were doubled or tripled in the 2x2 table. 

#### ROC on D14_D5 ratios

SNVs excluding splice region

```{r ROC0.1, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

library(pROC)

data_long <- wide %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") )%>% filter(EventType %in% c("Synonymous", "StopGain")) %>% mutate(functional.score = R1_D14_D5.1) %>% filter(uPOS %in% wide2$uPOS)

if(length(filenames) >1){
  temp <- wide %>% filter(spliceR == FALSE)%>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") ) %>% filter(EventType %in% c("Synonymous", "StopGain")) %>% mutate(functional.score = R2_D14_D5.1)
  data_long <- rbind(data_long, temp)
}

if(length(filenames) >2){
  temp <- wide %>% filter(spliceR == FALSE)%>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") ) %>% filter(EventType %in% c("Synonymous", "StopGain")) %>% mutate(functional.score = R3_D14_D5.1)
  data_long <- rbind(data_long, temp)
}
if(length(filenames) >3){
  temp <- wide %>% filter(spliceR == FALSE)%>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G") ) %>% filter(EventType %in% c("Synonymous", "StopGain")) %>% mutate(functional.score = R4_D14_D5.1)
  data_long <- rbind(data_long, temp)
}
par(mfrow = c(1, 1))
roc_score = roc(controls = data_long$functional.score[data_long$EventType == "Synonymous"], cases = data_long$functional.score[data_long$EventType == "StopGain"], direction = ">")  # AUC score
plot(roc_score, main="ROC curve")
text(0.2, 0.2, paste0("AUC: ", round(roc_score$auc, 3)))

out <- data.frame(functional.score = roc_score$"threshold", sens =  roc_score$"sensitivities", specs = roc_score$"specificities")

out$avg <- out$sens+out$specs
ind <- which(out$avg == max(out$avg))
if(length(ind) == 1){
  cutoff1 <- unique(out$functional.score[ind])
  sen1 <- out$sens[ind]
  spec1 <- out$specs[ind]

} else{
  cutoff1 <- unique(out$functional.score[ind[1]])
  sen1 <- out$sens[ind[1]]
  spec1 <- out$specs[ind[1]]
}

auc1 <- roc_score$auc

data_long$pred.class <- ifelse(data_long$functional.score < cutoff1, "StopGain", "Synonymous")

tb1 <- table(truth = data_long$EventType, pred =  data_long$pred.class)
tb1

```

`r cutoff1` was used for the performance. sensitivity is `r sen1`, specificity is `r spec1`. 
Note, pre-classification was done on replicates pooled together. The number of the variants were doubled or tripled in the 2x2 table. 


#### Mixture model on D14_lib_adjusted

Based on distribution from SNA only StopGain and synonymous without splicing variants, also exclude variants on Chunling's list, build mixture model.

NOTE::: 

The functional score cut offs are from the posterior probability curve. It highly depends on the actual data we have. In some senarios, we may not be able to get the values of corresponding functional scores of certain percentage. 

0. IF multiple experiments, we will pool together. 

1. IF the curve or posterior probability tails up to the right, there will not be a good value for 1 or 2 or 5% cutoff. 

2. IF the front end points are sparse, there may not be 99%, 98% or 95% cutoff neither. 



```{r MixModel1, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

##normalmixEM function of the mixtools package in R
library(mixtools)
f1 <- f2 <- f3 <- f4 <- NA

temp <- temp.1 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% 
  arrange(functional.score) %>% filter(uPOS %in% wide2$uPOS)

m1 <- mean(temp$functional.score[temp$EventType != "StopGain"], na.rm = TRUE)
tmp <- temp$functional.score[temp$EventType != "StopGain"]
tmp2 <- tmp ## tmp[which(tmp > median(tmp)-sd(tmp) & tmp < median(tmp)+sd(tmp))]

v1 <- sd(tmp2, na.rm = TRUE)


m2 <- mean(temp$functional.score[temp$EventType == "StopGain"])
v2 <- sd(temp$functional.score[temp$EventType == "StopGain"])
l1 <- sum(temp$EventType == "StopGain")/sum(temp$EventType != "StopGain")
an.error.occured <- FALSE
tryCatch( {
  x <- normalmixEM(temp$functional.score, 
                   lambda = l1, k = 2,
                   mean.constr = c(m2, m1),
                  sd.constr = c(v2, v1),
                 mu = c(m2, m1), 
                 sigma = c(v2, v1))

par(mfrow = c(1, 1))
plot(temp$functional.score, x$posterior[,1], col = ifelse(temp$EventType == "StopGain", 2, 1))
f1 <- temp$functional.score[x$posterior[,1]>0.01][order(temp$functional.score[x$posterior[,1]>0.01], decreasing = T)][1]
f2 <- temp$functional.score[x$posterior[,1]>0.99][order(temp$functional.score[x$posterior[,1]>0.99], decreasing = T)][1]

f3 <- temp$functional.score[x$posterior[,1]>0.02][order(temp$functional.score[x$posterior[,1]>0.02], decreasing = T)][1]
f4 <- temp$functional.score[x$posterior[,1]>0.98][order(temp$functional.score[x$posterior[,1]>0.98], decreasing = T)][1]


abline(v = f1)
abline(v = f2)
abline(h = 0.01)
abline(h = 0.99)

abline(v = f3, lty = 2)
abline(v = f4, lty = 2)
abline(h = 0.02, lty = 2)
abline(h = 0.98, lty = 2)

par(mfrow = c(1, 1))
plot(x, which=2, xlab2= "", yaxt= "n", xlim = c(range(x$x)[1]-2, range(x$x)[2]+2), main2 ="Two components with 1-99% prob(nonfunctional)", lwd2=0.8, border = "azure3")
lines(density(x$x), lty=2, lwd=0.8)

abline(v = f1)
abline(v = f2)
abline(v = f3, lty = 2)
abline(v = f4, lty = 2)

}
          , error = function(e) {an.error.occured <<- TRUE})
print(an.error.occured)


```

`r paste0("There is",  ifelse(an.error.occured, " AN ", " NO "), "Error Ocurred")`.

`r f2` was the functional score corresponding to the probability of non functional at 99% and `r f1` was the functional score corresponding to the probability of non functional at 1%.

`r f4` was the functional score corresponding to the probability of non functional at 98% and `r f3` was the functional score corresponding to the probability of non functional at 2%.

If we use these two thresholds to classfiy the variants of synonymous, stopgain and missense. 

First, 1-99% cutoff
Second, 2-98% cutoff

Note: Splice regions are included in Missense, not stopgain nor synonymous

```{r MixModel1.2, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

temp <- wide %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G"))


if(!an.error.occured){
  temp$ctype.R1 <- ifelse(temp$D14_lib_ratio_adjusted_R1.1>f1, "neutural", 
                       ifelse(temp$D14_lib_ratio_adjusted_R1.1>f2, "intermediate", "pathogenic"))
  
  
  temp$ctype2.R1 <- ifelse(temp$D14_lib_ratio_adjusted_R1.1>f3, "neutural", 
                       ifelse(temp$D14_lib_ratio_adjusted_R1.1>f4, "intermediate", "pathogenic"))
  
  
  temp$ctype.R1 <- factor(temp$ctype.R1, levels = c("pathogenic", "intermediate", "neutural"))
  temp$ctype2.R1 <- factor(temp$ctype2.R1, levels = c("pathogenic", "intermediate", "neutural"))
  temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))
  
  temp.check <- temp %>% filter((EventType %in% c("StopGain", "Synonymous") & spliceR == FALSE) | EventType == "Missense")
    
  table(truth = temp.check$EventType, pred = temp.check$ctype.R1)
  
  table(truth = temp.check$EventType, pred = temp.check$ctype2.R1)
  
  if(length(filenames)>1){
    temp$ctype.R2 <- ifelse(temp$D14_lib_ratio_adjusted_R2.1>f1, "neutural", 
                       ifelse(temp$D14_lib_ratio_adjusted_R2.1>f2, "intermediate", "pathogenic"))
    
    
    temp$ctype2.R2 <- ifelse(temp$D14_lib_ratio_adjusted_R2.1>f3, "neutural", 
                             ifelse(temp$D14_lib_ratio_adjusted_R2.1>f4, "intermediate", "pathogenic"))
    
    
    temp$ctype.R2 <- factor(temp$ctype.R2, levels = c("pathogenic", "intermediate", "neutural"))
    temp$ctype2.R2 <- factor(temp$ctype2.R2, levels = c("pathogenic", "intermediate", "neutural"))
    temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))
    
    temp.check <- temp %>% filter((EventType %in% c("StopGain", "Synonymous") & spliceR == FALSE) | EventType == "Missense")
    
    table(truth =  temp.check$EventType, pred = temp.check$ctype.R2)
    
    table(truth = temp.check$EventType, pred = temp.check$ctype2.R2)
    
  }
  
  if(length(filenames)>2){
    temp$ctype.R3 <- ifelse(temp$D14_lib_ratio_adjusted_R3.1>f1, "neutural", 
                       ifelse(temp$D14_lib_ratio_adjusted_R3.1>f2, "intermediate", "pathogenic"))
    
    
    temp$ctype2.R3 <- ifelse(temp$D14_lib_ratio_adjusted_R3.1>f3, "neutural", 
                             ifelse(temp$D14_lib_ratio_adjusted_R3.1>f4, "intermediate", "pathogenic"))
    
    
    temp$ctype.R3 <- factor(temp$ctype.R3, levels = c("pathogenic", "intermediate", "neutural"))
    temp$ctype2.R3 <- factor(temp$ctype2.R3, levels = c("pathogenic", "intermediate", "neutural"))
    temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))
    
    temp.check <- temp %>% filter((EventType %in% c("StopGain", "Synonymous") & spliceR == FALSE) | EventType == "Missense")
    table(truth = temp.check$EventType, pred = temp.check$ctype.R3)
    
    table(truth = temp.check$EventType, pred = temp.check$ctype2.R3)
    
  }
  
  if(length(filenames)>3){
    temp$ctype.R4 <- ifelse(temp$D14_lib_ratio_adjusted_R4.1>f1, "neutural", 
                       ifelse(temp$D14_lib_ratio_adjusted_R4.1>f2, "intermediate", "pathogenic"))
    
    
    temp$ctype2.R4 <- ifelse(temp$D14_lib_ratio_adjusted_R4.1>f3, "neutural", 
                             ifelse(temp$D14_lib_ratio_adjusted_R4.1>f4, "intermediate", "pathogenic"))
    
    
    temp$ctype.R4 <- factor(temp$ctype.R4, levels = c("pathogenic", "intermediate", "neutural"))
    temp$ctype2.R4 <- factor(temp$ctype2.R4, levels = c("pathogenic", "intermediate", "neutural"))
    temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))
    
    temp.check <- temp %>% filter((EventType %in% c("StopGain", "Synonymous") & spliceR == FALSE) | EventType == "Missense")
    table(truth = temp.check$EventType, pred = temp.check$ctype.R4)
    
    table(truth = temp.check$EventType, pred = temp.check$ctype2.R4)
    
  }
}

```

Pooled together if more than one Experiment.

```{r}

if(!an.error.occured){
  if(length(filenames)==2){
    table(truth =  c(temp$EventType, temp$EventType), pred = c(temp$ctype.R1, temp$ctype.R2))
    table(truth = c(temp$EventType, temp$EventType), pred = c(temp$ctype2.R1, temp$ctype2.R2))
  } 
  
  if(length(filenames) ==3){
    table(truth = c(temp$EventType, temp$EventType, temp$EventType), pred = c(temp$ctype.R1, temp$ctype.R2 ,temp$ctype.R3))
    table(truth = c(temp$EventType, temp$EventType, temp$EventType), pred = c(temp$ctype2.R1, temp$ctype2.R2 ,temp$ctype2.R3))
  }
  if(length(filenames) ==4){
    table(truth = c(temp$EventType, temp$EventType, temp$EventType, temp$EventType), pred = c(temp$ctype.R1, temp$ctype.R2 ,temp$ctype.R3, temp$ctype.R4))
    table(truth = c(temp$EventType, temp$EventType, temp$EventType, temp$EventType), pred = c(temp$ctype2.R1, temp$ctype2.R2 ,temp$ctype2.R3 ,temp$ctype2.R4))
  }
}
  

```

```{r}


## added pre-classification

if(length(filenames) ==1){
  final1 <- wide %>% rename(R1 = D14_lib_ratio_adjusted_R1.1) %>%
    mutate(pred.class_R1 = ifelse(R1 < cutoff0, "StopGain", "Synonymous")) 
}
if(length(filenames) ==2){
  final1 <- wide %>% rename(R1 = D14_lib_ratio_adjusted_R1.1,
                              R2 = D14_lib_ratio_adjusted_R2.1) %>%
    mutate(pred.class_R1 = ifelse(R1 < cutoff0, "StopGain", "Synonymous"),
           pred.class_R2 = ifelse(R2 < cutoff0, "StopGain", "Synonymous")) %>%
    mutate(diff12 = R1-R2) %>%
    mutate(meanR =  unlist(select(., R1:R2) %>% 
                           pmap(~ mean(c(...), na.rm = T)))) %>%
    mutate(sdR =  unlist(select(., R1:R2) %>% 
                         pmap(~ sd(c(...), na.rm = T))) ) %>% 
    mutate(cv =  NA) %>% 
    mutate(removeA = (abs(diff12) >C_fc) & pred.class_R1 != pred.class_R2) %>% 
      mutate(removeC = (abs(diff12) >C_fc))
  
} 

if(length(filenames) ==3 ){
  final1  <- wide %>% rename(R1 = D14_lib_ratio_adjusted_R1.1,
                             R2 = D14_lib_ratio_adjusted_R2.1, 
                             R3 = D14_lib_ratio_adjusted_R3.1) %>%
  mutate(pred.class_R1 = ifelse(R1 < cutoff0, "StopGain", "Synonymous"),
         pred.class_R2 = ifelse(R2 < cutoff0, "StopGain", "Synonymous"),
         pred.class_R3 = ifelse(R3 < cutoff0, "StopGain", "Synonymous")) %>%
  mutate(diff12 = R1-R2, diff23 = R2-R3, diff13 = R1-R3) %>%
  mutate(meanR =  unlist(select(., R1:R3) %>% 
                         pmap(~ mean(c(...), na.rm = T)))) %>%
  mutate(sdR =  unlist(select(., R1:R3) %>% 
                         pmap(~ sd(c(...), na.rm = T))) ) %>% 
  mutate(cv = NA) %>%
  mutate(removeA1 = ((abs(diff12) >C_fc | abs(diff13) >C_fc) & pred.class_R1 != pred.class_R2 & pred.class_R1 != pred.class_R3),
         removeA2 = ((abs(diff12) >C_fc | abs(diff23) >C_fc) & pred.class_R2 != pred.class_R1 & pred.class_R2 != pred.class_R3),
         removeA3 = ((abs(diff23) >C_fc | abs(diff13) >C_fc) & pred.class_R3 != pred.class_R2 & pred.class_R1 != pred.class_R3)) %>% 
  mutate(removeC1 = (abs(diff12) >C_fc & abs(diff13) >C_fc),
         removeC2 = (abs(diff12) >C_fc & abs(diff23) >C_fc),
         removeC3 = (abs(diff23) >C_fc & abs(diff13) >C_fc))
}

if(length(filenames) ==4 ){
  final1  <- wide %>% rename(R1 = D14_lib_ratio_adjusted_R1.1,
                             R2 = D14_lib_ratio_adjusted_R2.1, 
                             R3 = D14_lib_ratio_adjusted_R3.1,
                             R4 = D14_lib_ratio_adjusted_R4.1) %>%
  mutate(pred.class_R1 = ifelse(R1 < cutoff0, "StopGain", "Synonymous"),
         pred.class_R2 = ifelse(R2 < cutoff0, "StopGain", "Synonymous"),
         pred.class_R3 = ifelse(R3 < cutoff0, "StopGain", "Synonymous"),
         pred.class_R4 = ifelse(R3 < cutoff0, "StopGain", "Synonymous")) %>%
  mutate(diff12 = R1-R2, diff23 = R2-R3, diff13 = R1-R3, diff24 = R2-R4, diff14 = R1-R4, diff34 = R3-R4) %>%
  mutate(meanR =  unlist(select(., R1:R4) %>% 
                         pmap(~ mean(c(...), na.rm = T)))) %>%
  mutate(sdR =  unlist(select(., R1:R4) %>% 
                         pmap(~ sd(c(...), na.rm = T))) ) %>% 
  mutate(cv = NA) %>%
  mutate(removeA1 = ((abs(diff12) >C_fc | abs(diff13) >C_fc | abs(diff14) >C_fc) & pred.class_R1 != pred.class_R2 & pred.class_R1 != pred.class_R3 & pred.class_R1 != pred.class_R4),
         removeA2 = ((abs(diff12) >C_fc | abs(diff23) >C_fc | abs(diff24) >C_fc) & pred.class_R2 != pred.class_R1 & pred.class_R2 != pred.class_R3 & pred.class_R2 != pred.class_R4),
         removeA3 = ((abs(diff23) >C_fc | abs(diff13) >C_fc | abs(diff34) >C_fc) & pred.class_R3 != pred.class_R2 & pred.class_R1 != pred.class_R3 & pred.class_R3 != pred.class_R4),
         removeA4 = ((abs(diff24) >C_fc | abs(diff14) >C_fc | abs(diff34) >C_fc) & pred.class_R4 != pred.class_R1 & pred.class_R4 != pred.class_R2 & pred.class_R4 != pred.class_R3)) %>% 
  mutate(removeC1 = (abs(diff12) >C_fc & abs(diff13) >C_fc& abs(diff14) >C_fc),
         removeC2 = (abs(diff12) >C_fc & abs(diff23) >C_fc& abs(diff24) >C_fc),
         removeC3 = (abs(diff23) >C_fc & abs(diff13) >C_fc& abs(diff34) >C_fc),
         removeC4 = (abs(diff14) >C_fc & abs(diff24) >C_fc& abs(diff34) >C_fc))
}

## add scores.
if(length(filenames) == 1){
  final1$functional.scoreA <- final1$R1
  final1$functional.scoreB <- final1$R1
  final1$functional.scoreC <- final1$R1
}

if(length(filenames) ==2){

  temp <- final1 %>% select(R1, R2) 

  temp$R1[which(final1$removeA == TRUE)] <- NA
  temp$R2[which(final1$removeA == TRUE)] <- NA
  
  final1$functional.scoreA <- apply(temp, 1, mean, na.rm = TRUE)
  
  final1$functional.scoreB <- final1$meanR

  temp <- final1 %>% select(R1, R2) 
  temp$R1[which(final1$removeC == TRUE)] <- NA
  temp$R2[which(final1$removeC == TRUE)] <- NA
  
  final1$functional.scoreC <- apply(temp, 1, mean, na.rm = TRUE)
  
}
if(length(filenames) == 3){

  temp <- final1 %>% select(R1, R2, R3) 

  temp$R1[which(final1$removeA1 == TRUE)] <- NA
  temp$R2[which(final1$removeA2 == TRUE)] <- NA
  temp$R3[which(final1$removeA3 == TRUE)] <- NA
  
  
  final1$functional.scoreA <- apply(temp, 1, mean, na.rm = TRUE)

  final1$functional.scoreB <- final1$meanR
  
  temp <- final1 %>% select(R1, R2, R3) 
  temp$R1[which(final1$removeC1 == TRUE)] <- NA
  temp$R2[which(final1$removeC2 == TRUE)] <- NA
  temp$R3[which(final1$removeC3 == TRUE)] <- NA
  
  final1$functional.scoreC <- apply(temp, 1, mean, na.rm = TRUE)
}
if(length(filenames) == 4){

  temp <- final1 %>% select(R1, R2, R3, R4) 

  temp$R1[which(final1$removeA1 == TRUE)] <- NA
  temp$R2[which(final1$removeA2 == TRUE)] <- NA
  temp$R3[which(final1$removeA3 == TRUE)] <- NA
  temp$R4[which(final1$removeA4 == TRUE)] <- NA
  
  
  final1$functional.scoreA <- apply(temp, 1, mean, na.rm = TRUE)

  final1$functional.scoreB <- final1$meanR
 
  temp <- final1 %>% select(R1, R2, R3, R4) 
  temp$R1[which(final1$removeC1 == TRUE)] <- NA
  temp$R2[which(final1$removeC2 == TRUE)] <- NA
  temp$R3[which(final1$removeC3 == TRUE)] <- NA
  temp$R4[which(final1$removeC4 == TRUE)] <- NA
  
  final1$functional.scoreC <- apply(temp, 1, mean, na.rm = TRUE)
}
final1$cv[which(!is.na(final1$meanR))] <- unlist(final1$sdR[which(!is.na(final1$meanR))]) / unlist(final1$meanR[which(!is.na(final1$meanR))])* 100 


```




#### Mixture model on D14_D5

Based on distribution from StopGain and synonymous without splicing variants and exlude variants on Chunling's list, build mixture model.

NOTE::: 

The functional score cut offs are from the posterior probability curve. It highly depends on the actual data we have. In some senarios, we may not be able to get the values of corresponding functional scores of certain percentage. 

1. IF the curve or posterior probability tails up to the right, there will not be a good value for 1 or 2 or 5% cutoff. 

2. IF the front end points are sparse, there may not be 99%, 98% or 95% cutoff neither. 



```{r D14D5_MixModel1, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

##normalmixEM function of the mixtools package in R
library(mixtools)

temp <- temp.2 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% 
  arrange(functional.score) %>% filter(uPOS %in% wide2$uPOS)

m1 <- mean(temp$functional.score[temp$EventType != "StopGain"], na.rm = TRUE)
tmp <- temp$functional.score[temp$EventType != "StopGain"]
tmp2 <- tmp ## tmp[which(tmp > median(tmp)-sd(tmp) & tmp < median(tmp)+sd(tmp))]

v1 <- sd(tmp2, na.rm = T)

f1 <- f2 <- f3 <- f4 <- NA

m2 <- mean(temp$functional.score[temp$EventType == "StopGain"])
v2 <- sd(temp$functional.score[temp$EventType == "StopGain"])
l1 <- sum(temp$EventType == "StopGain")/sum(temp$EventType != "StopGain")
an.error.occured <- FALSE
tryCatch( {
  x <- normalmixEM(temp$functional.score, 
                   lambda = l1, k = 2,
                   mean.constr = c(m2, m1),
                  sd.constr = c(v2, v1),
                 mu = c(m2, m1), 
                 sigma = c(v2, v1))

par(mfrow = c(1, 1))
plot(temp$functional.score, x$posterior[,1], col = ifelse(temp$EventType == "StopGain", 2, 1))
f1 <- temp$functional.score[x$posterior[,1]>0.01][order(temp$functional.score[x$posterior[,1]>0.01], decreasing = T)][1]
f2 <- temp$functional.score[x$posterior[,1]>0.99][order(temp$functional.score[x$posterior[,1]>0.99], decreasing = T)][1]

f3 <- temp$functional.score[x$posterior[,1]>0.02][order(temp$functional.score[x$posterior[,1]>0.02], decreasing = T)][1]
f4 <- temp$functional.score[x$posterior[,1]>0.98][order(temp$functional.score[x$posterior[,1]>0.98], decreasing = T)][1]


abline(v = f1)
abline(v = f2)
abline(h = 0.01)
abline(h = 0.99)

abline(v = f3, lty = 2)
abline(v = f4, lty = 2)
abline(h = 0.02, lty = 2)
abline(h = 0.98, lty = 2)

par(mfrow = c(1, 1))
plot(x, which=2, xlab2= "", yaxt= "n", xlim = c(range(x$x)[1]-2, range(x$x)[2]+2), main2 ="Two components with 1-99% prob(nonfunctional)", lwd2=0.8, border = "azure3")
lines(density(x$x), lty=2, lwd=0.8)

abline(v = f1)
abline(v = f2)
abline(v = f3, lty = 2)
abline(v = f4, lty = 2)
 }
          , error = function(e) {an.error.occured <<- TRUE})
print(an.error.occured)
```

`r paste0("There is",  ifelse(an.error.occured, " an ", " NO "), "Error Ocurred when fitting mixture model")`.


`r f2` was the functional score corresponding to the probability of non functional at 99% and `r f1` was the functional score corresponding to the probability of non functional at 1%.

`r f4` was the functional score corresponding to the probability of non functional at 98% and `r f3` was the functional score corresponding to the probability of non functional at 2%.

If we use these two thresholds to classfiy the variants of synonymous, stopgain and missense. 

First, 1-99% cutoff
Second, 2-98% cutoff

Note: Splice regions are included in Missense, not stopgain nor synonymous

```{r D14_D5_MixModel1.2, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}
temp <- wide %>% filter(spliceR == FALSE) %>% filter(EventType %in% c("Synonymous", "StopGain")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G"))

temp2 <- wide %>% filter(EventType %in% c("Missense")) %>%  filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G"))


temp <- rbind(temp, temp2)

if(!an.error.occured){
  temp$ctype.R1 <- ifelse(temp$R1_D14_D5.1>f1, "neutural", 
                       ifelse(temp$R1_D14_D5.1>f2, "intermediate", "pathogenic"))
  
  
  temp$ctype2.R1 <- ifelse(temp$R1_D14_D5.1>f3, "neutural", 
                       ifelse(temp$R1_D14_D5.1>f4, "intermediate", "pathogenic"))
  
  
  temp$ctype.R1 <- factor(temp$ctype.R1, levels = c("pathogenic", "intermediate", "neutural"))
  temp$ctype2.R1 <- factor(temp$ctype2.R1, levels = c("pathogenic", "intermediate", "neutural"))
  temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))
  
  table( truth = temp$EventType, pred = temp$ctype.R1)
  
  table(truth = temp$EventType, pred = temp$ctype2.R1)
  
  if(length(filenames)>1){
    temp$ctype.R2 <- ifelse(temp$R2_D14_D5.1>f1, "neutural", 
                       ifelse(temp$R2_D14_D5.1>f2, "intermediate", "pathogenic"))
    
    
    temp$ctype2.R2 <- ifelse(temp$R2_D14_D5.1>f3, "neutural", 
                             ifelse(temp$R2_D14_D5.1>f4, "intermediate", "pathogenic"))
    
    
    temp$ctype.R2 <- factor(temp$ctype.R2, levels = c("pathogenic", "intermediate", "neutural"))
    temp$ctype2.R2 <- factor(temp$ctype2.R2, levels = c("pathogenic", "intermediate", "neutural"))
    temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))
    
    table(truth= temp$EventType, pred =  temp$ctype.R2)
    
    table(truth = temp$EventType, pred = temp$ctype2.R2)
    
  }
  
  if(length(filenames)>2){
    temp$ctype.R3 <- ifelse(temp$R3_D14_D5.1>f1, "neutural", 
                       ifelse(temp$R3_D14_D5.1>f2, "intermediate", "pathogenic"))
    
    
    temp$ctype2.R3 <- ifelse(temp$R3_D14_D5.1>f3, "neutural", 
                             ifelse(temp$R3_D14_D5.1>f4, "intermediate", "pathogenic"))
    
    
    temp$ctype.R3 <- factor(temp$ctype.R3, levels = c("pathogenic", "intermediate", "neutural"))
    temp$ctype2.R3 <- factor(temp$ctype2.R3, levels = c("pathogenic", "intermediate", "neutural"))
    temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))
    
    table(truth = temp$EventType, pred = temp$ctype.R3)
    
    table(truth = temp$EventType, pred = temp$ctype2.R3)
    
  }
  if(length(filenames)>3){
    temp$ctype.R4 <- ifelse(temp$R4_D14_D5.1>f1, "neutural", 
                       ifelse(temp$R4_D14_D5.1>f2, "intermediate", "pathogenic"))
    
    
    temp$ctype2.R4 <- ifelse(temp$R4_D14_D5.1>f3, "neutural", 
                             ifelse(temp$R4_D14_D5.1>f4, "intermediate", "pathogenic"))
    
    
    temp$ctype.R4 <- factor(temp$ctype.R4, levels = c("pathogenic", "intermediate", "neutural"))
    temp$ctype2.R4 <- factor(temp$ctype2.R4, levels = c("pathogenic", "intermediate", "neutural"))
    temp$EventType <-factor(temp$EventType, levels = c("StopGain", "Missense", "Synonymous"))
    
    table(truth = temp$EventType, pred = temp$ctype.R4)
    
    table(truth = temp$EventType, pred = temp$ctype2.R4)
    
  }
}

```


Pooled together if more than one experiment.

```{r}
if(!an.error.occured){
  if(length(filenames)==2){
    table(truth = c(temp$EventType, temp$EventType), pred = c(temp$ctype.R1, temp$ctype.R2))
    table(truth = c(temp$EventType, temp$EventType), pred = c(temp$ctype2.R1, temp$ctype2.R2))
  } 
  
  if(length(filenames) ==3){
    table(truth = c(temp$EventType, temp$EventType, temp$EventType), pred = c(temp$ctype.R1, temp$ctype.R2 ,temp$ctype.R3))
    table(truth = c(temp$EventType, temp$EventType, temp$EventType), pred = c(temp$ctype2.R1, temp$ctype2.R2 ,temp$ctype2.R3))
  }
  if(length(filenames) ==4){
    table(truth = c(temp$EventType, temp$EventType, temp$EventType, temp$EventType), pred = c(temp$ctype.R1, temp$ctype.R2 ,temp$ctype.R3,temp$ctype.R4))
    table(truth = c(temp$EventType, temp$EventType, temp$EventType, temp$EventType), pred = c(temp$ctype2.R1, temp$ctype2.R2 ,temp$ctype2.R3,temp$ctype2.R4))
  }
}
```

```{r}
## added pre-classification

if(length(filenames) ==1){
  final2 <- wide %>% rename(R1 = R1_D14_D5.1) %>%
    mutate(pred.class_R1 = ifelse(R1 < cutoff1, "StopGain", "Synonymous")) 
}
if(length(filenames) ==2){
  final2 <- wide %>% rename(R1 = R1_D14_D5.1,
                              R2 = R2_D14_D5.1) %>%
    mutate(pred.class_R1 = ifelse(R1 < cutoff1, "StopGain", "Synonymous"),
           pred.class_R2 = ifelse(R2 < cutoff1, "StopGain", "Synonymous")) %>%
    mutate(diff12 = R1-R2) %>%
    mutate(meanR =  unlist(select(., R1:R2) %>% 
                           pmap(~ mean(c(...), na.rm = T)))) %>%
    mutate(sdR =  unlist(select(., R1:R2) %>% 
                         pmap(~ sd(c(...), na.rm = T))) ) %>% 
    mutate(cv =  NA) %>% 
    mutate(removeA = (abs(diff12) >C_fc) & pred.class_R1 != pred.class_R2) %>% 
      mutate(removeC = (abs(diff12) >C_fc))
  
} 

if(length(filenames) ==3 ){
  final2  <- wide %>% rename(R1 = R1_D14_D5.1,
                             R2 = R2_D14_D5.1, 
                             R3 = R3_D14_D5.1) %>%
  mutate(pred.class_R1 = ifelse(R1 < cutoff1, "StopGain", "Synonymous"),
         pred.class_R2 = ifelse(R2 < cutoff1, "StopGain", "Synonymous"),
         pred.class_R3 = ifelse(R3 < cutoff1, "StopGain", "Synonymous")) %>%
  mutate(diff12 = R1-R2, diff23 = R2-R3, diff13 = R1-R3) %>%
  mutate(meanR =  unlist(select(., R1:R3) %>% 
                         pmap(~ mean(c(...), na.rm = T)))) %>%
  mutate(sdR =  unlist(select(., R1:R3) %>% 
                         pmap(~ sd(c(...), na.rm = T))) ) %>% 
  mutate(cv = NA) %>%
  mutate(removeA1 = ((abs(diff12) >C_fc | abs(diff13) >C_fc) & pred.class_R1 != pred.class_R2 & pred.class_R1 != pred.class_R3),
         removeA2 = ((abs(diff12) >C_fc | abs(diff23) >C_fc) & pred.class_R2 != pred.class_R1 & pred.class_R2 != pred.class_R3),
         removeA3 = ((abs(diff23) >C_fc | abs(diff13) >C_fc) & pred.class_R3 != pred.class_R2 & pred.class_R1 != pred.class_R3)) %>% 
  mutate(removeC1 = (abs(diff12) >C_fc & abs(diff13) >C_fc),
         removeC2 = (abs(diff12) >C_fc & abs(diff23) >C_fc),
         removeC3 = (abs(diff23) >C_fc & abs(diff13) >C_fc))
}

if(length(filenames) ==4 ){
  final2  <- wide %>% rename(R1 = R1_D14_D5.1,
                             R2 = R2_D14_D5.1, 
                             R3 = R3_D14_D5.1,
                             R4 = R4_D14_D5.1) %>%
  mutate(pred.class_R1 = ifelse(R1 < cutoff1, "StopGain", "Synonymous"),
         pred.class_R2 = ifelse(R2 < cutoff1, "StopGain", "Synonymous"),
         pred.class_R3 = ifelse(R3 < cutoff1, "StopGain", "Synonymous"),
         pred.class_R4 = ifelse(R4 < cutoff1, "StopGain", "Synonymous")) %>%
  mutate(diff12 = R1-R2, diff23 = R2-R3, diff13 = R1-R3,diff14 = R1-R4, diff24 = R2-R4, diff34 = R3-R4) %>%
  mutate(meanR =  unlist(select(., R1:R4) %>% 
                         pmap(~ mean(c(...), na.rm = T)))) %>%
  mutate(sdR =  unlist(select(., R1:R4) %>% 
                         pmap(~ sd(c(...), na.rm = T))) ) %>% 
  mutate(cv = NA) %>%
  mutate(removeA1 = ((abs(diff12) >C_fc | abs(diff13) >C_fc | abs(diff14) >C_fc) & pred.class_R1 != pred.class_R2 & pred.class_R1 != pred.class_R3 & pred.class_R1 != pred.class_R4),
         removeA2 = ((abs(diff12) >C_fc | abs(diff23) >C_fc | abs(diff24) >C_fc) & pred.class_R2 != pred.class_R1 & pred.class_R2 != pred.class_R3 & pred.class_R2 != pred.class_R4),
         removeA3 = ((abs(diff23) >C_fc | abs(diff13) >C_fc | abs(diff34) >C_fc) & pred.class_R3 != pred.class_R2 & pred.class_R1 != pred.class_R3 & pred.class_R3 != pred.class_R4),
         removeA4 = ((abs(diff24) >C_fc | abs(diff14) >C_fc | abs(diff34) >C_fc) & pred.class_R4 != pred.class_R1 & pred.class_R4 != pred.class_R2 & pred.class_R4 != pred.class_R3)) %>% 
  mutate(removeC1 = (abs(diff12) >C_fc & abs(diff13) >C_fc& abs(diff14) >C_fc),
         removeC2 = (abs(diff12) >C_fc & abs(diff23) >C_fc& abs(diff24) >C_fc),
         removeC3 = (abs(diff23) >C_fc & abs(diff13) >C_fc& abs(diff34) >C_fc),
         removeC4 = (abs(diff14) >C_fc & abs(diff24) >C_fc& abs(diff34) >C_fc))
}

## add scores.
if(length(filenames) == 1){
  final2$functional.scoreA <- final2$R1
  final2$functional.scoreB <- final2$R1
  final2$functional.scoreC <- final2$R1
}

if(length(filenames) ==2){

  temp <- final2 %>% select(R1, R2) 

  temp$R1[which(final2$removeA == TRUE)] <- NA
  temp$R2[which(final2$removeA == TRUE)] <- NA
  
  final2$functional.scoreA <- apply(temp, 1, mean, na.rm = TRUE)
  
  final2$functional.scoreB <- final2$meanR

  temp <- final2 %>% select(R1, R2)
  temp$R1[which(final2$removeC == TRUE)] <- NA
  temp$R2[which(final2$removeC == TRUE)] <- NA
  
  final2$functional.scoreC <- apply(temp, 1, mean, na.rm = TRUE)
  
}
if(length(filenames) == 3){

  temp <- final2 %>% select(R1, R2, R3) 

  temp$R1[which(final2$removeA1 == TRUE)] <- NA
  temp$R2[which(final2$removeA2 == TRUE)] <- NA
  temp$R3[which(final2$removeA3 == TRUE)] <- NA
  
  
  final2$functional.scoreA <- apply(temp, 1, mean, na.rm = TRUE)

  final2$functional.scoreB <- final2$meanR
  
  temp <- final2 %>% select(R1, R2, R3) 
  temp$R1[which(final2$removeC1 == TRUE)] <- NA
  temp$R2[which(final2$removeC2 == TRUE)] <- NA
  temp$R3[which(final2$removeC3 == TRUE)] <- NA
  
  final2$functional.scoreC <- apply(temp, 1, mean, na.rm = TRUE)
}
if(length(filenames) == 4){

  temp <- final2 %>% select(R1, R2, R3,R4) 

  temp$R1[which(final2$removeA1 == TRUE)] <- NA
  temp$R2[which(final2$removeA2 == TRUE)] <- NA
  temp$R3[which(final2$removeA3 == TRUE)] <- NA
  temp$R4[which(final2$removeA4 == TRUE)] <- NA
  
  
  final2$functional.scoreA <- apply(temp, 1, mean, na.rm = TRUE)

  final2$functional.scoreB <- final2$meanR
  
  temp <- final2 %>% select(R1, R2, R3,R4) 

  temp$R1[which(final2$removeC1 == TRUE)] <- NA
  temp$R2[which(final2$removeC2 == TRUE)] <- NA
  temp$R3[which(final2$removeC3 == TRUE)] <- NA
  temp$R4[which(final2$removeC4 == TRUE)] <- NA
  
  final2$functional.scoreC <- apply(temp, 1, mean, na.rm = TRUE)
}
final2$cv[which(!is.na(final2$meanR))] <- unlist(final2$sdR[which(!is.na(final2$meanR))]) / unlist(final2$meanR[which(!is.na(final2$meanR))])* 100 



```



### E. single variant Functional Scores

single variant Functional scores are average of individual experiment scores based on pre-classification (removeA1, removeA2, removeA3, removeA4).
By contrast, version B are average of individual experiment pre-functional scores without removing any data.

#### Distributions A

Note: ScoreA is mean of replicates after applying removeA1, 2, 3, 4

StopGain and Synonymous only, Including splice region variants 

```{r}

p1 <- final1 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>%  filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%
  ggplot( aes(x=functional.scoreA, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1) + scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14vsLib")


p2 <- final2 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%
  ggplot( aes(x=functional.scoreA, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1) +  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14_D5")
# 
# combined <- p1 + p2 & theme(legend.position = "bottom")
# (combined + plot_layout(guides = "collect"))
fig <- subplot(ggplotly(p1), ggplotly(p2)) %>% 
  layout(title = 'w. splice region variants: Left: D14_Lib, Right: D14_D5')
fig
```


StopGain and Synonymous SNV only, Excluding splice region variants 


```{r}

p1 <- final1 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>% arrange(functional.scoreA) %>%
  ggplot( aes(x=functional.scoreA, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1) + 
  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14vsLib single variant functional score A")


p2 <- final1 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR == FALSE) %>%filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>% arrange(functional.scoreA) %>%
  ggplot( aes(x=functional.scoreA, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1)  +
  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14_D5 single variant functional score A")

# combined <- p1 + p2 & theme(legend.position = "bottom")
# (combined + plot_layout(guides = "collect"))
fig <- subplot(ggplotly(p1), ggplotly(p2)) %>% 
  layout(title = 'wo. splice region variants: Left: D14_Lib, Right: D14_D5')
fig

```

D14vLib splice region SNV variants 

```{r}
options(digits = 3)
final1  %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR == TRUE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%select(EventType  , functional.scoreA)
```

D14vD5 splice region SNV variants 

```{r}
final2  %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(spliceR == TRUE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%select(EventType  , functional.scoreA)
```

#### Distributions B

Note: ScoreB is mean of replicates without removing any variant.

StopGain and Synonymous only, Including splice region variants 

```{r}
Rcol <- colnames(final1)[which(colnames(final1) %in% c("R1", "R2", "R3", "R4"))]
p1 <- final1 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%
  ggplot( aes(x=functional.scoreB, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1) + scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14vsLib single variant functional.scoreB")


p2 <- final2 %>% filter(EventType %in% c("StopGain", "Synonymous"))%>%filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%
  ggplot( aes(x=functional.scoreB, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1) +  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14_D5 single variant functional.scoreB")

# combined <- p1 + p2 & theme(legend.position = "bottom")
# (combined + plot_layout(guides = "collect"))

fig <- subplot(ggplotly(p1), ggplotly(p2)) %>% 
  layout(title = 'w. splice region variants: Left: D14_Lib, Right: D14_D5')
fig

```


StopGain and Synonymous SNV only, Excluding splice region variants 



```{r}

p1 <- final1 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%filter(spliceR == FALSE) %>% arrange(functional.scoreB) %>% 
  ggplot( aes(x=functional.scoreB, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1) + 
  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14vsLib functional scoreB")


p2 <- final2 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%filter(spliceR == FALSE) %>% arrange(functional.scoreB) %>% 
  ggplot( aes(x=functional.scoreB, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1)  +
  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14_D5 functional scoreB")

# combined <- p1 + p2 & theme(legend.position = "bottom")
# (combined + plot_layout(guides = "collect"))

fig <- subplot(ggplotly(p1), ggplotly(p2)) %>% 
  layout(title = 'wo. splice region variants: Left: D14_Lib, Right: D14_D5')
fig

```

D14vLib splice region variants 

```{r}
final1  %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%filter(spliceR == TRUE) %>% select(EventType  , functional.scoreB)
```

D14vD5 splice region variants 

```{r}
final2  %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%filter(spliceR == TRUE) %>% select(EventType  , functional.scoreB)
```



#### Distributions C
Note: ScoreC is mean of replicates after applying removeC1, 2, 3,4
StopGain and Synonymous only, Including splice region variants 

```{r}
Rcol <- colnames(final1)[which(colnames(final1) %in% c("R1", "R2", "R3", "R4"))]
p1 <- final1 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%
  ggplot( aes(x=functional.scoreC, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1) + scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14vsLib single variant functional.scoreC")


p2 <- final2 %>% filter(EventType %in% c("StopGain", "Synonymous"))%>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%
  ggplot( aes(x=functional.scoreC, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1) +  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14_D5 single variant functional.scoreC")

# combined <- p1 + p2 & theme(legend.position = "bottom")
# (combined + plot_layout(guides = "collect"))

fig <- subplot(ggplotly(p1), ggplotly(p2)) %>% 
  layout(title = 'w. splice region variants: Left: D14_Lib, Right: D14_D5')
fig

```


StopGain and Synonymous SNV only, Excluding splice region variants 


```{r}

p1 <- final1 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%filter(spliceR == FALSE) %>% arrange(functional.scoreB) %>% 
  ggplot( aes(x=functional.scoreB, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1) + 
  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14vsLib functional scoreB")


p2 <- final2 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%filter(spliceR == FALSE) %>% arrange(functional.scoreB) %>% 
  ggplot( aes(x=functional.scoreB, fill=EventType)) +
  geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity',binwidth = 0.1)  +
  scale_fill_manual(values=c( "red", "#69b3a2")) +  ggtitle ( "D14_D5 functional scoreB")

# combined <- p1 + p2 & theme(legend.position = "bottom")
# (combined + plot_layout(guides = "collect"))

fig <- subplot(ggplotly(p1), ggplotly(p2)) %>% 
  layout(title = 'wo. splice region variants: Left: D14_Lib, Right: D14_D5')
fig

```

D14vLib splice region variants 

```{r}
final1  %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%filter(spliceR == TRUE) %>% select(EventType  , functional.scoreC)
```

D14vD5 splice region variants 

```{r}
final2  %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%filter(spliceR == TRUE) %>% select(EventType  , functional.scoreC)
```


### F. Classification on single variant Functional Scores

#### ROC on D14_lib single variant functional score A

SNVs excluding splice region and variants on Chunling's list

```{r ROC0A, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

library(pROC)

data_long <- final1 %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%filter(EventType %in% c("Synonymous", "StopGain")) %>% arrange(functional.scoreA) %>% mutate(functional.score =functional.scoreA ) %>% filter(uPOS %in% wide2$uPOS)
##data_long$EventType <- droplevels(data_long$EventType)
par(mfrow = c(1, 1))
roc_score = roc(controls = data_long$functional.score[data_long$EventType == "Synonymous"], cases = data_long$functional.score[data_long$EventType == "StopGain"], direction = ">")  # AUC score
plot(roc_score, main="ROC curve")
text(0.2, 0.2, paste0("AUC: ", round(roc_score$auc, 3)))

out <- data.frame(functional.score = roc_score$"threshold", sens =  roc_score$"sensitivities", specs = roc_score$"specificities")

out$avg <- out$sens+out$specs

ind <- which(out$avg == max(out$avg))
if(length(ind) == 1){
  cutoff0A <- unique(out$functional.score[ind])
  sen0A <- out$sens[ind]
  spec0A <- out$specs[ind]

} else{
  cutoff0A <- unique(out$functional.score[ind[1]])
  sen0A <- out$sens[ind[1]]
  spec0A <- out$specs[ind[1]]
}

auc0A <- roc_score$auc

data_long$pred.class <- ifelse(data_long$functional.score < cutoff0A, "StopGain", "Synonymous")

tb0A <- table(truth = data_long$EventType, pred = data_long$pred.class)


```

`r cutoff0A` was used for the performance. sensitivity is `r sen0A`, specificity is `r spec0A`. 
Note, The number of the variants are the actural variants in the  table. 

#### ROC on D14_lib single variant functional score B

SNVs excluding splice region and exclude variants on chunling's list

```{r ROC0B, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}


data_long <- final1 %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>% filter(EventType %in% c("Synonymous", "StopGain")) %>%  mutate(functional.score = functional.scoreB) %>% arrange(functional.score) %>% filter(uPOS %in% wide2$uPOS)

##data_long$EventType <- droplevels(data_long$EventType)
par(mfrow = c(1, 1))
roc_score = roc(controls = data_long$functional.score[data_long$EventType == "Synonymous"], cases = data_long$functional.score[data_long$EventType == "StopGain"], direction = ">")  # AUC score
plot(roc_score, main="ROC curve")
text(0.2, 0.2, paste0("AUC: ", round(roc_score$auc, 3)))

out <- data.frame(functional.score = roc_score$"threshold", sens =  roc_score$"sensitivities", specs = roc_score$"specificities")

out$avg <- out$sens+out$specs
ind <- which(out$avg == max(out$avg))
if(length(ind) == 1){
  cutoff0B <- unique(out$functional.score[ind])
  sen0B <- out$sens[ind]
  spec0B <- out$specs[ind]

} else{
  cutoff0B <- unique(out$functional.score[ind[1]])
  sen0B <- out$sens[ind[1]]
  spec0B <- out$specs[ind[1]]
}

auc0B <- roc_score$auc

data_long$pred.class <- ifelse(data_long$functional.score < cutoff0B, "StopGain", "Synonymous")

tb0B <- table(truth = data_long$EventType, pred = data_long$pred.class)
tb0B

```

`r cutoff0B` was used for the performance. sensitivity is `r sen0B`, specificity is `r spec0B`. 
Note, The number of the variants are the actual variants in the  table.


#### ROC on D14_lib single variant functional score C

SNVs excluding splice region and exlude variants on chunling's list

```{r ROC0C, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

data_long <- final1 %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>% filter(EventType %in% c("Synonymous", "StopGain")) %>% arrange(functional.scoreC) %>% mutate(functional.score = functional.scoreC) %>% filter(uPOS %in% wide2$uPOS)

##data_long$EventType <- droplevels(data_long$EventType)
par(mfrow = c(1, 1))
roc_score = roc(controls = data_long$functional.score[data_long$EventType == "Synonymous"], cases = data_long$functional.score[data_long$EventType == "StopGain"], direction = ">")  # AUC score
plot(roc_score, main="ROC curve")
text(0.2, 0.2, paste0("AUC: ", round(roc_score$auc, 3)))

out <- data.frame(functional.score = roc_score$"threshold", sens =  roc_score$"sensitivities", specs = roc_score$"specificities")

out$avg <- out$sens+out$specs
ind <- which(out$avg == max(out$avg))
if(length(ind) == 1){
  cutoff0C <- unique(out$functional.score[ind])
  sen0C <- out$sens[ind]
  spec0C <- out$specs[ind]

} else{
  cutoff0C <- unique(out$functional.score[ind[1]])
  sen0C <- out$sens[ind[1]]
  spec0C <- out$specs[ind[1]]
}

auc0C <- roc_score$auc

data_long$pred.class <- ifelse(data_long$functional.score < cutoff0C, "StopGain", "Synonymous")

tb0C <- table(truth = data_long$EventType, pred = data_long$pred.class)
tb0C

```

`r cutoff0C` was used for the performance. sensitivity is `r sen0C`, specificity is `r spec0C`. 
Note, The number of the variants are the actual variants in the  table.

#### ROC on D14_D5 ratios single variant functional score A

SNVs excluding splice region and variants on chunling's list

```{r ROC0A.1, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}


data_long <- final2 %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>% filter(EventType %in% c("Synonymous", "StopGain")) %>% mutate(functional.score = functional.scoreA) %>% arrange(functional.score) %>% filter(uPOS %in% wide2$uPOS)
##data_long$EventType <- droplevels(data_long$EventType)
par(mfrow = c(1, 1))
roc_score = roc(controls = data_long$functional.score[data_long$EventType == "Synonymous"], cases = data_long$functional.score[data_long$EventType == "StopGain"], direction = ">")  # AUC score
plot(roc_score, main="ROC curve")
text(0.2, 0.2, paste0("AUC: ", round(roc_score$auc, 3)))

out <- data.frame(functional.score = roc_score$"threshold", sens =  roc_score$"sensitivities", specs = roc_score$"specificities")

out$avg <- out$sens+out$specs

cutoff1A<- unique(out$functional.score[which(out$avg == max(out$avg))])
sen1A <- out$sens[which(out$avg == max(out$avg))]
spec1A <- out$specs[which(out$avg == max(out$avg))]
auc1A <- roc_score$auc

data_long$pred.class <- ifelse(data_long$functional.score < cutoff1A, "StopGain", "Synonymous")

tb1A <- table(truth = data_long$EventType, pred = data_long$pred.class)
tb1A
```

`r cutoff1A` was used for the performance. sensitivity is `r sen1A`, specificity is `r spec1A`. 
The number of the variants is the number of actual variants in the table. 



#### ROC on D14_D5 ratios single variant functional score B

SNVs excluding splice region and variants on chunling's list

```{r ROC0B.1, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}


library(pROC)

data_long <- final2 %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>% filter(EventType %in% c("Synonymous", "StopGain")) %>%  mutate(functional.score = functional.scoreB) %>% arrange(functional.score) %>% filter(uPOS %in% wide2$uPOS)
##data_long$EventType <- droplevels(data_long$EventType)
par(mfrow = c(1, 1))
roc_score = roc(controls = data_long$functional.score[data_long$EventType == "Synonymous"], cases = data_long$functional.score[data_long$EventType == "StopGain"], direction = ">")  # AUC score
plot(roc_score, main="ROC curve")
text(0.2, 0.2, paste0("AUC: ", round(roc_score$auc, 3)))

out <- data.frame(functional.score = roc_score$"threshold", sens =  roc_score$"sensitivities", specs = roc_score$"specificities")

out$avg <- out$sens+out$specs

cutoff1B<- unique(out$functional.score[which(out$avg == max(out$avg))])
sen1B <- out$sens[which(out$avg == max(out$avg))]
spec1B <- out$specs[which(out$avg == max(out$avg))]
auc1B <- roc_score$auc

data_long$pred.class <- ifelse(data_long$functional.score < cutoff1B, "StopGain", "Synonymous")

tb1B <- table(truth = data_long$EventType, pred = data_long$pred.class)
tb1B

```

`r cutoff1B` was used for the performance. sensitivity is `r sen1B`, specificity is `r spec1B`. 
The number of the variants is the number of actual variants in the table. 


#### ROC on D14_D5 ratios single variant functional score C

SNVs excluding splice region and variants on chunling's list

```{r ROC0C.1, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

data_long <- final2 %>% filter(spliceR == FALSE) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>% filter(EventType %in% c("Synonymous", "StopGain")) %>%  mutate(functional.score = functional.scoreC) %>% arrange(functional.score) %>% filter(uPOS %in% wide2$uPOS)
##data_long$EventType <- droplevels(data_long$EventType)
par(mfrow = c(1, 1))
roc_score = roc(controls = data_long$functional.score[data_long$EventType == "Synonymous"], cases = data_long$functional.score[data_long$EventType == "StopGain"], direction = ">")  # AUC score
plot(roc_score, main="ROC curve")
text(0.2, 0.2, paste0("AUC: ", round(roc_score$auc, 3)))

out <- data.frame(functional.score = roc_score$"threshold", sens =  roc_score$"sensitivities", specs = roc_score$"specificities")

out$avg <- out$sens+out$specs

cutoff1C<- unique(out$functional.score[which(out$avg == max(out$avg))])
sen1C <- out$sens[which(out$avg == max(out$avg))]
spec1C <- out$specs[which(out$avg == max(out$avg))]
auc1C <- roc_score$auc

data_long$pred.class <- ifelse(data_long$functional.score < cutoff1C, "StopGain", "Synonymous")

tb1C <- table(truth = data_long$EventType, pred = data_long$pred.class)
tb1C

```

`r cutoff1C` was used for the performance. sensitivity is `r sen1C`, specificity is `r spec1C`. 
The number of the variants is the number of actual variants in the table. 



#### Mixture model on D14_lib single variant functional scoreA


```{r MixModel1A, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

##normalmixEM function of the mixtools package in R
library(mixtools)
f1A <- f2A <- f3A <- f4A <- NA

temp <- final1 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%
  arrange(functional.scoreA) %>% rename(functional.score = functional.scoreA)%>% filter(!is.na(functional.score)) %>% filter(spliceR == FALSE) %>% filter(uPOS %in% wide2$uPOS)

m1 <- mean(temp$functional.score[temp$EventType != "StopGain"], na.rm = TRUE)
tmp <- temp$functional.score[temp$EventType != "StopGain"]
tmp2 <- tmp ## tmp[which(tmp > median(tmp)-sd(tmp) & tmp < median(tmp)+sd(tmp))]

v1 <- sd(tmp2, na.rm = TRUE)

m2 <- mean(temp$functional.score[temp$EventType == "StopGain"])
v2 <- sd(temp$functional.score[temp$EventType == "StopGain"])
l1 <- sum(temp$EventType == "StopGain")/sum(temp$EventType != "StopGain")
another.error.occured <- FALSE
tryCatch( {
  x <- normalmixEM(temp$functional.score, 
                   lambda = l1, k = 2,
                   mean.constr = c(m2, m1),
                  sd.constr = c(v2, v1),
                 mu = c(m2, m1), 
                 sigma = c(v2, v1))

par(mfrow = c(1, 1))
plot(temp$functional.score, x$posterior[,1], col = ifelse(temp$EventType == "StopGain", 2, 1))
f1A <- temp$functional.score[x$posterior[,1]>0.01][order(temp$functional.score[x$posterior[,1]>0.01], decreasing = T)][1]
f2A <- temp$functional.score[x$posterior[,1]>0.99][order(temp$functional.score[x$posterior[,1]>0.99], decreasing = T)][1]

f3A <- temp$functional.score[x$posterior[,1]>0.02][order(temp$functional.score[x$posterior[,1]>0.02], decreasing = T)][1]
f4A <- temp$functional.score[x$posterior[,1]>0.98][order(temp$functional.score[x$posterior[,1]>0.98], decreasing = T)][1]


abline(v = f1A)
abline(v = f2A)
abline(h = 0.01)
abline(h = 0.99)

abline(v = f3A, lty = 2)
abline(v = f4A, lty = 2)
abline(h = 0.02, lty = 2)
abline(h = 0.98, lty = 2)

par(mfrow = c(1, 1))
plot(x, which=2, xlab2= "", yaxt= "n", xlim = c(range(x$x)[1]-2, range(x$x)[2]+2), main2 ="Two components with 1-99% prob(nonfunctional)", lwd2=0.8, border = "azure3")
lines(density(x$x), lty=2, lwd=0.8)

abline(v = f1A)
abline(v = f2A)
abline(v = f3A, lty = 2)
abline(v = f4A, lty = 2)

}
          , error = function(e) {another.error.occured <<- TRUE})
print(another.error.occured)

# pdf(paste0(params$Output_dir, params$Gene, "_", Exn, "_MM_A.pdf"), width = 4, height = 4)
# plot(x, which=2, xlab2= "", yaxt= "n", xlim = c(range(x$x)[1]-2, range(x$x)[2]+2), main2 ="Two components Mixture Model", lwd2=0.8, border = "azure3")
# 
# abline(v = f1A)
# abline(v = f2A)
# abline(v = f3A, lty = 2)
# abline(v = f4A, lty = 2)
# dev.off()
# 
# tiff(paste0(params$Output_dir, params$Gene, "_", Exn, "_MM_A.tiff"), width = 4, height = 4, units = "in", res = 1200)
# plot(x, which=2, xlab2= "", yaxt= "n", xlim = c(range(x$x)[1]-2, range(x$x)[2]+2), main2 ="Two components Mixture Model", lwd2=0.8, border = "azure3", cex.lab=0.8, cex.axis=0.8, cex.main=0.8)
# 
# abline(v = f1A)
# abline(v = f2A)
# abline(v = f3A, lty = 2)
# abline(v = f4A, lty = 2)
# dev.off()


```

`r paste0("There is",  ifelse(another.error.occured, " AN ", " NO "), "Error Ocurred")`.

`r f2A` was the functional score corresponding to the probability of non functional at 99% and `r f1A` was the functional score corresponding to the probability of non functional at 1%.

`r f4A` was the functional score corresponding to the probability of non functional at 98% and `r f3A` was the functional score corresponding to the probability of non functional at 2%.

If we use these two thresholds to classfiy the variants of synonymous, stopgain and missense. 

First, 1-99% cutoff
Second, 2-98% cutoff

Note: Splice regions are included in Missense, not stopgain nor synonymous

```{r MixModel1A.2, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}


if(!another.error.occured){
  final1$ctypeA <- ifelse(final1$functional.scoreA>f1A, "neutural", 
                       ifelse(final1$functional.scoreA>f2A, "intermediate", "pathogenic"))
  final1$ctypeA2 <- ifelse(final1$functional.scoreA>f3A, "neutural", 
                       ifelse(final1$functional.scoreA>f4A, "intermediate", "pathogenic"))
  
  temp.check <- final1 %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>% filter((EventType %in% c("StopGain", "Synonymous") & spliceR == FALSE) | EventType == "Missense")
    
  table( truth = temp.check$EventType, pred = temp.check$ctypeA)
  table(truth =  temp.check$EventType, pred = temp.check$ctypeA2)
}

```

#### Mixture model on D14_lib single variant functional scoreB


```{r MixModel1B, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

##normalmixEM function of the mixtools package in R
library(mixtools)
f1B <- f2B <- f3B <- f4B <- NA

temp <- final1 %>% filter(EventType %in% c("StopGain", "Synonymous"))  %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%
  arrange(functional.scoreB) %>% rename(functional.score = functional.scoreB)%>% filter(!is.na(functional.score)) %>% filter(spliceR == FALSE) %>% filter(uPOS %in% wide2$uPOS)

m1 <- mean(temp$functional.score[temp$EventType != "StopGain"], na.rm = TRUE)
tmp <- temp$functional.score[temp$EventType != "StopGain"]
tmp2 <- tmp ## tmp[which(tmp > median(tmp)-sd(tmp) & tmp < median(tmp)+sd(tmp))]

v1 <- sd(tmp2, na.rm = TRUE)

m2 <- mean(temp$functional.score[temp$EventType == "StopGain"])
v2 <- sd(temp$functional.score[temp$EventType == "StopGain"])
l1 <- sum(temp$EventType == "StopGain")/sum(temp$EventType != "StopGain")
another.error.occured <- FALSE
tryCatch( {
  x <- normalmixEM(temp$functional.score, 
                   lambda = l1, k = 2,
                   mean.constr = c(m2, m1),
                  sd.constr = c(v2, v1),
                 mu = c(m2, m1), 
                 sigma = c(v2, v1))

par(mfrow = c(1, 1))
plot(temp$functional.score, x$posterior[,1], col = ifelse(temp$EventType == "StopGain", 2, 1))
f1B <- temp$functional.score[x$posterior[,1]>0.01][order(temp$functional.score[x$posterior[,1]>0.01], decreasing = T)][1]
f2B <- temp$functional.score[x$posterior[,1]>0.99][order(temp$functional.score[x$posterior[,1]>0.99], decreasing = T)][1]

f3B <- temp$functional.score[x$posterior[,1]>0.02][order(temp$functional.score[x$posterior[,1]>0.02], decreasing = T)][1]
f4B <- temp$functional.score[x$posterior[,1]>0.98][order(temp$functional.score[x$posterior[,1]>0.98], decreasing = T)][1]


abline(v = f1B)
abline(v = f2B)
abline(h = 0.01)
abline(h = 0.99)

abline(v = f3B, lty = 2)
abline(v = f4B, lty = 2)
abline(h = 0.02, lty = 2)
abline(h = 0.98, lty = 2)

par(mfrow = c(1, 1))
plot(x, which=2, xlab2= "", yaxt= "n", xlim = c(range(x$x)[1]-2, range(x$x)[2]+2), main2 ="Two components with 1-99% prob(nonfunctional)", lwd2=0.8, border = "azure3")
lines(density(x$x), lty=2, lwd=0.8)

abline(v = f1B)
abline(v = f2B)
abline(v = f3B, lty = 2)
abline(v = f4B, lty = 2)

}
          , error = function(e) {another.error.occured <<- TRUE})
print(another.error.occured)


```

`r paste0("There is",  ifelse(another.error.occured, " AN ", " NO "), "Error Ocurred")`.

`r f2B` was the functional score corresponding to the probability of non functional at 99% and `r f1B` was the functional score corresponding to the probability of non functional at 1%.

`r f4B` was the functional score corresponding to the probability of non functional at 98% and `r f3B` was the functional score corresponding to the probability of non functional at 2%.

If we use these two thresholds to classfiy the variants of synonymous, stopgain and missense. 

First, 1-99% cutoff
Second, 2-98% cutoff

Note: Splice regions are included in Missense, not stopgain nor synonymous

```{r MixModel1B.2, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}


if(!another.error.occured){
  final1$ctypeB <- ifelse(final1$functional.scoreB>f1B, "neutural", 
                       ifelse(final1$functional.scoreB>f2B, "intermediate", "pathogenic"))
  final1$ctypeB2 <- ifelse(final1$functional.scoreB>f3B, "neutural", 
                       ifelse(final1$functional.scoreB>f4B, "intermediate", "pathogenic"))
  
  temp.check <- final1 %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%filter((EventType %in% c("StopGain", "Synonymous") & spliceR == FALSE) | EventType == "Missense")
    
  table(temp.check$ctypeB, temp.check$EventType)
  table(temp.check$ctypeB2, temp.check$EventType)
}

```

#### Mixture model on D14_lib single variant functional scoreC


```{r MixModel1C, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

##normalmixEM function of the mixtools package in R
library(mixtools)
f1C <- f2C <- f3C <- f4C <- NA

temp <- final1 %>% filter(EventType %in% c("StopGain", "Synonymous"))  %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%
  arrange(functional.scoreC) %>% rename(functional.score = functional.scoreC) %>% filter(!is.na(functional.score)) %>% filter(spliceR == FALSE) %>% filter(uPOS %in% wide2$uPOS)

m1 <- mean(temp$functional.score[temp$EventType != "StopGain"], na.rm = TRUE)
tmp <- temp$functional.score[temp$EventType != "StopGain"]
tmp2 <- tmp ## tmp[which(tmp > median(tmp)-sd(tmp) & tmp < median(tmp)+sd(tmp))]

v1 <- sd(tmp2, na.rm = TRUE)

m2 <- mean(temp$functional.score[temp$EventType == "StopGain"])
v2 <- sd(temp$functional.score[temp$EventType == "StopGain"])
l1 <- sum(temp$EventType == "StopGain")/sum(temp$EventType != "StopGain")
another.error.occured <- FALSE
tryCatch( {
  x <- normalmixEM(temp$functional.score, 
                   lambda = l1, k = 2,
                   mean.constr = c(m2, m1),
                  sd.constr = c(v2, v1),
                 mu = c(m2, m1), 
                 sigma = c(v2, v1))

par(mfrow = c(1, 1))
plot(temp$functional.score, x$posterior[,1], col = ifelse(temp$EventType == "StopGain", 2, 1))
f1C <- temp$functional.score[x$posterior[,1]>0.01][order(temp$functional.score[x$posterior[,1]>0.01], decreasing = T)][1]
f2C <- temp$functional.score[x$posterior[,1]>0.99][order(temp$functional.score[x$posterior[,1]>0.99], decreasing = T)][1]

f3C <- temp$functional.score[x$posterior[,1]>0.02][order(temp$functional.score[x$posterior[,1]>0.02], decreasing = T)][1]
f4C <- temp$functional.score[x$posterior[,1]>0.98][order(temp$functional.score[x$posterior[,1]>0.98], decreasing = T)][1]


abline(v = f1C)
abline(v = f2C)
abline(h = 0.01)
abline(h = 0.99)

abline(v = f3C, lty = 2)
abline(v = f4C, lty = 2)
abline(h = 0.02, lty = 2)
abline(h = 0.98, lty = 2)

par(mfrow = c(1, 1))
plot(x, which=2, xlab2= "", yaxt= "n", xlim = c(range(x$x)[1]-2, range(x$x)[2]+2), main2 ="Two components with 1-99% prob(nonfunctional)", lwd2=0.8, border = "azure3")
lines(density(x$x), lty=2, lwd=0.8)

abline(v = f1C)
abline(v = f2C)
abline(v = f3C, lty = 2)
abline(v = f4C, lty = 2)

}
          , error = function(e) {another.error.occured <<- TRUE})
print(another.error.occured)


```

`r paste0("There is",  ifelse(another.error.occured, " AN ", " NO "), "Error Ocurred")`.

`r f2C` was the functional score corresponding to the probability of non functional at 99% and `r f1C` was the functional score corresponding to the probability of non functional at 1%.

`r f4C` was the functional score corresponding to the probability of non functional at 98% and `r f3C` was the functional score corresponding to the probability of non functional at 2%.

If we use these two thresholds to classfiy the variants of synonymous, stopgain and missense. 

First, 1-99% cutoff
Second, 2-98% cutoff

Note: Splice regions are included in Missense, not stopgain nor synonymous

```{r MixModel1C.2, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}


if(!another.error.occured){
  final1$ctypeC <- ifelse(final1$functional.scoreC>f1C, "neutural", 
                       ifelse(final1$functional.scoreC>f2C, "intermediate", "pathogenic"))
  final1$ctypeC2 <- ifelse(final1$functional.scoreC>f3C, "neutural", 
                       ifelse(final1$functional.scoreC>f4C, "intermediate", "pathogenic"))
  
  temp.check <- final1 %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>% filter((EventType %in% c("StopGain", "Synonymous") & spliceR == FALSE) | EventType == "Missense")
    
  table(temp.check$ctypeC, temp.check$EventType)
  table(temp.check$ctypeC2, temp.check$EventType)
}

write.csv(final1, paste0(params$Output_dir, Gene, "_", Exn, "_", vset, "_", loessSubset, "_", EventCount_Filter, "_cutoff", round(cutoff0, 2), "_se", round(sen0, 2),"_sp", round(spec0, 2), "_D14vLib.csv"), row.names = FALSE)
```

#### Mixture model on D14_D5 single variant functional score A

```{r MixModel1A_D14_D5, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

##normalmixEM function of the mixtools package in R
library(mixtools)
f1A.2 <- f2A.2 <- f3A.2 <- f4A.2 <- NA

temp <- final2 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%
  arrange(functional.scoreA) %>% rename(functional.score = functional.scoreA) %>% filter(!is.na(functional.score))%>% filter(spliceR == FALSE) %>% filter(uPOS %in% wide2$uPOS)

m1 <- mean(temp$functional.score[temp$EventType != "StopGain"], na.rm = TRUE)
tmp <- temp$functional.score[temp$EventType != "StopGain"]
tmp2 <- tmp ## tmp[which(tmp > median(tmp)-sd(tmp) & tmp < median(tmp)+sd(tmp))]

v1 <- sd(tmp2, na.rm = TRUE)

m2 <- mean(temp$functional.score[temp$EventType == "StopGain"])
v2 <- sd(temp$functional.score[temp$EventType == "StopGain"])
l1 <- sum(temp$EventType == "StopGain")/sum(temp$EventType != "StopGain")
another.error.occured <- FALSE
tryCatch( {
  x <- normalmixEM(temp$functional.score, 
                   lambda = l1, k = 2,
                   mean.constr = c(m2, m1),
                  sd.constr = c(v2, v1),
                 mu = c(m2, m1), 
                 sigma = c(v2, v1))

par(mfrow = c(1, 1))
plot(temp$functional.score, x$posterior[,1], col = ifelse(temp$EventType == "StopGain", 2, 1))
f1A.2 <- temp$functional.score[x$posterior[,1]>0.01][order(temp$functional.score[x$posterior[,1]>0.01], decreasing = T)][1]
f2A.2 <- temp$functional.score[x$posterior[,1]>0.99][order(temp$functional.score[x$posterior[,1]>0.99], decreasing = T)][1]

f3A.2 <- temp$functional.score[x$posterior[,1]>0.02][order(temp$functional.score[x$posterior[,1]>0.02], decreasing = T)][1]
f4A.2 <- temp$functional.score[x$posterior[,1]>0.98][order(temp$functional.score[x$posterior[,1]>0.98], decreasing = T)][1]


abline(v = f1A.2)
abline(v = f2A.2)
abline(h = 0.01)
abline(h = 0.99)

abline(v = f3A.2, lty = 2)
abline(v = f4A.2, lty = 2)
abline(h = 0.02, lty = 2)
abline(h = 0.98, lty = 2)

par(mfrow = c(1, 1))
plot(x, which=2, xlab2= "", yaxt= "n", xlim = c(range(x$x)[1]-2, range(x$x)[2]+2), main2 ="Two components with 1-99% prob(nonfunctional)", lwd2=0.8, border = "azure3")
lines(density(x$x), lty=2, lwd=0.8)

abline(v = f1A.2)
abline(v = f2A.2)
abline(v = f3A.2, lty = 2)
abline(v = f4A.2, lty = 2)

}
          , error = function(e) {another.error.occured <<- TRUE})
print(another.error.occured)


```

`r paste0("There is",  ifelse(another.error.occured, " AN ", " NO "), "Error Ocurred")`.

`r f2A.2` was the functional score corresponding to the probability of non functional at 99% and `r f1A.2` was the functional score corresponding to the probability of non functional at 1%.

`r f4A.2` was the functional score corresponding to the probability of non functional at 98% and `r f3A.2` was the functional score corresponding to the probability of non functional at 2%.

If we use these two thresholds to classfiy the variants of synonymous, stopgain and missense. 

First, 1-99% cutoff
Second, 2-98% cutoff

Note: Splice regions are included in Missense, not stopgain nor synonymous

```{r MixModel1A.2_D14_D5, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}


if(!another.error.occured){
  final2$ctypeA <- ifelse(final2$functional.scoreA>f1A.2, "neutural", 
                       ifelse(final2$functional.scoreA>f2A.2, "intermediate", "pathogenic"))
  final2$ctypeA2 <- ifelse(final2$functional.scoreA>f3A.2, "neutural", 
                       ifelse(final1$functional.scoreA>f4A.2, "intermediate", "pathogenic"))
  
  temp.check <- final1 %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%filter((EventType %in% c("StopGain", "Synonymous") & spliceR == FALSE) | EventType == "Missense")
    
  table(truth =  temp.check$EventType, pred = temp.check$ctypeA)
  table(truth = temp.check$EventType, pred = temp.check$ctypeA2)
}

```

#### Mixture model on D14_D5 single variant functional score B

```{r MixModel1B.2_D14_D5, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

##normalmixEM function of the mixtools package in R
library(mixtools)
f1B.2 <- f2B.2 <- f3B.2 <- f4B.2 <- NA

temp <- final2 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%
  arrange(functional.scoreB) %>% rename(functional.score = functional.scoreB) %>% filter(!is.na(functional.score)) %>% filter(spliceR == FALSE) %>% filter(uPOS %in% wide2$uPOS)
 

m1 <- mean(temp$functional.score[temp$EventType != "StopGain"], na.rm = TRUE)
tmp <- temp$functional.score[temp$EventType != "StopGain"]
tmp2 <- tmp ## tmp[which(tmp > median(tmp)-sd(tmp) & tmp < median(tmp)+sd(tmp))]

v1 <- sd(tmp2, na.rm = TRUE)
m2 <- mean(temp$functional.score[temp$EventType == "StopGain"])
v2 <- sd(temp$functional.score[temp$EventType == "StopGain"])
l1 <- sum(temp$EventType == "StopGain")/sum(temp$EventType != "StopGain")
another.error.occured <- FALSE
tryCatch( {
  x <- normalmixEM(temp$functional.score, 
                   lambda = l1, k = 2,
                   mean.constr = c(m2, m1),
                  sd.constr = c(v2, v1),
                 mu = c(m2, m1), 
                 sigma = c(v2, v1))

par(mfrow = c(1, 1))
plot(temp$functional.score, x$posterior[,1], col = ifelse(temp$EventType == "StopGain", 2, 1))
f1B.2 <- temp$functional.score[x$posterior[,1]>0.01][order(temp$functional.score[x$posterior[,1]>0.01], decreasing = T)][1]
f2B.2 <- temp$functional.score[x$posterior[,1]>0.99][order(temp$functional.score[x$posterior[,1]>0.99], decreasing = T)][1]

f3B.2 <- temp$functional.score[x$posterior[,1]>0.02][order(temp$functional.score[x$posterior[,1]>0.02], decreasing = T)][1]
f4B.2 <- temp$functional.score[x$posterior[,1]>0.98][order(temp$functional.score[x$posterior[,1]>0.98], decreasing = T)][1]


abline(v = f1B.2)
abline(v = f2B.2)
abline(h = 0.01)
abline(h = 0.99)

abline(v = f3B.2, lty = 2)
abline(v = f4B.2, lty = 2)
abline(h = 0.02, lty = 2)
abline(h = 0.98, lty = 2)

par(mfrow = c(1, 1))
plot(x, which=2, xlab2= "", yaxt= "n", xlim = c(range(x$x)[1]-2, range(x$x)[2]+2), main2 ="Two components with 1-99% prob(nonfunctional)", lwd2=0.8, border = "azure3")
lines(density(x$x), lty=2, lwd=0.8)

abline(v = f1B.2)
abline(v = f2B.2)
abline(v = f3B.2, lty = 2)
abline(v = f4B.2, lty = 2)

}
          , error = function(e) {another.error.occured <<- TRUE})
print(another.error.occured)


```

`r paste0("There is",  ifelse(another.error.occured, " AN ", " NO "), "Error Ocurred")`.

`r f2B.2` was the functional score corresponding to the probability of non functional at 99% and `r f1B.2` was the functional score corresponding to the probability of non functional at 1%.

`r f4B.2` was the functional score corresponding to the probability of non functional at 98% and `r f3B.2` was the functional score corresponding to the probability of non functional at 2%.

If we use these two thresholds to classfiy the variants of synonymous, stopgain and missense. 

First, 1-99% cutoff
Second, 2-98% cutoff

Note: Splice regions are included in Missense, not stopgain nor synonymous

```{r MixModel1B.2.2_D14_D5, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

if(!another.error.occured){
  final2$ctypeB <- ifelse(final2$functional.scoreB>f1B.2, "neutural", 
                       ifelse(final2$functional.scoreB>f2B.2, "intermediate", "pathogenic"))
  final2$ctypeB2 <- ifelse(final2$functional.scoreB>f3B.2, "neutural", 
                       ifelse(final1$functional.scoreB>f4B.2, "intermediate", "pathogenic"))
  
  temp.check <- final2 %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>% filter((EventType %in% c("StopGain", "Synonymous") & spliceR == FALSE) | EventType == "Missense")
    
  table(truth = temp.check$EventType, pred = temp.check$ctypeB)
  table(truth = temp.check$EventType, pred = temp.check$ctypeB2)
}

```


#### Mixture model on D14_D5 single variant functional score C

```{r MixModel1C.2_D14_D5, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

##normalmixEM function of the mixtools package in R
library(mixtools)
f1C.2 <- f2C.2 <- f3C.2 <- f4C.2 <- NA

temp <- final2 %>% filter(EventType %in% c("StopGain", "Synonymous")) %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%
  arrange(functional.scoreC) %>% rename(functional.score = functional.scoreC) %>% filter(!is.na(functional.score)) %>% filter(spliceR == FALSE) %>% filter(uPOS %in% wide2$uPOS)


m1 <- mean(temp$functional.score[temp$EventType != "StopGain"], na.rm = TRUE)
tmp <- temp$functional.score[temp$EventType != "StopGain"]
tmp2 <- tmp ## tmp[which(tmp > median(tmp)-sd(tmp) & tmp < median(tmp)+sd(tmp))]

v1 <- sd(tmp2, na.rm = TRUE)

m2 <- mean(temp$functional.score[temp$EventType == "StopGain"])
v2 <- sd(temp$functional.score[temp$EventType == "StopGain"])
l1 <- sum(temp$EventType == "StopGain")/sum(temp$EventType != "StopGain")
another.error.occured <- FALSE
tryCatch( {
  x <- normalmixEM(temp$functional.score, 
                   lambda = l1, k = 2,
                   mean.constr = c(m2, m1),
                  sd.constr = c(v2, v1),
                 mu = c(m2, m1), 
                 sigma = c(v2, v1))

par(mfrow = c(1, 1))
plot(temp$functional.score, x$posterior[,1], col = ifelse(temp$EventType == "StopGain", 2, 1))
f1C.2 <- temp$functional.score[x$posterior[,1]>0.01][order(temp$functional.score[x$posterior[,1]>0.01], decreasing = T)][1]
f2C.2 <- temp$functional.score[x$posterior[,1]>0.99][order(temp$functional.score[x$posterior[,1]>0.99], decreasing = T)][1]

f3C.2 <- temp$functional.score[x$posterior[,1]>0.02][order(temp$functional.score[x$posterior[,1]>0.02], decreasing = T)][1]
f4C.2 <- temp$functional.score[x$posterior[,1]>0.98][order(temp$functional.score[x$posterior[,1]>0.98], decreasing = T)][1]


abline(v = f1C.2)
abline(v = f2C.2)
abline(h = 0.01)
abline(h = 0.99)

abline(v = f3C.2, lty = 2)
abline(v = f4C.2, lty = 2)
abline(h = 0.02, lty = 2)
abline(h = 0.98, lty = 2)

par(mfrow = c(1, 1))
plot(x, which=2, xlab2= "", yaxt= "n", xlim = c(range(x$x)[1]-2, range(x$x)[2]+2), main2 ="Two components with 1-99% prob(nonfunctional)", lwd2=0.8, border = "azure3")
lines(density(x$x), lty=2, lwd=0.8)

abline(v = f1C.2)
abline(v = f2C.2)
abline(v = f3C.2, lty = 2)
abline(v = f4C.2, lty = 2)

}
          , error = function(e) {another.error.occured <<- TRUE})
print(another.error.occured)


```

`r paste0("There is",  ifelse(another.error.occured, " AN ", " NO "), "Error Ocurred")`.

`r f2C.2` was the functional score corresponding to the probability of non functional at 99% and `r f1C.2` was the functional score corresponding to the probability of non functional at 1%.

`r f4C.2` was the functional score corresponding to the probability of non functional at 98% and `r f3C.2` was the functional score corresponding to the probability of non functional at 2%.

If we use these two thresholds to classfiy the variants of synonymous, stopgain and missense. 

First, 1-99% cutoff
Second, 2-98% cutoff

Note: Splice regions are included in Missense, not stopgain nor synonymous

```{r MixModel1C.2.2_D14_D5, include=TRUE, fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

if(!another.error.occured){
  final2$ctypeC <- ifelse(final2$functional.scoreC>f1C.2, "neutural", 
                       ifelse(final2$functional.scoreC>f2C.2, "intermediate", "pathogenic"))
  final2$ctypeC2 <- ifelse(final2$functional.scoreC>f3C.2, "neutural", 
                       ifelse(final1$functional.scoreC>f4C.2, "intermediate", "pathogenic"))
  
  temp.check <- final2 %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>%filter((EventType %in% c("StopGain", "Synonymous") & spliceR == FALSE) | EventType == "Missense")
    
  table(truth = temp.check$EventType, pred = temp.check$ctypeC)
  table(truth = temp.check$EventType, pred = temp.check$ctypeC2)
}

write.csv(final2, paste0(params$Output_dir, Gene, "_", Exn, "_", vset, "_", loessSubset, "_", EventCount_Filter, "_cutoff", round(cutoff1, 2), "_se", round(sen1, 2),"_sp", round(spec1, 2), "_D14vD5.csv"), row.names = FALSE)
```

```{r}
## output the metrics in one file
## 4 scores
metrics <- data.frame(Exon = rep(Exn,4),
                      Rrep	= rep(length(filenames),4), 
                      N_score=c(length(filenames)*nrow(final1), sum(!is.na(final1$functional.scoreA)), sum(!is.na(final1$functional.scoreB)), sum(!is.na(final1$functional.scoreC))),
                      EvenFilter = rep(EventCount_Filter, 4),
                      vSet = rep(vset, 4),
                      Method = rep(loessSubset, 4),
                      score = c("pre-FS", "single variant FS A", "single variant FS B", "single variant FS C"),
                      AUC_D14_lib= c(as.numeric(auc0),as.numeric(auc0A),as.numeric(auc0B),as.numeric(auc0C)),
                      sens_D14_lib=c(sen0, sen0A, sen0B, sen0C),
                      spec_D14_lib=c(spec0, spec0A, spec0B, spec0C),
                      miscall_D14_lib=c(sum(tb0) - sum(diag(tb0)), sum(tb0A) - sum(diag(tb0A)), sum(tb0B) - sum(diag(tb0B)), sum(tb0C) - sum(diag(tb0C))),
                      confusion.matrix_D14_lib = c(paste(as.vector((c(tb0[1,], tb0[2,]))), collapse = " "),
                                                   paste(as.vector((c(tb0A[1,], tb0A[2,]))), collapse = " "),
                                                   paste(as.vector((c(tb0B[1,], tb0B[2,]))), collapse = " "),
                                                   paste(as.vector((c(tb0C[1,], tb0C[2,]))), collapse = " ")),
                      AUC_D14_D5=c(as.numeric(auc1),as.numeric(auc1A),as.numeric(auc1B),as.numeric(auc1C)),
                      sens_D14_D5=c(sen1, sen1A, sen1B, sen1C),
                      spec_D14_D5=c(spec1, spec1A, spec1B, spec1C),
                      miscall_D14_D5= c(sum(tb1) - sum(diag(tb1)), sum(tb1A) - sum(diag(tb1A)), sum(tb1B) - sum(diag(tb1B)), sum(tb1C) - sum(diag(tb1C))),
                      confusion.matrix_D14_D5=c(paste(as.vector((c(tb1[1,], tb1[2,]))), collapse = " "),
                                                paste(as.vector((c(tb1A[1,], tb1A[2,]))), collapse = " "),
                                                paste(as.vector((c(tb1B[1,], tb1B[2,]))), collapse = " "),
                                                paste(as.vector((c(tb1C[1,], tb1C[2,]))), collapse = " "))
)

write.csv(metrics, paste0(params$Output_dir, Gene, "_", Exn, "_", vset, "_", loessSubset, "_", EventCount_Filter, "_metrics.csv"), row.names = FALSE)
evalthis <- (length(filenames)>1)
```

### G. Scatter plot for replicates only

SNVs StopGain and Synomymous, and Missense (including splice region variants)

#### D14_Lib

```{r, eval = evalthis, include = TRUE}
test3 <- final1 %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>% filter(EventType %in% c("Missense", "Synonymous",  "StopGain")) %>% select(EventType, any_of(Rcol))

test3 <- test3[complete.cases(test3),]
etype <- unique(test3$EventType)
mycol3 <- c( "green",  "red", "blue")
check <- mycol3[match(test3$EventType, etype)]

##scatterplot3js(x = test3$R1, y = test3$R2, z = test3$R3,  color= check, size = 0.2, axisLabels=c("R1", "R2", "R3"))
if(length(filenames) == 2){
  fig <- plot_ly(test3, x = ~R1, y = ~R2, color = ~EventType, colors = mycol3, size = 0.5)
} 

if(length(filenames) == 3){
  fig <- plot_ly(test3, x = ~R1, y = ~R2, z = ~R3, color = ~EventType, colors = mycol3, size = 0.5)
}

if(length(filenames) == 4){
  fig1 <- plot_ly(test3, x = ~R1, y = ~R2, z = ~R3, color = ~EventType, colors = mycol3, size = 0.5)
  fig2 <- plot_ly(test3, x = ~R1, y = ~R3, z = ~R4, color = ~EventType, colors = mycol3, size = 0.5)
  fig3 <- plot_ly(test3, x = ~R2, y = ~R3, z = ~R4, color = ~EventType, colors = mycol3, size = 0.5)
}

if(length(filenames) == 4){
  fig1 %>% add_markers()
  fig2 %>% add_markers()
  fig3 %>% add_markers()
}else{
  fig <- fig %>% add_markers()

  fig

}
evalthis <- (length(filenames)>1)
```


#### D14_D5

```{r, eval = evalthis, include = TRUE}

test3 <- final2 %>% filter(REF %in% c("A", "T", "C", "G") & ALT %in% c("A", "T", "C", "G")) %>% filter(EventType %in% c("Missense", "Synonymous",  "StopGain"))  %>% select(EventType, any_of(Rcol))

test3 <- test3[complete.cases(test3),]
etype <- unique(test3$EventType)
mycol3 <- c( "green",  "red", "blue")
check <- mycol3[match(test3$EventType, etype)]

if(length(filenames) == 2){
  fig <- plot_ly(test3, x = ~R1, y = ~R2, color = ~EventType, colors = mycol3, size = 0.5)
} 

if(length(filenames) == 3){
  fig <- plot_ly(test3, x = ~R1, y = ~R2, z = ~R3, color = ~EventType, colors = mycol3, size = 0.5)
}

if(length(filenames) == 4){
  fig1 %>% add_markers()
  fig2 %>% add_markers()
  fig3 %>% add_markers()
}else{
  fig <- fig %>% add_markers()

  fig

}
```
